{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Existing Model Accuracy\n",
    "\n",
    "In this document, I explore the accuracies of various models that predict virus-host interactions solely from DNA sequences.\n",
    "\n",
    "To date (07/07/2022), I ran VirHostMatcher-Net, vHULK, CHERRY and RaFAH. I will explore the outputs of each oef those tools one a time. In addition, the input data used can be divided into 3 datasets:\n",
    "1. NCBI phages, which is a mix between collected host range from literature and interactions extracted from GenBank files\n",
    "2. Nahant Collection, a complete host range set between phages and Vibrio species\n",
    "3. Staph Study, also a complete host range set between phages and Staph species from different environments.\n",
    "\n",
    "When calculating the accuracy of the aforementioned tools, due to their approach in making the predictions (that is, predicting taxa rather than interaction), there will be 3 possibilities:\n",
    "1. The prediction is correct (predicted host = known host)\n",
    "2. The prediction is false (predicted host = known non-infection)\n",
    "3. Cannot determine if the prediction is correct or incorrect (the pair was not tested in the lab)\n",
    "\n",
    "\n",
    "## Setup\n",
    "\n",
    "First, let's load modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/gbastien/Documents/Research/Projects/ExistingVHPModels/analysis\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Metadata\n",
    "\n",
    "I also need to load some various metadata tables. For instance, depending on the tool, the naming system used to keep track of which virus infect which host may be different (using filenames versus accession number for example).\n",
    "\n",
    "Below I load metadata tables that I will be using throughout the document.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>filename</th>\n",
       "      <th>accession</th>\n",
       "      <th>phagename</th>\n",
       "      <th>phageGCAid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GCA_021090665.1_ASM2109066v1_genomic.fna</td>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>GCA_021090665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GCA_021090675.1_ASM2109067v1_genomic.fna</td>\n",
       "      <td>MZ417316.1</td>\n",
       "      <td>PG-2021_12</td>\n",
       "      <td>GCA_021090675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GCA_021090685.1_ASM2109068v1_genomic.fna</td>\n",
       "      <td>MZ417317.1</td>\n",
       "      <td>PG-2021_14</td>\n",
       "      <td>GCA_021090685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>GCA_021090695.1_ASM2109069v1_genomic.fna</td>\n",
       "      <td>MZ417318.1</td>\n",
       "      <td>PG-2021_15</td>\n",
       "      <td>GCA_021090695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>GCA_021090705.1_ASM2109070v1_genomic.fna</td>\n",
       "      <td>MZ417319.1</td>\n",
       "      <td>PG-2021_16</td>\n",
       "      <td>GCA_021090705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>GCA_021090715.1_ASM2109071v1_genomic.fna</td>\n",
       "      <td>MZ417320.1</td>\n",
       "      <td>PG-2021_17</td>\n",
       "      <td>GCA_021090715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>GCA_021090725.1_ASM2109072v1_genomic.fna</td>\n",
       "      <td>MZ417321.1</td>\n",
       "      <td>PG-2021_18</td>\n",
       "      <td>GCA_021090725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>GCA_021090735.1_ASM2109073v1_genomic.fna</td>\n",
       "      <td>MZ417322.1</td>\n",
       "      <td>PG-2021_19</td>\n",
       "      <td>GCA_021090735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>GCA_021090745.1_ASM2109074v1_genomic.fna</td>\n",
       "      <td>MZ417323.1</td>\n",
       "      <td>PG-2021_1</td>\n",
       "      <td>GCA_021090745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>GCA_021090755.1_ASM2109075v1_genomic.fna</td>\n",
       "      <td>MZ417324.1</td>\n",
       "      <td>PG-2021_22</td>\n",
       "      <td>GCA_021090755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>GCA_021090765.1_ASM2109076v1_genomic.fna</td>\n",
       "      <td>MZ417325.1</td>\n",
       "      <td>PG-2021_23</td>\n",
       "      <td>GCA_021090765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>GCA_021090775.1_ASM2109077v1_genomic.fna</td>\n",
       "      <td>MZ417326.1</td>\n",
       "      <td>PG-2021_27</td>\n",
       "      <td>GCA_021090775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>GCA_021090785.1_ASM2109078v1_genomic.fna</td>\n",
       "      <td>MZ417327.1</td>\n",
       "      <td>PG-2021_29</td>\n",
       "      <td>GCA_021090785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>GCA_021090795.1_ASM2109079v1_genomic.fna</td>\n",
       "      <td>MZ417328.1</td>\n",
       "      <td>PG-2021_2</td>\n",
       "      <td>GCA_021090795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>GCA_021090805.1_ASM2109080v1_genomic.fna</td>\n",
       "      <td>MZ417329.1</td>\n",
       "      <td>PG-2021_31</td>\n",
       "      <td>GCA_021090805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>GCA_021090815.1_ASM2109081v1_genomic.fna</td>\n",
       "      <td>MZ417330.1</td>\n",
       "      <td>PG-2021_33</td>\n",
       "      <td>GCA_021090815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>GCA_021090825.1_ASM2109082v1_genomic.fna</td>\n",
       "      <td>MZ417331.1</td>\n",
       "      <td>PG-2021_35</td>\n",
       "      <td>GCA_021090825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>GCA_021090835.1_ASM2109083v1_genomic.fna</td>\n",
       "      <td>MZ417332.1</td>\n",
       "      <td>PG-2021_38</td>\n",
       "      <td>GCA_021090835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>GCA_021090845.1_ASM2109084v1_genomic.fna</td>\n",
       "      <td>MZ417333.1</td>\n",
       "      <td>PG-2021_40</td>\n",
       "      <td>GCA_021090845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>GCA_021090855.1_ASM2109085v1_genomic.fna</td>\n",
       "      <td>MZ417334.1</td>\n",
       "      <td>PG-2021_41</td>\n",
       "      <td>GCA_021090855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>GCA_021090865.1_ASM2109086v1_genomic.fna</td>\n",
       "      <td>MZ417335.1</td>\n",
       "      <td>PG-2021_43</td>\n",
       "      <td>GCA_021090865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>GCA_021090875.1_ASM2109087v1_genomic.fna</td>\n",
       "      <td>MZ417336.1</td>\n",
       "      <td>PG-2021_46</td>\n",
       "      <td>GCA_021090875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>GCA_021090885.1_ASM2109088v1_genomic.fna</td>\n",
       "      <td>MZ417337.1</td>\n",
       "      <td>PG-2021_47</td>\n",
       "      <td>GCA_021090885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>GCA_021090895.1_ASM2109089v1_genomic.fna</td>\n",
       "      <td>MZ417338.1</td>\n",
       "      <td>PG-2021_4</td>\n",
       "      <td>GCA_021090895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>GCA_021090905.1_ASM2109090v1_genomic.fna</td>\n",
       "      <td>MZ417339.1</td>\n",
       "      <td>PG-2021_5</td>\n",
       "      <td>GCA_021090905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>GCA_021090915.1_ASM2109091v1_genomic.fna</td>\n",
       "      <td>MZ417340.1</td>\n",
       "      <td>PG-2021_64</td>\n",
       "      <td>GCA_021090915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>GCA_021090925.1_ASM2109092v1_genomic.fna</td>\n",
       "      <td>MZ417341.1</td>\n",
       "      <td>PG-2021_67</td>\n",
       "      <td>GCA_021090925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>GCA_021090935.1_ASM2109093v1_genomic.fna</td>\n",
       "      <td>MZ417342.1</td>\n",
       "      <td>PG-2021_68</td>\n",
       "      <td>GCA_021090935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>GCA_021090945.1_ASM2109094v1_genomic.fna</td>\n",
       "      <td>MZ417343.1</td>\n",
       "      <td>PG-2021_74</td>\n",
       "      <td>GCA_021090945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>GCA_021090955.1_ASM2109095v1_genomic.fna</td>\n",
       "      <td>MZ417344.1</td>\n",
       "      <td>PG-2021_76</td>\n",
       "      <td>GCA_021090955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>GCA_021090965.1_ASM2109096v1_genomic.fna</td>\n",
       "      <td>MZ417345.1</td>\n",
       "      <td>PG-2021_84</td>\n",
       "      <td>GCA_021090965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>GCA_021090975.1_ASM2109097v1_genomic.fna</td>\n",
       "      <td>MZ417346.1</td>\n",
       "      <td>PG-2021_86</td>\n",
       "      <td>GCA_021090975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>GCA_021090985.1_ASM2109098v1_genomic.fna</td>\n",
       "      <td>MZ417347.1</td>\n",
       "      <td>PG-2021_87</td>\n",
       "      <td>GCA_021090985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>GCA_021090995.1_ASM2109099v1_genomic.fna</td>\n",
       "      <td>MZ417348.1</td>\n",
       "      <td>PG-2021_88</td>\n",
       "      <td>GCA_021090995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>GCA_021091005.1_ASM2109100v1_genomic.fna</td>\n",
       "      <td>MZ417349.1</td>\n",
       "      <td>PG-2021_89</td>\n",
       "      <td>GCA_021091005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>GCA_021091015.1_ASM2109101v1_genomic.fna</td>\n",
       "      <td>MZ417350.1</td>\n",
       "      <td>PG-2021_8</td>\n",
       "      <td>GCA_021091015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>GCA_021091025.1_ASM2109102v1_genomic.fna</td>\n",
       "      <td>MZ417351.1</td>\n",
       "      <td>PG-2021_90</td>\n",
       "      <td>GCA_021091025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>GCA_021091035.1_ASM2109103v1_genomic.fna</td>\n",
       "      <td>MZ417352.1</td>\n",
       "      <td>PG-2021_91</td>\n",
       "      <td>GCA_021091035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>GCA_021091045.1_ASM2109104v1_genomic.fna</td>\n",
       "      <td>MZ417353.1</td>\n",
       "      <td>PG-2021_93</td>\n",
       "      <td>GCA_021091045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>GCA_021091055.1_ASM2109105v1_genomic.fna</td>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>GCA_021091055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0                                  filename   accession  \\\n",
       "0            0  GCA_021090665.1_ASM2109066v1_genomic.fna  MZ417315.1   \n",
       "1            1  GCA_021090675.1_ASM2109067v1_genomic.fna  MZ417316.1   \n",
       "2            2  GCA_021090685.1_ASM2109068v1_genomic.fna  MZ417317.1   \n",
       "3            3  GCA_021090695.1_ASM2109069v1_genomic.fna  MZ417318.1   \n",
       "4            4  GCA_021090705.1_ASM2109070v1_genomic.fna  MZ417319.1   \n",
       "5            5  GCA_021090715.1_ASM2109071v1_genomic.fna  MZ417320.1   \n",
       "6            6  GCA_021090725.1_ASM2109072v1_genomic.fna  MZ417321.1   \n",
       "7            7  GCA_021090735.1_ASM2109073v1_genomic.fna  MZ417322.1   \n",
       "8            8  GCA_021090745.1_ASM2109074v1_genomic.fna  MZ417323.1   \n",
       "9            9  GCA_021090755.1_ASM2109075v1_genomic.fna  MZ417324.1   \n",
       "10          10  GCA_021090765.1_ASM2109076v1_genomic.fna  MZ417325.1   \n",
       "11          11  GCA_021090775.1_ASM2109077v1_genomic.fna  MZ417326.1   \n",
       "12          12  GCA_021090785.1_ASM2109078v1_genomic.fna  MZ417327.1   \n",
       "13          13  GCA_021090795.1_ASM2109079v1_genomic.fna  MZ417328.1   \n",
       "14          14  GCA_021090805.1_ASM2109080v1_genomic.fna  MZ417329.1   \n",
       "15          15  GCA_021090815.1_ASM2109081v1_genomic.fna  MZ417330.1   \n",
       "16          16  GCA_021090825.1_ASM2109082v1_genomic.fna  MZ417331.1   \n",
       "17          17  GCA_021090835.1_ASM2109083v1_genomic.fna  MZ417332.1   \n",
       "18          18  GCA_021090845.1_ASM2109084v1_genomic.fna  MZ417333.1   \n",
       "19          19  GCA_021090855.1_ASM2109085v1_genomic.fna  MZ417334.1   \n",
       "20          20  GCA_021090865.1_ASM2109086v1_genomic.fna  MZ417335.1   \n",
       "21          21  GCA_021090875.1_ASM2109087v1_genomic.fna  MZ417336.1   \n",
       "22          22  GCA_021090885.1_ASM2109088v1_genomic.fna  MZ417337.1   \n",
       "23          23  GCA_021090895.1_ASM2109089v1_genomic.fna  MZ417338.1   \n",
       "24          24  GCA_021090905.1_ASM2109090v1_genomic.fna  MZ417339.1   \n",
       "25          25  GCA_021090915.1_ASM2109091v1_genomic.fna  MZ417340.1   \n",
       "26          26  GCA_021090925.1_ASM2109092v1_genomic.fna  MZ417341.1   \n",
       "27          27  GCA_021090935.1_ASM2109093v1_genomic.fna  MZ417342.1   \n",
       "28          28  GCA_021090945.1_ASM2109094v1_genomic.fna  MZ417343.1   \n",
       "29          29  GCA_021090955.1_ASM2109095v1_genomic.fna  MZ417344.1   \n",
       "30          30  GCA_021090965.1_ASM2109096v1_genomic.fna  MZ417345.1   \n",
       "31          31  GCA_021090975.1_ASM2109097v1_genomic.fna  MZ417346.1   \n",
       "32          32  GCA_021090985.1_ASM2109098v1_genomic.fna  MZ417347.1   \n",
       "33          33  GCA_021090995.1_ASM2109099v1_genomic.fna  MZ417348.1   \n",
       "34          34  GCA_021091005.1_ASM2109100v1_genomic.fna  MZ417349.1   \n",
       "35          35  GCA_021091015.1_ASM2109101v1_genomic.fna  MZ417350.1   \n",
       "36          36  GCA_021091025.1_ASM2109102v1_genomic.fna  MZ417351.1   \n",
       "37          37  GCA_021091035.1_ASM2109103v1_genomic.fna  MZ417352.1   \n",
       "38          38  GCA_021091045.1_ASM2109104v1_genomic.fna  MZ417353.1   \n",
       "39          39  GCA_021091055.1_ASM2109105v1_genomic.fna  MZ417354.1   \n",
       "\n",
       "     phagename     phageGCAid  \n",
       "0   PG-2021_10  GCA_021090665  \n",
       "1   PG-2021_12  GCA_021090675  \n",
       "2   PG-2021_14  GCA_021090685  \n",
       "3   PG-2021_15  GCA_021090695  \n",
       "4   PG-2021_16  GCA_021090705  \n",
       "5   PG-2021_17  GCA_021090715  \n",
       "6   PG-2021_18  GCA_021090725  \n",
       "7   PG-2021_19  GCA_021090735  \n",
       "8    PG-2021_1  GCA_021090745  \n",
       "9   PG-2021_22  GCA_021090755  \n",
       "10  PG-2021_23  GCA_021090765  \n",
       "11  PG-2021_27  GCA_021090775  \n",
       "12  PG-2021_29  GCA_021090785  \n",
       "13   PG-2021_2  GCA_021090795  \n",
       "14  PG-2021_31  GCA_021090805  \n",
       "15  PG-2021_33  GCA_021090815  \n",
       "16  PG-2021_35  GCA_021090825  \n",
       "17  PG-2021_38  GCA_021090835  \n",
       "18  PG-2021_40  GCA_021090845  \n",
       "19  PG-2021_41  GCA_021090855  \n",
       "20  PG-2021_43  GCA_021090865  \n",
       "21  PG-2021_46  GCA_021090875  \n",
       "22  PG-2021_47  GCA_021090885  \n",
       "23   PG-2021_4  GCA_021090895  \n",
       "24   PG-2021_5  GCA_021090905  \n",
       "25  PG-2021_64  GCA_021090915  \n",
       "26  PG-2021_67  GCA_021090925  \n",
       "27  PG-2021_68  GCA_021090935  \n",
       "28  PG-2021_74  GCA_021090945  \n",
       "29  PG-2021_76  GCA_021090955  \n",
       "30  PG-2021_84  GCA_021090965  \n",
       "31  PG-2021_86  GCA_021090975  \n",
       "32  PG-2021_87  GCA_021090985  \n",
       "33  PG-2021_88  GCA_021090995  \n",
       "34  PG-2021_89  GCA_021091005  \n",
       "35   PG-2021_8  GCA_021091015  \n",
       "36  PG-2021_90  GCA_021091025  \n",
       "37  PG-2021_91  GCA_021091035  \n",
       "38  PG-2021_93  GCA_021091045  \n",
       "39   PG-2021_9  GCA_021091055  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "StaphStudy_virusnames = pd.read_csv('../data/other/StaphStudy_virusnames.tsv', sep='\\t')\n",
    "\n",
    "# Add new column to contain the GCA filename of phage (only the GCA and numbers and nothing after it)\n",
    "filenames = StaphStudy_virusnames['filename'].tolist()\n",
    "GCAid = []\n",
    "for name in filenames:\n",
    "    GCAid.append(name.split('.')[0])\n",
    "StaphStudy_virusnames['phageGCAid'] = GCAid\n",
    "\n",
    "StaphStudy_virusnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGid</th>\n",
       "      <th>GCAid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MG592461.1</td>\n",
       "      <td>GCA_003927395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MG592607.1</td>\n",
       "      <td>GCA_003930195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MG592555.1</td>\n",
       "      <td>GCA_003929195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MG592500.1</td>\n",
       "      <td>GCA_003928155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MG592496.1</td>\n",
       "      <td>GCA_003928075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>MG592542.1</td>\n",
       "      <td>GCA_003928955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>MG592538.1</td>\n",
       "      <td>GCA_003928875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>MG592597.1</td>\n",
       "      <td>GCA_003929995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>MG592647.1</td>\n",
       "      <td>GCA_003930995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>MG592445.1</td>\n",
       "      <td>GCA_003852955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>283 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGid          GCAid\n",
       "0    MG592461.1  GCA_003927395\n",
       "1    MG592607.1  GCA_003930195\n",
       "2    MG592555.1  GCA_003929195\n",
       "3    MG592500.1  GCA_003928155\n",
       "4    MG592496.1  GCA_003928075\n",
       "..          ...            ...\n",
       "278  MG592542.1  GCA_003928955\n",
       "279  MG592538.1  GCA_003928875\n",
       "280  MG592597.1  GCA_003929995\n",
       "281  MG592647.1  GCA_003930995\n",
       "282  MG592445.1  GCA_003852955\n",
       "\n",
       "[283 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list sequence files\n",
    "datadir = '../data/sequences/phages/NahantCollection/'\n",
    "files = os.listdir(datadir)\n",
    "\n",
    "MGid = []\n",
    "GCAid = []\n",
    "\n",
    "# iterate through each file and get record id\n",
    "for file in files:\n",
    "    if file.endswith('.fasta'):\n",
    "        GCAid.append(file.split('.')[0])\n",
    "        loc = datadir + file\n",
    "        for record in SeqIO.parse(loc, 'fasta'):\n",
    "            MGid.append(record.id)\n",
    "\n",
    "# make dataframe\n",
    "NahantCollection_virusnames = pd.DataFrame()\n",
    "NahantCollection_virusnames['MGid'] = MGid\n",
    "NahantCollection_virusnames['GCAid'] = GCAid\n",
    "\n",
    "NahantCollection_virusnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Taxonomy\n",
    "\n",
    "Lastly, I need to load the taxonomy file for bacteria and archaea, which will be needed to determine the accuracy of the tools at different taxa level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus_species</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acaryochloris_marina</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Acaryochloridaceae</td>\n",
       "      <td>Acaryochloris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acetobacter_aceti</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirrllales</td>\n",
       "      <td>Acetobacteraceae</td>\n",
       "      <td>Acetobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acetobacter_ascendens</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirrllales</td>\n",
       "      <td>Acetobacteraceae</td>\n",
       "      <td>Acetobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acetobacter_orientalis</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirrllales</td>\n",
       "      <td>Acetobacteraceae</td>\n",
       "      <td>Acetobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acetobacter_oryzifermentans</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodospirrllales</td>\n",
       "      <td>Acetobacteraceae</td>\n",
       "      <td>Acetobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3278</th>\n",
       "      <td>Salinivibrio_costicola</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Salinivibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279</th>\n",
       "      <td>Enterovibrio_norvegicus</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterovibrio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3280</th>\n",
       "      <td>Phormidium_foveolarum</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Pseudanabaenales</td>\n",
       "      <td>Leptolyingbyaceae</td>\n",
       "      <td>Phormidium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3281</th>\n",
       "      <td>Faecalibacterium_prausnitzii</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Eubacteriales</td>\n",
       "      <td>Oscillosipiraceae</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3282</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhizobiales</td>\n",
       "      <td>Hyphomicrobiaceae</td>\n",
       "      <td>Candidatus Pelagibacter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3283 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Genus_species    Domain          Phylum  \\\n",
       "0             Acaryochloris_marina  Bacteria   Cyanobacteria   \n",
       "1                Acetobacter_aceti  Bacteria  Proteobacteria   \n",
       "2            Acetobacter_ascendens  Bacteria  Proteobacteria   \n",
       "3           Acetobacter_orientalis  Bacteria  Proteobacteria   \n",
       "4      Acetobacter_oryzifermentans  Bacteria  Proteobacteria   \n",
       "...                            ...       ...             ...   \n",
       "3278        Salinivibrio_costicola  Bacteria  Proteobacteria   \n",
       "3279       Enterovibrio_norvegicus  Bacteria  Proteobacteria   \n",
       "3280         Phormidium_foveolarum  Bacteria   Cyanobacteria   \n",
       "3281  Faecalibacterium_prausnitzii  Bacteria       Bacillota   \n",
       "3282                           NaN  Bacteria  Proteobacteria   \n",
       "\n",
       "                    Class             Order              Family  \\\n",
       "0           Cyanobacteria   Synechococcales  Acaryochloridaceae   \n",
       "1     Alphaproteobacteria  Rhodospirrllales    Acetobacteraceae   \n",
       "2     Alphaproteobacteria  Rhodospirrllales    Acetobacteraceae   \n",
       "3     Alphaproteobacteria  Rhodospirrllales    Acetobacteraceae   \n",
       "4     Alphaproteobacteria  Rhodospirrllales    Acetobacteraceae   \n",
       "...                   ...               ...                 ...   \n",
       "3278  Gammaproteobacteria       Vibrionales        Vibrionaceae   \n",
       "3279  Gammaproteobacteria       Vibrionales        Vibrionaceae   \n",
       "3280        Cyanobacteria  Pseudanabaenales   Leptolyingbyaceae   \n",
       "3281           Clostridia     Eubacteriales   Oscillosipiraceae   \n",
       "3282  Alphaproteobacteria       Rhizobiales   Hyphomicrobiaceae   \n",
       "\n",
       "                        Genus  \n",
       "0               Acaryochloris  \n",
       "1                 Acetobacter  \n",
       "2                 Acetobacter  \n",
       "3                 Acetobacter  \n",
       "4                 Acetobacter  \n",
       "...                       ...  \n",
       "3278             Salinivibrio  \n",
       "3279             Enterovibrio  \n",
       "3280               Phormidium  \n",
       "3281         Faecalibacterium  \n",
       "3282  Candidatus Pelagibacter  \n",
       "\n",
       "[3283 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxa = pd.read_csv('../data/other/bacteria_archaea_taxonomy.csv')\n",
    "taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Host Range Only\n",
    "\n",
    "I also need to extract only host range data. Currently, my NCBI data is a mix between host range data collected from literature, and host range extracted from virus GenBank files. The issue is that host range retrieved from virus GenBank files is the predominant data that was used to create the models that I am testing in this document. So the value of checking models' accuracies on data that was used to train the model is simply none.\n",
    "\n",
    "So in the following code chunk, I am extracting host range data that was solely collected from literature, rather than virus GenBank files!\n",
    "\n",
    "The best way to approach it to consider all viruses in the NCBI database. Then remove from the NCBI database any pair that are in the virus GenBank files.\n",
    "\n",
    "First, I am going to load all the host range data. This also include the Nahant Collection and the Staph Study.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# all host range data\n",
    "allHR = pd.read_csv('../data/hostrange/allhostrange.csv')\n",
    "\n",
    "# add taxa levels information\n",
    "Genus = []\n",
    "Family = []\n",
    "Order = []\n",
    "Class = []\n",
    "Phylum = []\n",
    "Domain = []\n",
    "\n",
    "for _, rows in allHR.iterrows():\n",
    "    genus_species = rows['hostname']\n",
    "    genus = genus_species.replace('_', ' ').split(' ')[0]\n",
    "    Genus.append(genus)\n",
    "\n",
    "    subset = taxa[taxa['Genus'] == genus]\n",
    "\n",
    "    Family.append(subset['Family'].values[0])\n",
    "    Order.append(subset['Order'].values[0])\n",
    "    Class.append(subset['Class'].values[0])\n",
    "    Phylum.append(subset['Phylum'].values[0])\n",
    "    Domain.append(subset['Domain'].values[0])\n",
    "\n",
    "allHR['genus'] = Genus\n",
    "allHR['family'] = Family\n",
    "allHR['order'] = Order\n",
    "allHR['class'] = Class\n",
    "allHR['phylum'] = Phylum\n",
    "allHR['domain'] = Domain\n",
    "\n",
    "# subset allHR to retrieve NCBI host range, Nahant Collection, and Staph Study\n",
    "NCBIHR = allHR[allHR['data'] == 'NCBI_HR']\n",
    "NahantCollectionHR = allHR[allHR['data'] == 'NahantCollection']\n",
    "StaphStudyHR = allHR[allHR['data'] == 'StaphStudy']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "So now I have the host range for each dataset. Now I am going to focus on formatting the NCBIHR so I only have host range from literature instead of GenBank files. In the past, I parsed the phage GenBank file so I am going to use that table. The resulting dataframe is named LiteratureHR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# read parsed GB file\n",
    "parsed_phageGB = pd.read_csv('../data/GenBank/NCBIphages_parsedGB.csv')\n",
    "\n",
    "# initialize two lists\n",
    "phagelist = []\n",
    "hostlist = []\n",
    "\n",
    "# iterate through the parsed_phageGB\n",
    "for _, row in parsed_phageGB.iterrows():\n",
    "    # get phage info\n",
    "    phage = row['accession']\n",
    "    # get phage's host info\n",
    "    temp = row['host'].split(' ')\n",
    "    if len(temp) < 2:\n",
    "        continue\n",
    "    host = temp[0] + '_' + temp[1]\n",
    "    # store phage and host in respective list\n",
    "    phagelist.append(phage)\n",
    "    hostlist.append(host)\n",
    "\n",
    "# create dataframe\n",
    "toremove = {phagelist[i]: hostlist[i] for i in range(len(phagelist))}\n",
    "\n",
    "# copy NCBI hr, then remove undesired pairs -- the ones extracted from GenBank files\n",
    "LiteratureHR = NCBIHR.copy()\n",
    "todrop = []\n",
    "for index, row in LiteratureHR.iterrows():\n",
    "    # check if phage and pair exist in dictionary. if it does, add index that will need to be removed\n",
    "    phage = row['phagename']\n",
    "    host = row['hostname']\n",
    "    if phage in toremove:\n",
    "        if host in toremove[phage]:\n",
    "            todrop.append(index)\n",
    "\n",
    "LiteratureHR.drop(todrop, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Let's see the resulting size of each of the dataframes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in NCBIHR is 2588\n",
      "Number of rows in LiteratureHR is 552\n",
      "Number of rows in NahantCollectionHR is 5208\n",
      "Number of rows in StaphStudyHR is 1053\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of rows in NCBIHR is {NCBIHR.shape[0]}')\n",
    "print(f'Number of rows in LiteratureHR is {LiteratureHR.shape[0]}')\n",
    "print(f'Number of rows in NahantCollectionHR is {NahantCollectionHR.shape[0]}')\n",
    "print(f'Number of rows in StaphStudyHR is {StaphStudyHR.shape[0]}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Functions\n",
    "\n",
    "All my functions to automate some coding will be here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Assessing virus-host predictions tools\n",
    "\n",
    "Before I start to determine the accuracy of each tool, I am going to create an empty dataframe to store the results. I will use this dataframe to plot the results!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# initialize dataframe\n",
    "results = pd.DataFrame(columns=['Tool', 'Dataset', 'Taxa_lvl', 'Correct', 'Wrong', 'Unknown'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "\n",
    "### VirHostMatcher-Net\n",
    "\n",
    "I am going to explore the accuracies of VirHostMatcher-Net. The output for VirHostMatcher-Net is a single file per prediction. So each phage will have their own individual prediction. The other interesting part of VirHostMatcher-Net is that it also return multiple potential hosts ranked from most likely to less likely. This will allow the following question: did VHM-net successfully predict the host after x amount of prediction? It's something to keep in mind for later, though for now I will just focus on the best prediction!\n",
    "\n",
    "\n",
    "#### Staph Study\n",
    "\n",
    "Let's start with the Staph Study, being the smallest dataset.\n",
    "\n",
    "When looking at one of the prediction file, the phage name is not included, but it is present in the filename.\n",
    "\n",
    "Below, I am formatting the resulting predictions!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# make a list of files\n",
    "datadir = '../data/predictions/VirHostMatcher-Net/StaphStudy_predictions/predictions/'\n",
    "files = os.listdir(datadir)\n",
    "\n",
    "# initialize dataframe\n",
    "VHMnet_StaphStudy_preds = pd.DataFrame()\n",
    "\n",
    "# load and stitch the files together\n",
    "for file in files:\n",
    "    # location of file\n",
    "    loc = datadir + file\n",
    "    # read file, extract phage name, add new column, store file in VHMnet_StaphStudy_preds\n",
    "    df = pd.read_csv(loc)\n",
    "    phage = file.split('.')[0]\n",
    "    df['phagename'] = phage\n",
    "\n",
    "    if VHMnet_StaphStudy_preds.empty:\n",
    "        VHMnet_StaphStudy_preds = df\n",
    "    else:\n",
    "        VHMnet_StaphStudy_preds = pd.concat([VHMnet_StaphStudy_preds, df])\n",
    "\n",
    "# reformat the hostname column to fit the format that I have in the host range\n",
    "hostnames = VHMnet_StaphStudy_preds['hostSpecies'].tolist()\n",
    "newnames = []\n",
    "for name in hostnames:\n",
    "    temp = name.split(' ')\n",
    "    string = temp[0] + '_' + temp[1]\n",
    "    newnames.append(string)\n",
    "VHMnet_StaphStudy_preds['hostSpecies'] = newnames\n",
    "\n",
    "phagenames = VHMnet_StaphStudy_preds['phagename']\n",
    "phagePGid = []\n",
    "for name in phagenames:\n",
    "    PGid = StaphStudy_virusnames.loc[StaphStudy_virusnames['phageGCAid'] == name, 'phagename'].values[0]\n",
    "    phagePGid.append(PGid)\n",
    "VHMnet_StaphStudy_preds['phagePGid'] = phagePGid\n",
    "\n",
    "# retrieve best predictions\n",
    "VHMnet_StaphStudy_bestpreds = VHMnet_StaphStudy_preds.iloc[::10, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "With the data stiched and formatted correctly, I can now calculate the accuracy of the tool at predicting hosts from the Staph Study host range, and I can also determine the accuracy at different taxa level!\n",
    "\n",
    "First, I will calculate the accuracy rate at the finest taxa resolution, which is at the species level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n",
      "9\n",
      "1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/394387278.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine if the predictions were correct/incorrect/unknown\n",
    "for _, row in VHMnet_StaphStudy_bestpreds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['hostSpecies']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine if correct, wrong, or unknown pred\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "\n",
    "# store results in dataframe\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Let's try the predictions at the genus level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38\n",
      "0\n",
      "1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/667673099.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/667673099.py:31: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/667673099.py:32: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/667673099.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : 39, 'Wrong': 0, 'Unknown': 0}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/667673099.py:34: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : 39, 'Wrong': 0, 'Unknown': 0}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_StaphStudy_bestpreds.iterrows():\n",
    "    phage = row['phagePGid']\n",
    "    hostpred_genus = row['hostSpecies'].split('_')[0]\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "\n",
    "# store results in dataframe\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : 39, 'Wrong': 0, 'Unknown': 0}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : 39, 'Wrong': 0, 'Unknown': 0}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Jumping from species level to genus level, the accuracy nearly becomes 100%. There is still 1 unknown at that level, which is Lactococcus (no lactococcus were tested in the Staph study). There is no reason to keep writing code to test more taxa level.\n",
    "\n",
    "#### Nahant Collection\n",
    "\n",
    "Let's do the Nahant Collection next!\n",
    "\n",
    "Sone phage predictions files had 11 predictions instead of 10 because of ties. So I used a different approach to retrieve the best predictions (see at the end of the code chunk)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3785272904.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  NahantCollectionHR['hostname'] = updated\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hostNCBIName</th>\n",
       "      <th>hostSuperkingdom</th>\n",
       "      <th>hostPhylum</th>\n",
       "      <th>hostClass</th>\n",
       "      <th>hostOrder</th>\n",
       "      <th>hostFamily</th>\n",
       "      <th>hostGenus</th>\n",
       "      <th>hostSpecies</th>\n",
       "      <th>hostName</th>\n",
       "      <th>score</th>\n",
       "      <th>...</th>\n",
       "      <th>posSV_pct</th>\n",
       "      <th>negSV_pct</th>\n",
       "      <th>crispr_pct</th>\n",
       "      <th>acc_phylum</th>\n",
       "      <th>acc_class</th>\n",
       "      <th>acc_order</th>\n",
       "      <th>acc_family</th>\n",
       "      <th>acc_genus</th>\n",
       "      <th>phagename</th>\n",
       "      <th>phageMGid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_001469215.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Alteromonadales</td>\n",
       "      <td>Pseudoalteromonadaceae</td>\n",
       "      <td>Pseudoalteromonas</td>\n",
       "      <td>Pseudoalteromonas_sp.</td>\n",
       "      <td>Pseudoalteromonas sp. H105</td>\n",
       "      <td>0.9737</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9725</td>\n",
       "      <td>0.9725</td>\n",
       "      <td>0.9471</td>\n",
       "      <td>0.9154</td>\n",
       "      <td>0.8541</td>\n",
       "      <td>GCA_003930595</td>\n",
       "      <td>MG592627.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GCF_000017705.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_campbellii</td>\n",
       "      <td>Vibrio campbellii ATCC BAA-1116</td>\n",
       "      <td>0.9021</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9617</td>\n",
       "      <td>0.9617</td>\n",
       "      <td>0.9291</td>\n",
       "      <td>0.8889</td>\n",
       "      <td>0.8084</td>\n",
       "      <td>GCA_003926295</td>\n",
       "      <td>MG592399.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GCF_001011015.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_parahaemolyticus</td>\n",
       "      <td>Vibrio parahaemolyticus</td>\n",
       "      <td>0.9426</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9722</td>\n",
       "      <td>0.9722</td>\n",
       "      <td>0.9404</td>\n",
       "      <td>0.9046</td>\n",
       "      <td>0.8310</td>\n",
       "      <td>GCA_003344165</td>\n",
       "      <td>MG592529.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>GCF_000354175.2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_alginolyticus</td>\n",
       "      <td>Vibrio alginolyticus NBRC 15630 = ATCC 17749</td>\n",
       "      <td>0.9593</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9735</td>\n",
       "      <td>0.9735</td>\n",
       "      <td>0.9429</td>\n",
       "      <td>0.9082</td>\n",
       "      <td>0.8469</td>\n",
       "      <td>GCA_003926635</td>\n",
       "      <td>MG592416.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>GCF_001469215.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Alteromonadales</td>\n",
       "      <td>Pseudoalteromonadaceae</td>\n",
       "      <td>Pseudoalteromonas</td>\n",
       "      <td>Pseudoalteromonas_sp.</td>\n",
       "      <td>Pseudoalteromonas sp. H105</td>\n",
       "      <td>0.9768</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9745</td>\n",
       "      <td>0.9745</td>\n",
       "      <td>0.9489</td>\n",
       "      <td>0.9170</td>\n",
       "      <td>0.8553</td>\n",
       "      <td>GCA_003929475</td>\n",
       "      <td>MG592570.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2785</th>\n",
       "      <td>GCF_000354175.2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_alginolyticus</td>\n",
       "      <td>Vibrio alginolyticus NBRC 15630 = ATCC 17749</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9740</td>\n",
       "      <td>0.9740</td>\n",
       "      <td>0.9501</td>\n",
       "      <td>0.9176</td>\n",
       "      <td>0.8590</td>\n",
       "      <td>GCA_003927315</td>\n",
       "      <td>MG592456.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2795</th>\n",
       "      <td>GCF_000017705.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_campbellii</td>\n",
       "      <td>Vibrio campbellii ATCC BAA-1116</td>\n",
       "      <td>0.9939</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>0.9998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9848</td>\n",
       "      <td>0.9848</td>\n",
       "      <td>0.9596</td>\n",
       "      <td>0.9343</td>\n",
       "      <td>0.8864</td>\n",
       "      <td>GCA_003930295</td>\n",
       "      <td>MG592612.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2805</th>\n",
       "      <td>GCF_000354175.2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_alginolyticus</td>\n",
       "      <td>Vibrio alginolyticus NBRC 15630 = ATCC 17749</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9830</td>\n",
       "      <td>0.9830</td>\n",
       "      <td>0.9558</td>\n",
       "      <td>0.9320</td>\n",
       "      <td>0.9116</td>\n",
       "      <td>GCA_003926595</td>\n",
       "      <td>MG592414.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2815</th>\n",
       "      <td>GCF_000354175.2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrio_alginolyticus</td>\n",
       "      <td>Vibrio alginolyticus NBRC 15630 = ATCC 17749</td>\n",
       "      <td>0.9960</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.9997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9854</td>\n",
       "      <td>0.9854</td>\n",
       "      <td>0.9592</td>\n",
       "      <td>0.9388</td>\n",
       "      <td>0.8950</td>\n",
       "      <td>GCA_003929835</td>\n",
       "      <td>MG592589.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2825</th>\n",
       "      <td>GCF_000169995.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Vibrionales</td>\n",
       "      <td>NAmissing</td>\n",
       "      <td>NAmissing</td>\n",
       "      <td>Vibrionales_bacterium</td>\n",
       "      <td>Vibrionales bacterium SWAT-3</td>\n",
       "      <td>0.9656</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9994</td>\n",
       "      <td>0.9990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>0.9731</td>\n",
       "      <td>0.9441</td>\n",
       "      <td>0.9089</td>\n",
       "      <td>0.8468</td>\n",
       "      <td>GCA_003927855</td>\n",
       "      <td>MG592485.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>283 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         hostNCBIName hostSuperkingdom      hostPhylum            hostClass  \\\n",
       "0     GCF_001469215.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "10    GCF_000017705.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "20    GCF_001011015.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "30    GCF_000354175.2         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "40    GCF_001469215.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "...               ...              ...             ...                  ...   \n",
       "2785  GCF_000354175.2         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "2795  GCF_000017705.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "2805  GCF_000354175.2         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "2815  GCF_000354175.2         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "2825  GCF_000169995.1         Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "\n",
       "            hostOrder              hostFamily          hostGenus  \\\n",
       "0     Alteromonadales  Pseudoalteromonadaceae  Pseudoalteromonas   \n",
       "10        Vibrionales            Vibrionaceae             Vibrio   \n",
       "20        Vibrionales            Vibrionaceae             Vibrio   \n",
       "30        Vibrionales            Vibrionaceae             Vibrio   \n",
       "40    Alteromonadales  Pseudoalteromonadaceae  Pseudoalteromonas   \n",
       "...               ...                     ...                ...   \n",
       "2785      Vibrionales            Vibrionaceae             Vibrio   \n",
       "2795      Vibrionales            Vibrionaceae             Vibrio   \n",
       "2805      Vibrionales            Vibrionaceae             Vibrio   \n",
       "2815      Vibrionales            Vibrionaceae             Vibrio   \n",
       "2825      Vibrionales               NAmissing          NAmissing   \n",
       "\n",
       "                  hostSpecies                                      hostName  \\\n",
       "0       Pseudoalteromonas_sp.                    Pseudoalteromonas sp. H105   \n",
       "10          Vibrio_campbellii               Vibrio campbellii ATCC BAA-1116   \n",
       "20    Vibrio_parahaemolyticus                       Vibrio parahaemolyticus   \n",
       "30       Vibrio_alginolyticus  Vibrio alginolyticus NBRC 15630 = ATCC 17749   \n",
       "40      Pseudoalteromonas_sp.                    Pseudoalteromonas sp. H105   \n",
       "...                       ...                                           ...   \n",
       "2785     Vibrio_alginolyticus  Vibrio alginolyticus NBRC 15630 = ATCC 17749   \n",
       "2795        Vibrio_campbellii               Vibrio campbellii ATCC BAA-1116   \n",
       "2805     Vibrio_alginolyticus  Vibrio alginolyticus NBRC 15630 = ATCC 17749   \n",
       "2815     Vibrio_alginolyticus  Vibrio alginolyticus NBRC 15630 = ATCC 17749   \n",
       "2825    Vibrionales_bacterium                  Vibrionales bacterium SWAT-3   \n",
       "\n",
       "       score  ...  posSV_pct  negSV_pct  crispr_pct  acc_phylum  acc_class  \\\n",
       "0     0.9737  ...     1.0000     0.9996         NaN      0.9725     0.9725   \n",
       "10    0.9021  ...     0.9999     0.9998         NaN      0.9617     0.9617   \n",
       "20    0.9426  ...     1.0000     0.9999         NaN      0.9722     0.9722   \n",
       "30    0.9593  ...     1.0000     0.9996         NaN      0.9735     0.9735   \n",
       "40    0.9768  ...     1.0000     0.9997         NaN      0.9745     0.9745   \n",
       "...      ...  ...        ...        ...         ...         ...        ...   \n",
       "2785  0.9809  ...     1.0000     0.9999         NaN      0.9740     0.9740   \n",
       "2795  0.9939  ...     0.9999     0.9998         NaN      0.9848     0.9848   \n",
       "2805  0.9972  ...     1.0000     0.9996         NaN      0.9830     0.9830   \n",
       "2815  0.9960  ...     1.0000     0.9997         NaN      0.9854     0.9854   \n",
       "2825  0.9656  ...     0.9994     0.9990         NaN      0.9731     0.9731   \n",
       "\n",
       "      acc_order  acc_family  acc_genus      phagename   phageMGid  \n",
       "0        0.9471      0.9154     0.8541  GCA_003930595  MG592627.1  \n",
       "10       0.9291      0.8889     0.8084  GCA_003926295  MG592399.1  \n",
       "20       0.9404      0.9046     0.8310  GCA_003344165  MG592529.1  \n",
       "30       0.9429      0.9082     0.8469  GCA_003926635  MG592416.1  \n",
       "40       0.9489      0.9170     0.8553  GCA_003929475  MG592570.1  \n",
       "...         ...         ...        ...            ...         ...  \n",
       "2785     0.9501      0.9176     0.8590  GCA_003927315  MG592456.1  \n",
       "2795     0.9596      0.9343     0.8864  GCA_003930295  MG592612.1  \n",
       "2805     0.9558      0.9320     0.9116  GCA_003926595  MG592414.1  \n",
       "2815     0.9592      0.9388     0.8950  GCA_003929835  MG592589.1  \n",
       "2825     0.9441      0.9089     0.8468  GCA_003927855  MG592485.1  \n",
       "\n",
       "[283 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make a list of files\n",
    "datadir = '../data/predictions/VirHostMatcher-Net/NahantCollection_predictions/predictions/'\n",
    "files = os.listdir(datadir)\n",
    "\n",
    "# intialize dataframe\n",
    "VHMnet_NahantCollection_preds = pd.DataFrame()\n",
    "\n",
    "# load and stitch the files together\n",
    "for file in files:\n",
    "    loc = datadir + file\n",
    "    df = pd.read_csv(loc)\n",
    "    phage = file.split('.')[0]\n",
    "    df['phagename'] = phage\n",
    "\n",
    "    if VHMnet_NahantCollection_preds.empty:\n",
    "        VHMnet_NahantCollection_preds = df\n",
    "    else:\n",
    "        VHMnet_NahantCollection_preds = pd.concat([VHMnet_NahantCollection_preds, df], ignore_index=True)\n",
    "\n",
    "# reformat the hostname column to fit the format that I have in the host range\n",
    "hostnames = VHMnet_NahantCollection_preds['hostSpecies'].tolist()\n",
    "newnames = []\n",
    "for name in hostnames:\n",
    "    temp = name.split(' ')\n",
    "    string = temp[0] + '_' + temp[1]\n",
    "    newnames.append(string)\n",
    "VHMnet_NahantCollection_preds['hostSpecies'] = newnames\n",
    "\n",
    "phagenames = VHMnet_NahantCollection_preds['phagename']\n",
    "phageMGid = []\n",
    "for name in phagenames:\n",
    "    MGid = NahantCollection_virusnames.loc[NahantCollection_virusnames['GCAid'] == name, 'MGid'].values[0]\n",
    "    phageMGid.append(MGid)\n",
    "VHMnet_NahantCollection_preds['phageMGid'] = phageMGid\n",
    "\n",
    "# reformat host range dataframe to have the _ between species and genus\n",
    "genus_species = NahantCollectionHR['hostname'].tolist()\n",
    "updated = []\n",
    "for host in genus_species:\n",
    "    temp = host.split(' ')\n",
    "    updated.append(temp[0] + '_' + temp[1])\n",
    "NahantCollectionHR['hostname'] = updated\n",
    "\n",
    "# retrieve best predictions for each phage\n",
    "seen = []\n",
    "tokeep = []\n",
    "for index, row in VHMnet_NahantCollection_preds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    if phage not in seen:\n",
    "        seen.append(phage)\n",
    "        tokeep.append(index)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "VHMnet_NahantCollection_bestpreds = VHMnet_NahantCollection_preds.iloc[tokeep]\n",
    "VHMnet_NahantCollection_bestpreds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Alright, time to determine the accuracy of VHM net on the Nahant Collection!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "139\n",
      "109\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/987085758.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# intialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine if the predictions were correct/incorrect/unknown\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostpred = row['hostSpecies']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range for the current phage\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine if correct, wrong, or unknown pred\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "184\n",
      "10\n",
      "54\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1781950459.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_genus = row['hostSpecies'].split('_')[0]\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "186\n",
      "8\n",
      "54\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3780411202.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_family = row['hostFamily']\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred_family not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "193\n",
      "31\n",
      "24\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2857586431.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_order = row['hostOrder']\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred_order not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "224"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/782480625.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "9\n",
      "15\n",
      "248\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_class = row['hostClass']\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostpred_class not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "226\n",
      "9\n",
      "13\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1604593755.py:30: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NahantCollection_bestpreds.iterrows():\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_phylum = row['hostPhylum']\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostpred_phylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interactions = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "            if 'Inf' in interactions:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interactions:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'VHMnet', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Literature NCBI\n",
    "\n",
    "At last I am doing the literature NCBI -- that is, host range collected from literature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# make a list of files\n",
    "datadir = '../data/predictions/VirHostMatcher-Net/NCBIphages_predictions/predictions/'\n",
    "files = os.listdir(datadir)\n",
    "\n",
    "# intiialize dataframe\n",
    "VHMnet_NCBI_preds = pd.DataFrame()\n",
    "\n",
    "# load and stitch the files together\n",
    "for file in files:\n",
    "    loc = datadir + file\n",
    "    df = pd.read_csv(loc)\n",
    "    phage = file.split('.')[0]\n",
    "    df['phagename'] = phage\n",
    "\n",
    "    if VHMnet_NCBI_preds.empty:\n",
    "        VHMnet_NCBI_preds = df\n",
    "    else:\n",
    "        VHMnet_NCBI_preds = pd.concat([VHMnet_NCBI_preds, df], ignore_index=True)\n",
    "\n",
    "# reformat the hostname column to fit the format that I have in the host range\n",
    "hostnames = VHMnet_NCBI_preds['hostSpecies'].tolist()\n",
    "newnames = []\n",
    "for name in hostnames:\n",
    "    temp = name.split(' ')\n",
    "    string = temp[0] + '_' + temp[1]\n",
    "    newnames.append(string)\n",
    "VHMnet_NCBI_preds['hostSpecies'] = newnames\n",
    "\n",
    "# retrieve best predictions\n",
    "seen = []\n",
    "tokeep = []\n",
    "for index, row in VHMnet_NCBI_preds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    if phage not in seen:\n",
    "        seen.append(phage)\n",
    "        tokeep.append(index)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "VHMnet_NCBI_bestpreds = VHMnet_NCBI_preds.iloc[tokeep]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Determine accuracy of the tool at the species level!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "7\n",
      "75\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# intialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine if the predictions were correct/incorrect/unknown\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagename']\n",
    "    hostpred = row['hostSpecies']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['hostname'].tolist()\n",
    "            if hostpred not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n",
      "29\n",
      "43\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_genus = row['hostGenus']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['genus'].tolist()\n",
    "            if hostpred_genus not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "31\n",
      "36\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_family = row['hostFamily']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['family'].tolist()\n",
    "            if hostpred_family not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n",
      "38\n",
      "24\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_order = row['hostOrder']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['order'].tolist()\n",
    "            if hostpred_order not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "43\n",
      "14\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_class = row['hostClass']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['class'].tolist()\n",
    "            if hostpred_class not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "45\n",
      "12\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_phylum = row['hostPhylum']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['phylum'].tolist()\n",
    "            if hostpred_phylum not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35\n",
      "50\n",
      "0\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through dataframe\n",
    "for _, row in VHMnet_NCBI_bestpreds.iterrows():\n",
    "    phage = row['phagename']\n",
    "    hostpred_domain = row['hostSuperkingdom']\n",
    "\n",
    "    # subset host range dataframe for current phage\n",
    "    if phage not in LiteratureHR['phagename'].tolist():\n",
    "        continue\n",
    "    else:\n",
    "        subset = LiteratureHR[LiteratureHR['phagename'] == phage]\n",
    "\n",
    "        if subset.empty:\n",
    "            continue\n",
    "        else:\n",
    "            known_hosts = subset['domain'].tolist()\n",
    "            if hostpred_domain not in known_hosts:\n",
    "                unknown += 1\n",
    "            else:\n",
    "                interaction = subset.loc[subset['domain'] == hostpred_domain, 'infection'].tolist()\n",
    "                if 'Inf' in interaction:\n",
    "                    correct += 1\n",
    "                elif 'NoInf' in interaction:\n",
    "                    wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Summary and plots\n",
    "The bar plots were done in R. Check RStudio files in the same directory as this files for the plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# TODO: Bar plots of accuracies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### vHULK\n",
    "\n",
    "Now I will move on to vHULK. Same thing, I will start with the Staph Study. This tool returned a single file for all the predictions for all phages of interest. It returns two predictions, at the species level and at the genus level.\n",
    "\n",
    "\n",
    "#### Staph Study\n",
    "\n",
    "Let's read the first csv file which contains all the predictions for the Staph Study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "vHULK_StaphStudy_preds = pd.read_csv('../data/predictions/vHULK/vHULK_StaphStudy_predictions.csv')\n",
    "\n",
    "# reformat the hostname column to fit the format that I have in the host range dataset\n",
    "phagenames = vHULK_StaphStudy_preds['BIN/genome'].tolist()\n",
    "newnames = []\n",
    "for name in phagenames:\n",
    "    GCAid = name + '.fna'\n",
    "    PGid = StaphStudy_virusnames.loc[StaphStudy_virusnames['filename'] == GCAid, 'phagename'].values[0]\n",
    "    newnames.append(PGid)\n",
    "vHULK_StaphStudy_preds['phagePGid'] = newnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "File is formatted and ready to go! Now let's calculate vHULK accuracy at the species level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n",
      "10\n",
      "0\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/888182815.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['final_prediction']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Let's do at the genus level next!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39\n",
      "0\n",
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3920379083.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3920379083.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3920379083.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3920379083.py:38: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3920379083.py:39: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred_genus = row['final_prediction'].split('_')[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "'100% accuracy at the genus level!\n",
    "\n",
    "#### Nahant Collection\n",
    "\n",
    "Now let's move on to the Nahant collection.\n",
    "\n",
    "First thing, let's read the prediction file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "vHULK_NahantCollection_preds = pd.read_csv('../data/predictions/vHULK/vHULK_NahantCollection_predictions.csv')\n",
    "\n",
    "# reformat the hostname column to fit the format that I have in the host range dataset\n",
    "phagenames = vHULK_NahantCollection_preds['BIN/genome'].tolist()\n",
    "newnames = []\n",
    "for name in phagenames:\n",
    "    GCAid = name.split('.')[0]\n",
    "    MGid = NahantCollection_virusnames.loc[NahantCollection_virusnames['GCAid'] == GCAid, 'MGid'].values[0]\n",
    "    newnames.append(MGid)\n",
    "vHULK_NahantCollection_preds['phageMGid'] = newnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Prediction file is formatted and good to go! Interestingly, the column final_prediction contains many 'None' prediction. So I guess I will use the pred_species_softmax instead? Not sure how to score/assess the tool otherwise.\n",
    "\n",
    "\n",
    "\n",
    "Let's start with determining accuracy at the species level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "55\n",
      "193\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2246082667.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostpred = row['pred_species_softmax']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57\n",
      "11\n",
      "180\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1968442726.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostpred_genus = row['pred_species_softmax'].split('_')[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n",
      "7\n",
      "180\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1032444566.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostgenus = row['pred_species_softmax'].split('_')[0]\n",
    "    hostpred_family = taxa.loc[taxa['Genus'] == hostgenus, 'Family'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred_family not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Order level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n",
      "7\n",
      "180\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/266898483.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostgenus = row['pred_species_softmax'].split('_')[0]\n",
    "    hostpred_order = taxa.loc[taxa['Genus'] == hostgenus, 'Order'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred_order not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Class level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "10\n",
      "13\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1299406411.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostgenus = row['pred_species_softmax'].split('_')[0]\n",
    "    hostpred_class = taxa.loc[taxa['Genus'] == hostgenus, 'Class'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostpred_class not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Phylum level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "226\n",
      "10\n",
      "12\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2242359109.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in vHULK_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phageMGid']\n",
    "    hostgenus = row['pred_species_softmax'].split('_')[0]\n",
    "    hostpred_phylum = taxa.loc[taxa['Genus'] == hostgenus, 'Phylum'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostpred_phylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'vHULK', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### CHERRY\n",
    "\n",
    "Let's move on the CHERRY. As usual, let's star twith the Staph Study! Same thing as vHULK, all the predictions were stored in a single file.\n",
    "\n",
    "#### Staph Study\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "CHERRY_StaphStudy_preds = pd.read_csv('../data/predictions/CHERRY/CHERRYpredictions_StaphStudy.csv')\n",
    "\n",
    "# reformat the hostname column to fit the format that I have\n",
    "hostnames = CHERRY_StaphStudy_preds['Top_1_label'].tolist()\n",
    "newnames = []\n",
    "for name in hostnames:\n",
    "    split = name.split(' ')\n",
    "    newnames.append(split[0] + '_' + split[1])\n",
    "CHERRY_StaphStudy_preds['hostprediction'] = newnames\n",
    "\n",
    "# reformat phage name to fit the format that I have\n",
    "phagenames = CHERRY_StaphStudy_preds['contig_name']\n",
    "newnames = []\n",
    "for name in phagenames:\n",
    "    PGid = StaphStudy_virusnames.loc[StaphStudy_virusnames['accession'] == name, 'phagename'].values[0]\n",
    "    newnames.append(PGid)\n",
    "CHERRY_StaphStudy_preds['phagePGid'] = newnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "File formatted and I am ready to determine their accuracy at different taxa level!\n",
    "\n",
    "Let's start with species level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\n",
      "3\n",
      "22\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/752492902.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['hostprediction']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "0\n",
      "22\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/754444028.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred_genus = row['hostprediction'].split('_')[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "0\n",
      "22\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2400072594.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_family = taxa.loc[taxa['Genus'] == hostgenus, 'Family'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred_family not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "0\n",
      "22\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2562228612.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_order = taxa.loc[taxa['Genus'] == hostgenus, 'Order'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred_order not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Class level\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/349774287.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0\n",
      "22\n",
      "39\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_class = taxa.loc[taxa['Genus'] == hostgenus, 'Class'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostpred_class not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Phylum level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34\n",
      "0\n",
      "5\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3158824013.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_phylum = taxa.loc[taxa['Genus'] == hostgenus, 'Phylum'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostpred_phylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Nahant Collection\n",
    "\n",
    "Let's move on to the Nahant Collection!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "CHERRY_NahantCollection_preds = pd.read_csv('../data/predictions/CHERRY/CHERRYpredictions_NahantCollection.csv')\n",
    "\n",
    "# reformat the hostname column to fit the format that I have\n",
    "hostnames = CHERRY_NahantCollection_preds['Top_1_label'].tolist()\n",
    "newnames = []\n",
    "for name in hostnames:\n",
    "    split = name.split(' ')\n",
    "    newnames.append(split[0] + '_' + split[1])\n",
    "CHERRY_NahantCollection_preds['hostprediction'] = newnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Species level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n",
      "97\n",
      "145\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/3937366512.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostpred = row['hostprediction']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "117\n",
      "21\n",
      "110\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2564772484.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostpred_genus = row['hostprediction'].split('_')[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred_genus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred_genus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "128\n",
      "10\n",
      "110\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/224799748.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_family = taxa.loc[taxa['Genus'] == hostgenus, 'Family'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred_family not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Order taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "128\n",
      "37\n",
      "83\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/2025759894.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_order = taxa.loc[taxa['Genus'] == hostgenus, 'Order'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred_order not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Class taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "222\n",
      "9\n",
      "17\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/1171195113.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_class = taxa.loc[taxa['Genus'] == hostgenus, 'Class'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostpred_class not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Phylum level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "224\n",
      "9\n",
      "15\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_91984/696753010.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in CHERRY_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['contig_name']\n",
    "    hostgenus = row['hostprediction'].split('_')[0]\n",
    "    hostpred_phylum = taxa.loc[taxa['Genus'] == hostgenus, 'Phylum'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostpred_phylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'CHERRY', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### RAFAH\n",
    "\n",
    "This is the last tool I ran. Same thing, let's explore its accuracy rate in predicting taxa for the Staph Study and the Nahant Collection. There are no predictions at the species level, rather it starts at the genus level.\n",
    "\n",
    "#### Staph Study\n",
    "\n",
    "Load file and format it as needed!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "RAFAH_StaphStudy_preds = pd.read_csv('../data/predictions/RaFAH/RaFAH_StaphStudy_Host_Predictions.tsv', sep='\\t')\n",
    "\n",
    "# reformat the phage name to fit the format that I have\n",
    "phagenames = RAFAH_StaphStudy_preds['Sequence']\n",
    "newnames = []\n",
    "for name in phagenames:\n",
    "    PGid = StaphStudy_virusnames.loc[StaphStudy_virusnames['accession'] == name, 'phagename'].values[0]\n",
    "    newnames.append(PGid)\n",
    "RAFAH_StaphStudy_preds['phagePGid'] = newnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Predicted_Host</th>\n",
       "      <th>Winner_Score</th>\n",
       "      <th>Abiotrophia</th>\n",
       "      <th>Acetobacter</th>\n",
       "      <th>Acholeplasma</th>\n",
       "      <th>Achromobacter</th>\n",
       "      <th>Acidaminococcus</th>\n",
       "      <th>Acidianus</th>\n",
       "      <th>Acidiferrobacter</th>\n",
       "      <th>...</th>\n",
       "      <th>Xenorhabdus</th>\n",
       "      <th>Xylanimonas</th>\n",
       "      <th>Xylella</th>\n",
       "      <th>Yersinia</th>\n",
       "      <th>Yokenella</th>\n",
       "      <th>Youngiibacter</th>\n",
       "      <th>Zhongshania</th>\n",
       "      <th>Zymomonas</th>\n",
       "      <th>Zymophilus</th>\n",
       "      <th>phagePGid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MZ417316.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MZ417317.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MZ417318.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MZ417319.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.994</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MZ417320.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MZ417321.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MZ417322.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.859</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MZ417323.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.949</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MZ417324.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MZ417325.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.907</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MZ417326.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.852</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MZ417327.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.976</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MZ417328.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MZ417329.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.974</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MZ417330.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.974</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MZ417331.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.987</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MZ417332.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MZ417333.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.948</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MZ417334.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MZ417335.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MZ417336.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.873</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MZ417337.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MZ417338.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MZ417339.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.877</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MZ417340.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MZ417341.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.807</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MZ417342.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.812</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MZ417343.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.877</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MZ417344.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.974</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MZ417345.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.994</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MZ417346.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.968</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MZ417347.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.771</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MZ417348.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.877</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MZ417349.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.985</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MZ417350.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MZ417351.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.996</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MZ417352.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MZ417353.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.966</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PG-2021_9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40 rows × 621 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Sequence  Predicted_Host  Winner_Score  Abiotrophia  Acetobacter  \\\n",
       "0   MZ417315.1  Staphylococcus         0.995            0            0   \n",
       "1   MZ417316.1  Staphylococcus         0.986            0            0   \n",
       "2   MZ417317.1  Staphylococcus         0.986            0            0   \n",
       "3   MZ417318.1  Staphylococcus         0.986            0            0   \n",
       "4   MZ417319.1  Staphylococcus         0.994            0            0   \n",
       "5   MZ417320.1  Staphylococcus         0.995            0            0   \n",
       "6   MZ417321.1  Staphylococcus         0.961            0            0   \n",
       "7   MZ417322.1  Staphylococcus         0.859            0            0   \n",
       "8   MZ417323.1  Staphylococcus         0.949            0            0   \n",
       "9   MZ417324.1  Staphylococcus         0.986            0            0   \n",
       "10  MZ417325.1  Staphylococcus         0.907            0            0   \n",
       "11  MZ417326.1  Staphylococcus         0.852            0            0   \n",
       "12  MZ417327.1  Staphylococcus         0.976            0            0   \n",
       "13  MZ417328.1  Staphylococcus         0.983            0            0   \n",
       "14  MZ417329.1  Staphylococcus         0.974            0            0   \n",
       "15  MZ417330.1  Staphylococcus         0.974            0            0   \n",
       "16  MZ417331.1  Staphylococcus         0.987            0            0   \n",
       "17  MZ417332.1  Staphylococcus         0.990            0            0   \n",
       "18  MZ417333.1  Staphylococcus         0.948            0            0   \n",
       "19  MZ417334.1  Staphylococcus         0.986            0            0   \n",
       "20  MZ417335.1  Staphylococcus         0.986            0            0   \n",
       "21  MZ417336.1  Staphylococcus         0.873            0            0   \n",
       "22  MZ417337.1  Staphylococcus         0.993            0            0   \n",
       "23  MZ417338.1  Staphylococcus         0.808            0            0   \n",
       "24  MZ417339.1  Staphylococcus         0.877            0            0   \n",
       "25  MZ417340.1  Staphylococcus         0.983            0            0   \n",
       "26  MZ417341.1  Staphylococcus         0.807            0            0   \n",
       "27  MZ417342.1  Staphylococcus         0.812            0            0   \n",
       "28  MZ417343.1  Staphylococcus         0.877            0            0   \n",
       "29  MZ417344.1  Staphylococcus         0.974            0            0   \n",
       "30  MZ417345.1  Staphylococcus         0.994            0            0   \n",
       "31  MZ417346.1  Staphylococcus         0.968            0            0   \n",
       "32  MZ417347.1  Staphylococcus         0.771            0            0   \n",
       "33  MZ417348.1  Staphylococcus         0.877            0            0   \n",
       "34  MZ417349.1  Staphylococcus         0.985            0            0   \n",
       "35  MZ417350.1  Staphylococcus         0.983            0            0   \n",
       "36  MZ417351.1  Staphylococcus         0.996            0            0   \n",
       "37  MZ417352.1  Staphylococcus         0.993            0            0   \n",
       "38  MZ417353.1  Staphylococcus         0.966            0            0   \n",
       "39  MZ417354.1  Staphylococcus         0.995            0            0   \n",
       "\n",
       "    Acholeplasma  Achromobacter  Acidaminococcus  Acidianus  Acidiferrobacter  \\\n",
       "0              0          0.000                0          0                 0   \n",
       "1              0          0.000                0          0                 0   \n",
       "2              0          0.000                0          0                 0   \n",
       "3              0          0.000                0          0                 0   \n",
       "4              0          0.000                0          0                 0   \n",
       "5              0          0.000                0          0                 0   \n",
       "6              0          0.000                0          0                 0   \n",
       "7              0          0.001                0          0                 0   \n",
       "8              0          0.000                0          0                 0   \n",
       "9              0          0.000                0          0                 0   \n",
       "10             0          0.000                0          0                 0   \n",
       "11             0          0.000                0          0                 0   \n",
       "12             0          0.000                0          0                 0   \n",
       "13             0          0.000                0          0                 0   \n",
       "14             0          0.000                0          0                 0   \n",
       "15             0          0.000                0          0                 0   \n",
       "16             0          0.000                0          0                 0   \n",
       "17             0          0.000                0          0                 0   \n",
       "18             0          0.000                0          0                 0   \n",
       "19             0          0.000                0          0                 0   \n",
       "20             0          0.000                0          0                 0   \n",
       "21             0          0.000                0          0                 0   \n",
       "22             0          0.000                0          0                 0   \n",
       "23             0          0.000                0          0                 0   \n",
       "24             0          0.000                0          0                 0   \n",
       "25             0          0.000                0          0                 0   \n",
       "26             0          0.000                0          0                 0   \n",
       "27             0          0.000                0          0                 0   \n",
       "28             0          0.000                0          0                 0   \n",
       "29             0          0.000                0          0                 0   \n",
       "30             0          0.000                0          0                 0   \n",
       "31             0          0.000                0          0                 0   \n",
       "32             0          0.000                0          0                 0   \n",
       "33             0          0.000                0          0                 0   \n",
       "34             0          0.000                0          0                 0   \n",
       "35             0          0.000                0          0                 0   \n",
       "36             0          0.000                0          0                 0   \n",
       "37             0          0.000                0          0                 0   \n",
       "38             0          0.000                0          0                 0   \n",
       "39             0          0.000                0          0                 0   \n",
       "\n",
       "    ...  Xenorhabdus  Xylanimonas  Xylella  Yersinia  Yokenella  \\\n",
       "0   ...            0        0.000        0     0.000          0   \n",
       "1   ...            0        0.000        0     0.000          0   \n",
       "2   ...            0        0.000        0     0.000          0   \n",
       "3   ...            0        0.000        0     0.000          0   \n",
       "4   ...            0        0.000        0     0.000          0   \n",
       "5   ...            0        0.000        0     0.000          0   \n",
       "6   ...            0        0.000        0     0.001          0   \n",
       "7   ...            0        0.000        0     0.004          0   \n",
       "8   ...            0        0.000        0     0.000          0   \n",
       "9   ...            0        0.000        0     0.000          0   \n",
       "10  ...            0        0.000        0     0.000          0   \n",
       "11  ...            0        0.000        0     0.000          0   \n",
       "12  ...            0        0.000        0     0.000          0   \n",
       "13  ...            0        0.000        0     0.000          0   \n",
       "14  ...            0        0.000        0     0.001          0   \n",
       "15  ...            0        0.000        0     0.001          0   \n",
       "16  ...            0        0.000        0     0.000          0   \n",
       "17  ...            0        0.000        0     0.000          0   \n",
       "18  ...            0        0.000        0     0.000          0   \n",
       "19  ...            0        0.000        0     0.000          0   \n",
       "20  ...            0        0.000        0     0.000          0   \n",
       "21  ...            0        0.000        0     0.000          0   \n",
       "22  ...            0        0.000        0     0.000          0   \n",
       "23  ...            0        0.001        0     0.000          0   \n",
       "24  ...            0        0.000        0     0.000          0   \n",
       "25  ...            0        0.000        0     0.000          0   \n",
       "26  ...            0        0.001        0     0.000          0   \n",
       "27  ...            0        0.001        0     0.000          0   \n",
       "28  ...            0        0.000        0     0.000          0   \n",
       "29  ...            0        0.000        0     0.001          0   \n",
       "30  ...            0        0.000        0     0.000          0   \n",
       "31  ...            0        0.000        0     0.007          0   \n",
       "32  ...            0        0.000        0     0.008          0   \n",
       "33  ...            0        0.000        0     0.000          0   \n",
       "34  ...            0        0.000        0     0.000          0   \n",
       "35  ...            0        0.000        0     0.000          0   \n",
       "36  ...            0        0.000        0     0.000          0   \n",
       "37  ...            0        0.000        0     0.000          0   \n",
       "38  ...            0        0.000        0     0.000          0   \n",
       "39  ...            0        0.000        0     0.000          0   \n",
       "\n",
       "    Youngiibacter  Zhongshania  Zymomonas  Zymophilus   phagePGid  \n",
       "0               0            0          0           0  PG-2021_10  \n",
       "1               0            0          0           0  PG-2021_12  \n",
       "2               0            0          0           0  PG-2021_14  \n",
       "3               0            0          0           0  PG-2021_15  \n",
       "4               0            0          0           0  PG-2021_16  \n",
       "5               0            0          0           0  PG-2021_17  \n",
       "6               0            0          0           0  PG-2021_18  \n",
       "7               0            0          0           0  PG-2021_19  \n",
       "8               0            0          0           0   PG-2021_1  \n",
       "9               0            0          0           0  PG-2021_22  \n",
       "10              0            0          0           0  PG-2021_23  \n",
       "11              0            0          0           0  PG-2021_27  \n",
       "12              0            0          0           0  PG-2021_29  \n",
       "13              0            0          0           0   PG-2021_2  \n",
       "14              0            0          0           0  PG-2021_31  \n",
       "15              0            0          0           0  PG-2021_33  \n",
       "16              0            0          0           0  PG-2021_35  \n",
       "17              0            0          0           0  PG-2021_38  \n",
       "18              0            0          0           0  PG-2021_40  \n",
       "19              0            0          0           0  PG-2021_41  \n",
       "20              0            0          0           0  PG-2021_43  \n",
       "21              0            0          0           0  PG-2021_46  \n",
       "22              0            0          0           0  PG-2021_47  \n",
       "23              0            0          0           0   PG-2021_4  \n",
       "24              0            0          0           0   PG-2021_5  \n",
       "25              0            0          0           0  PG-2021_64  \n",
       "26              0            0          0           0  PG-2021_67  \n",
       "27              0            0          0           0  PG-2021_68  \n",
       "28              0            0          0           0  PG-2021_74  \n",
       "29              0            0          0           0  PG-2021_76  \n",
       "30              0            0          0           0  PG-2021_84  \n",
       "31              0            0          0           0  PG-2021_86  \n",
       "32              0            0          0           0  PG-2021_87  \n",
       "33              0            0          0           0  PG-2021_88  \n",
       "34              0            0          0           0  PG-2021_89  \n",
       "35              0            0          0           0   PG-2021_8  \n",
       "36              0            0          0           0  PG-2021_90  \n",
       "37              0            0          0           0  PG-2021_91  \n",
       "38              0            0          0           0  PG-2021_93  \n",
       "39              0            0          0           0   PG-2021_9  \n",
       "\n",
       "[40 rows x 621 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RAFAH_StaphStudy_preds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Genus species level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39\n",
      "0\n",
      "0\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/530910736.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/530910736.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/530910736.py:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/530910736.py:38: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/530910736.py:39: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostgenus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostgenus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Nahant Collection\n",
    "\n",
    "Load file and format it as necessary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# load file\n",
    "RAFAH_NahantCollection_preds = pd.read_csv('../data/predictions/RaFAH/RaFAH_NahantCollection_Host_Predictions.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "84\n",
      "11\n",
      "153\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2355729214.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Sequence']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostgenus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostgenus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Family taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n",
      "7\n",
      "153\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2521979754.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Sequence']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "    hostpred_family = taxa.loc[taxa['Genus'] == hostgenus, 'Family'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred_family not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred_family, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Order taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n",
      "31\n",
      "129\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2217446707.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Sequence']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "    hostpred_order = taxa.loc[taxa['Genus'] == hostgenus, 'Order'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred_order not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostpred_order, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Class taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "204\n",
      "8\n",
      "36\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2438883843.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Sequence']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "    hostpred_class = taxa.loc[taxa['Genus'] == hostgenus, 'Class'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostpred_class not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostpred_class, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Phylum taxa level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "230\n",
      "8\n",
      "10\n",
      "248\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3323951267.py:36: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in RAFAH_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Sequence']\n",
    "    hostgenus = row['Predicted_Host']\n",
    "    hostpred_phylum = taxa.loc[taxa['Genus'] == hostgenus, 'Phylum'].values[0]\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostpred_phylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostpred_phylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'RAFAH', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### iPHoP (considering all predictions)\n",
    "\n",
    "10/02/2022 - I tried a new tool to predict host. This is the most recent tool to date that uses other tools to predict host.\n",
    "\n",
    "Only going to focus on the Nahant Collection and the Staph Study since I figured out those are the only one interested in predicting for. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Staph Study\n",
    "\n",
    "Let's load the prediction file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus</th>\n",
       "      <th>Host genome</th>\n",
       "      <th>Host taxonomy</th>\n",
       "      <th>Main method</th>\n",
       "      <th>Confidence score</th>\n",
       "      <th>Additional methods</th>\n",
       "      <th>phagePGid</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>RS_GCF_002902385.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>98.1</td>\n",
       "      <td>iPHoP-RF;97.10</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>Staphylococcus intermedius</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>RS_GCF_900636325.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>97.6</td>\n",
       "      <td>iPHoP-RF;97.10</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>Staphylococcus delphini</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>RS_GCF_000010125.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>97.1</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>Staphylococcus saprophyticus</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>RS_GCF_000236925.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>97.1</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>Staphylococcus argenteus</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>MZ417315.1</td>\n",
       "      <td>RS_GCF_000338275.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>97.1</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-2021_10</td>\n",
       "      <td>Staphylococcus xylosus_B</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>RS_GCF_002902325.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>97.1</td>\n",
       "      <td>CRISPR;96.40</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>Staphylococcus capitis</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>RS_GCF_002902345.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>97.1</td>\n",
       "      <td>CRISPR;96.50</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>Staphylococcus arlettae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>RS_GCF_003697915.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>96.7</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>Staphylococcus pseudoxylosus</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>RS_GCF_001431205.1</td>\n",
       "      <td>d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>96.1</td>\n",
       "      <td>None</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>Staphylococcus sp001431205</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Firmicutes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>MZ417354.1</td>\n",
       "      <td>RS_GCF_003417065.1</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>blast</td>\n",
       "      <td>92.0</td>\n",
       "      <td>iPHoP-RF;54.90</td>\n",
       "      <td>PG-2021_9</td>\n",
       "      <td>Hymenobacter sp003417065</td>\n",
       "      <td>Hymenobacter</td>\n",
       "      <td>Hymenobacteraceae</td>\n",
       "      <td>Cytophagales</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidota</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>556 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Virus         Host genome  \\\n",
       "123  MZ417315.1  RS_GCF_002902385.1   \n",
       "124  MZ417315.1  RS_GCF_900636325.1   \n",
       "125  MZ417315.1  RS_GCF_000010125.1   \n",
       "126  MZ417315.1  RS_GCF_000236925.1   \n",
       "127  MZ417315.1  RS_GCF_000338275.1   \n",
       "..          ...                 ...   \n",
       "674  MZ417354.1  RS_GCF_002902325.1   \n",
       "675  MZ417354.1  RS_GCF_002902345.1   \n",
       "676  MZ417354.1  RS_GCF_003697915.1   \n",
       "677  MZ417354.1  RS_GCF_001431205.1   \n",
       "678  MZ417354.1  RS_GCF_003417065.1   \n",
       "\n",
       "                                         Host taxonomy Main method  \\\n",
       "123  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...      CRISPR   \n",
       "124  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...      CRISPR   \n",
       "125  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "126  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "127  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "..                                                 ...         ...   \n",
       "674  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "675  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "676  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "677  d__Bacteria;p__Firmicutes;c__Bacilli;o__Staphy...    iPHoP-RF   \n",
       "678  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       blast   \n",
       "\n",
       "     Confidence score Additional methods   phagePGid  \\\n",
       "123              98.1     iPHoP-RF;97.10  PG-2021_10   \n",
       "124              97.6     iPHoP-RF;97.10  PG-2021_10   \n",
       "125              97.1               None  PG-2021_10   \n",
       "126              97.1               None  PG-2021_10   \n",
       "127              97.1               None  PG-2021_10   \n",
       "..                ...                ...         ...   \n",
       "674              97.1       CRISPR;96.40   PG-2021_9   \n",
       "675              97.1       CRISPR;96.50   PG-2021_9   \n",
       "676              96.7               None   PG-2021_9   \n",
       "677              96.1               None   PG-2021_9   \n",
       "678              92.0     iPHoP-RF;54.90   PG-2021_9   \n",
       "\n",
       "                          Species           Genus             Family  \\\n",
       "123    Staphylococcus intermedius  Staphylococcus  Staphylococcaceae   \n",
       "124       Staphylococcus delphini  Staphylococcus  Staphylococcaceae   \n",
       "125  Staphylococcus saprophyticus  Staphylococcus  Staphylococcaceae   \n",
       "126      Staphylococcus argenteus  Staphylococcus  Staphylococcaceae   \n",
       "127      Staphylococcus xylosus_B  Staphylococcus  Staphylococcaceae   \n",
       "..                            ...             ...                ...   \n",
       "674        Staphylococcus capitis  Staphylococcus  Staphylococcaceae   \n",
       "675       Staphylococcus arlettae  Staphylococcus  Staphylococcaceae   \n",
       "676  Staphylococcus pseudoxylosus  Staphylococcus  Staphylococcaceae   \n",
       "677    Staphylococcus sp001431205  Staphylococcus  Staphylococcaceae   \n",
       "678      Hymenobacter sp003417065    Hymenobacter  Hymenobacteraceae   \n",
       "\n",
       "                Order        Class        Phylum  \n",
       "123  Staphylococcales      Bacilli    Firmicutes  \n",
       "124  Staphylococcales      Bacilli    Firmicutes  \n",
       "125  Staphylococcales      Bacilli    Firmicutes  \n",
       "126  Staphylococcales      Bacilli    Firmicutes  \n",
       "127  Staphylococcales      Bacilli    Firmicutes  \n",
       "..                ...          ...           ...  \n",
       "674  Staphylococcales      Bacilli    Firmicutes  \n",
       "675  Staphylococcales      Bacilli    Firmicutes  \n",
       "676  Staphylococcales      Bacilli    Firmicutes  \n",
       "677  Staphylococcales      Bacilli    Firmicutes  \n",
       "678      Cytophagales  Bacteroidia  Bacteroidota  \n",
       "\n",
       "[556 rows x 13 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load file\n",
    "iPHoP_StaphStudy_preds = pd.read_csv('../data/predictions/iPHoP/StaphStudy_Host_prediction_to_genome_m90.csv', sep=',')\n",
    "\n",
    "# reformat the phage name to fit the format that I have\n",
    "phagenames = iPHoP_StaphStudy_preds['Virus']\n",
    "newnames = []\n",
    "for name in phagenames:\n",
    "    if name not in StaphStudy_virusnames['accession'].tolist():\n",
    "        PGid = 'NA'\n",
    "    else:\n",
    "        PGid = StaphStudy_virusnames.loc[StaphStudy_virusnames['accession'] == name, 'phagename'].values[0]\n",
    "    newnames.append(PGid)\n",
    "iPHoP_StaphStudy_preds['phagePGid'] = newnames\n",
    "\n",
    "# reformat prediction \n",
    "Species = []\n",
    "Genus = []\n",
    "Family = []\n",
    "Order = []\n",
    "Class = []\n",
    "Phylum = []\n",
    "\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    taxa = row['Host taxonomy'].split(';')\n",
    "    Phylum.append(taxa[1].lstrip('p__'))\n",
    "    Class.append(taxa[2].lstrip('c__'))\n",
    "    Order.append(taxa[3].lstrip('o__'))\n",
    "    Family.append(taxa[4].lstrip('f__'))\n",
    "    Genus.append(taxa[5].lstrip('g__'))\n",
    "    Species.append(taxa[6].lstrip('s__'))\n",
    "\n",
    "iPHoP_StaphStudy_preds['Species'] = Species\n",
    "iPHoP_StaphStudy_preds['Genus'] = Genus\n",
    "iPHoP_StaphStudy_preds['Family'] = Family\n",
    "iPHoP_StaphStudy_preds['Order'] = Order\n",
    "iPHoP_StaphStudy_preds['Class'] = Class\n",
    "iPHoP_StaphStudy_preds['Phylum'] = Phylum\n",
    "\n",
    "iPHoP_StaphStudy_preds = iPHoP_StaphStudy_preds[iPHoP_StaphStudy_preds['phagePGid'] != 'NA']\n",
    "iPHoP_StaphStudy_preds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86\n",
      "114\n",
      "324\n",
      "524\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3018864843.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['Species'].replace(' ', '_')\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "503\n",
      "0\n",
      "21\n",
      "524\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1823049242.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['Genus']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostgenus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostgenus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "503\n",
      "0\n",
      "21\n",
      "524\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3798743924.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostfamily = row['Family']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostfamily not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostfamily, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Order level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "524\n",
      "524\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['Order']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostfamily, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having weird issues above because of taxa mistmatch (Stahpylococcale versus Bacillale). Hopefully I will not have those issues with the Nahant Collection. \n",
    "\n",
    "In the meantime, because all the host in the Staph Study were Bacillale, it as simple as comparing how many were predicted as Staphylococcale versus not. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2679172136.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "correct = 503\n",
    "wrong = 0\n",
    "unknown = 21\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Class level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "503\n",
      "0\n",
      "21\n",
      "524\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2660217046.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostclass = row['Class']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostclass not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostclass, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phylum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "503\n",
      "0\n",
      "21\n",
      "524\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/250894921.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostphylum = row['Phylum']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostphylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostphylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nahant Collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus</th>\n",
       "      <th>Host genome</th>\n",
       "      <th>Host taxonomy</th>\n",
       "      <th>Main method</th>\n",
       "      <th>Confidence score</th>\n",
       "      <th>Additional methods</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>MG592390.1</td>\n",
       "      <td>RS_GCF_006538565.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>96.4</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio comitans</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>MG592390.1</td>\n",
       "      <td>RS_GCF_000316985.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>95.7</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio sp000316985</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>MG592390.1</td>\n",
       "      <td>RS_GCF_013394185.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>95.7</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio sp013394185</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>MG592390.1</td>\n",
       "      <td>RS_GCF_000272045.2</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>95.4</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio crassostreae_C</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>MG592390.1</td>\n",
       "      <td>RS_GCF_001048535.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>95.4</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio crassostreae</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6163</th>\n",
       "      <td>MG592672.1</td>\n",
       "      <td>RS_GCF_003937805.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>91.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio pectenicida</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6164</th>\n",
       "      <td>MG592672.1</td>\n",
       "      <td>RS_GCF_000171815.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>90.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio sp000171815</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6165</th>\n",
       "      <td>MG592672.1</td>\n",
       "      <td>RS_GCF_000400365.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>90.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio jasicida</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6166</th>\n",
       "      <td>MG592672.1</td>\n",
       "      <td>RS_GCF_001048675.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>90.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio diabolicus</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6167</th>\n",
       "      <td>MG592672.1</td>\n",
       "      <td>RS_GCF_014596725.1</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>iPHoP-RF</td>\n",
       "      <td>90.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Vibrio sp014596725</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6045 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Virus         Host genome  \\\n",
       "123   MG592390.1  RS_GCF_006538565.1   \n",
       "124   MG592390.1  RS_GCF_000316985.1   \n",
       "125   MG592390.1  RS_GCF_013394185.1   \n",
       "126   MG592390.1  RS_GCF_000272045.2   \n",
       "127   MG592390.1  RS_GCF_001048535.1   \n",
       "...          ...                 ...   \n",
       "6163  MG592672.1  RS_GCF_003937805.1   \n",
       "6164  MG592672.1  RS_GCF_000171815.1   \n",
       "6165  MG592672.1  RS_GCF_000400365.1   \n",
       "6166  MG592672.1  RS_GCF_001048675.1   \n",
       "6167  MG592672.1  RS_GCF_014596725.1   \n",
       "\n",
       "                                          Host taxonomy Main method  \\\n",
       "123   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "124   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "125   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "126   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "127   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "...                                                 ...         ...   \n",
       "6163  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "6164  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "6165  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "6166  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "6167  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...    iPHoP-RF   \n",
       "\n",
       "      Confidence score Additional methods                Species   Genus  \\\n",
       "123               96.4               None        Vibrio comitans  Vibrio   \n",
       "124               95.7               None     Vibrio sp000316985  Vibrio   \n",
       "125               95.7               None     Vibrio sp013394185  Vibrio   \n",
       "126               95.4               None  Vibrio crassostreae_C  Vibrio   \n",
       "127               95.4               None    Vibrio crassostreae  Vibrio   \n",
       "...                ...                ...                    ...     ...   \n",
       "6163              91.1               None     Vibrio pectenicida  Vibrio   \n",
       "6164              90.1               None     Vibrio sp000171815  Vibrio   \n",
       "6165              90.1               None        Vibrio jasicida  Vibrio   \n",
       "6166              90.1               None      Vibrio diabolicus  Vibrio   \n",
       "6167              90.1               None     Vibrio sp014596725  Vibrio   \n",
       "\n",
       "            Family             Order                Class          Phylum  \n",
       "123   Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "124   Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "125   Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "126   Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "127   Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "...            ...               ...                  ...             ...  \n",
       "6163  Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "6164  Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "6165  Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "6166  Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "6167  Vibrionaceae  Enterobacterales  Gammaproteobacteria  Proteobacteria  \n",
       "\n",
       "[6045 rows x 12 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load file\n",
    "iPHoP_NahantCollection_preds = pd.read_csv('../data/predictions/iPHoP/NahantCollection_Host_prediction_to_genome_m90.csv', sep=',')\n",
    "\n",
    "# reformat prediction \n",
    "Species = []\n",
    "Genus = []\n",
    "Family = []\n",
    "Order = []\n",
    "Class = []\n",
    "Phylum = []\n",
    "\n",
    "for _, row in iPHoP_NahantCollection_preds.iterrows():\n",
    "    taxa = row['Host taxonomy'].split(';')\n",
    "    Phylum.append(taxa[1].lstrip('p__'))\n",
    "    Class.append(taxa[2].lstrip('c__'))\n",
    "    Order.append(taxa[3].lstrip('o__'))\n",
    "    Family.append(taxa[4].lstrip('f__'))\n",
    "    Genus.append(taxa[5].lstrip('g__'))\n",
    "    Species.append(taxa[6].lstrip('s__'))\n",
    "\n",
    "iPHoP_NahantCollection_preds['Species'] = Species\n",
    "iPHoP_NahantCollection_preds['Genus'] = Genus\n",
    "iPHoP_NahantCollection_preds['Family'] = Family\n",
    "iPHoP_NahantCollection_preds['Order'] = Order\n",
    "iPHoP_NahantCollection_preds['Class'] = Class\n",
    "iPHoP_NahantCollection_preds['Phylum'] = Phylum\n",
    "\n",
    "iPHoP_NahantCollection_preds = iPHoP_NahantCollection_preds[iPHoP_NahantCollection_preds['Virus'].str.match('MG')]\n",
    "iPHoP_NahantCollection_preds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81\n",
      "397\n",
      "4967\n",
      "5445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2494548787.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Species'].replace(' ', '_')\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5157\n",
      "277\n",
      "11\n",
      "5445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3307562875.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Genus']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5219\n",
      "221\n",
      "5\n",
      "5445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3916616848.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Family']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Order level\n",
    "\n",
    "Getting same issues with mismatch in the taxa names despite looking at the same predictions. Not sure where the people behind iPHoP got their taxa names. Potentially worth looking into. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Enterobacterales    6044\n",
       "                       1\n",
       "Name: Order, dtype: int64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iPHoP_NahantCollection_preds['Order'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/168101269.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/168101269.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/168101269.py:8: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "correct = 5445\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'iPHoP', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### iPHoP (considering only best prediction)\n",
    "\n",
    "Because iPHoP may return more than 1 predicted host taxa, the comparison with other models get muddled. I am going to re-calculate the accuracy of this tool but this time only going to consider the best hit. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Staph Study \n",
    "\n",
    "First, I need to subset the dataframe so it only contains the best prediction. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename column \n",
    "iPHoP_StaphStudy_preds.rename(columns={'Confidence score':'Confidence_score'}, inplace=True)\n",
    "iPHoP_StaphStudy_best_preds = iPHoP_StaphStudy_preds.loc[iPHoP_StaphStudy_preds.groupby('phagePGid', sort=False).Confidence_score.idxmax()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interestingly, some phages have 0 predictions. Only now do I realize it. I figured that out because number of phages expected is 39, but the dataframe has only 33 rows. Will need to find a way to score those missing phages somehow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species level "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Staphylococcus_intermedius\n",
      "Staphylococcus_epidermidis\n",
      "Staphylococcus_epidermidis\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_capitis\n",
      "Hymenobacter_sp003417065\n",
      "Staphylococcus_argenteus\n",
      "Staphylococcus_saprophyticus\n",
      "Staphylococcus_delphini\n",
      "Staphylococcus_delphini\n",
      "Hymenobacter_sp003417065\n",
      "Hymenobacter_sp003417065\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_delphini\n",
      "Staphylococcus_epidermidis\n",
      "Staphylococcus_epidermidis\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_petrasii_B\n",
      "Staphylococcus_delphini\n",
      "Staphylococcus_petrasii_B\n",
      "Hymenobacter_sp003417065\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_intermedius\n",
      "Staphylococcus_petrasii_B\n",
      "Staphylococcus_saprophyticus\n",
      "Staphylococcus_delphini\n",
      "Staphylococcus_lugdunensis\n",
      "Staphylococcus_petrasii_B\n",
      "Staphylococcus_haemolyticus\n",
      "Staphylococcus_intermedius\n",
      "4\n",
      "5\n",
      "23\n",
      "32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/2628611061.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['Species'].replace(' ', '_')\n",
    "    print(hostpred)\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "0\n",
      "4\n",
      "32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1603746211.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostgenus = row['Genus']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostgenus not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostgenus, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "0\n",
      "4\n",
      "32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3331252752.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostfamily = row['Family']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostfamily not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostfamily, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Order level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "32\n",
      "32\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostpred = row['Order']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['order'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['order'] == hostfamily, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Same issue again. Doing value count instead then "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Staphylococcales    29\n",
       "Cytophagales         4\n",
       "Name: Order, dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iPHoP_StaphStudy_best_preds['Order'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3550686202.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "correct = 28\n",
    "wrong = 0\n",
    "unknown = 4\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Class level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "0\n",
      "4\n",
      "32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/3930269946.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostclass = row['Class']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['class'].tolist()\n",
    "        if hostclass not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['class'] == hostclass, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phylum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "0\n",
      "4\n",
      "32\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1649784001.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    " # initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the predictions and determine accuracy\n",
    "for _, row in iPHoP_StaphStudy_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['phagePGid']\n",
    "    hostphylum = row['Phylum']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = StaphStudyHR[StaphStudyHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host range of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['phylum'].tolist()\n",
    "        if hostphylum not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['phylum'] == hostphylum, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'StaphStudy', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nahant Collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(193, 12)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rename column \n",
    "iPHoP_NahantCollection_preds.rename(columns={'Confidence score':'Confidence_score'}, inplace=True)\n",
    "iPHoP_NahantCollection_best_preds = iPHoP_NahantCollection_preds.loc[iPHoP_NahantCollection_preds.groupby('Virus', sort=False).Confidence_score.idxmax()]\n",
    "\n",
    "iPHoP_NahantCollection_best_preds.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "193 viruses have predictions out of 248. \n",
    "\n",
    "##### Species level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "13\n",
      "160\n",
      "173\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1396487523.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Species'].replace(' ', '_')\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['hostname'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['hostname'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Species', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "154\n",
      "13\n",
      "6\n",
      "173\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/489699161.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Genus']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['genus'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['genus'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Genus', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Family level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "160\n",
      "9\n",
      "4\n",
      "173\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1886209405.py:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "# initialize integers\n",
    "correct = 0\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "# iterate through the prediction and determine accuracy\n",
    "for _, row in iPHoP_NahantCollection_best_preds.iterrows():\n",
    "    # retrieve phage id and prediction\n",
    "    phage = row['Virus']\n",
    "    hostpred = row['Family']\n",
    "\n",
    "    # subset hostrange dataframe for current phage\n",
    "    subset = NahantCollectionHR[NahantCollectionHR['phagename'] == phage]\n",
    "\n",
    "    # check if I have host of phage of interest\n",
    "    if subset.empty:\n",
    "        continue\n",
    "    # determine accuracy\n",
    "    else:\n",
    "        known_hosts = subset['family'].tolist()\n",
    "        if hostpred not in known_hosts:\n",
    "            unknown += 1\n",
    "        else:\n",
    "            interaction = subset.loc[subset['family'] == hostpred, 'infection'].tolist()\n",
    "            if 'Inf' in interaction:\n",
    "                correct += 1\n",
    "            elif 'NoInf' in interaction:\n",
    "                wrong += 1\n",
    "\n",
    "print(correct)\n",
    "print(wrong)\n",
    "print(unknown)\n",
    "print(correct + wrong + unknown)\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Family', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Order level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Enterobacterales    192\n",
       "                      1\n",
       "Name: Order, dtype: int64"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iPHoP_NahantCollection_best_preds['Order'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1356318104.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1356318104.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
      "/var/folders/yy/x1x_rtqj13l_9jplp1l2hxgw0000gn/T/ipykernel_37643/1356318104.py:8: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "correct = 173\n",
    "wrong = 0\n",
    "unknown = 0\n",
    "\n",
    "\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Order', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Class', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n",
    "results = results.append({'Tool' : 'iPHoP_bestpred', 'Dataset' : 'NahantCollection', 'Taxa_lvl' : 'Phylum', 'Correct' : correct, 'Wrong': wrong, 'Unknown': unknown}, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Plotting the results!\n",
    "\n",
    "Adding a couple new columns to the results, such as the accuracy and the percent unknown.\n",
    "\n",
    "Also saving the results to analyze and plot in R since I am more comfortable with that language."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tool</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Taxa_lvl</th>\n",
       "      <th>Correct</th>\n",
       "      <th>Wrong</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>perc_accuracy</th>\n",
       "      <th>perc_unknown</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VHMnet</td>\n",
       "      <td>StaphStudy</td>\n",
       "      <td>Species</td>\n",
       "      <td>29</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.74359</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VHMnet</td>\n",
       "      <td>StaphStudy</td>\n",
       "      <td>Genus</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.974359</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VHMnet</td>\n",
       "      <td>StaphStudy</td>\n",
       "      <td>Family</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.974359</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VHMnet</td>\n",
       "      <td>StaphStudy</td>\n",
       "      <td>Order</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.974359</td>\n",
       "      <td>0.025641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VHMnet</td>\n",
       "      <td>StaphStudy</td>\n",
       "      <td>Class</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>iPHoP_bestpred</td>\n",
       "      <td>NahantCollection</td>\n",
       "      <td>Genus</td>\n",
       "      <td>154</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>0.890173</td>\n",
       "      <td>0.034682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>iPHoP_bestpred</td>\n",
       "      <td>NahantCollection</td>\n",
       "      <td>Family</td>\n",
       "      <td>160</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>0.924855</td>\n",
       "      <td>0.023121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>iPHoP_bestpred</td>\n",
       "      <td>NahantCollection</td>\n",
       "      <td>Order</td>\n",
       "      <td>173</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>iPHoP_bestpred</td>\n",
       "      <td>NahantCollection</td>\n",
       "      <td>Class</td>\n",
       "      <td>173</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>iPHoP_bestpred</td>\n",
       "      <td>NahantCollection</td>\n",
       "      <td>Phylum</td>\n",
       "      <td>173</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Tool           Dataset Taxa_lvl Correct Wrong Unknown  \\\n",
       "0           VHMnet        StaphStudy  Species      29     9       1   \n",
       "1           VHMnet        StaphStudy    Genus      38     0       1   \n",
       "2           VHMnet        StaphStudy   Family      38     0       1   \n",
       "3           VHMnet        StaphStudy    Order      38     0       1   \n",
       "4           VHMnet        StaphStudy    Class      39     0       0   \n",
       "..             ...               ...      ...     ...   ...     ...   \n",
       "65  iPHoP_bestpred  NahantCollection    Genus     154    13       6   \n",
       "66  iPHoP_bestpred  NahantCollection   Family     160     9       4   \n",
       "67  iPHoP_bestpred  NahantCollection    Order     173     0       0   \n",
       "68  iPHoP_bestpred  NahantCollection    Class     173     0       0   \n",
       "69  iPHoP_bestpred  NahantCollection   Phylum     173     0       0   \n",
       "\n",
       "   perc_accuracy perc_unknown  \n",
       "0        0.74359     0.025641  \n",
       "1       0.974359     0.025641  \n",
       "2       0.974359     0.025641  \n",
       "3       0.974359     0.025641  \n",
       "4            1.0          0.0  \n",
       "..           ...          ...  \n",
       "65      0.890173     0.034682  \n",
       "66      0.924855     0.023121  \n",
       "67           1.0          0.0  \n",
       "68           1.0          0.0  \n",
       "69           1.0          0.0  \n",
       "\n",
       "[70 rows x 8 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['perc_accuracy'] = results['Correct'] /  (results['Correct'] + results['Wrong'] + results['Unknown'])\n",
    "results['perc_unknown']  = results['Unknown'] / (results['Correct'] + results['Wrong'] + results['Unknown'])\n",
    "\n",
    "# also export the results\n",
    "results.to_csv('../data/output/modelsaccuracy_results.tsv', sep='\\t', index=False)\n",
    "\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY0AAAEJCAYAAABohnsfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAx3ElEQVR4nO3dd3wVVfrH8c+9CQkgBBBQsSA2HuxUywr2iqCs4qKwIq5gYcGuqIBtRRdQ7CCKgMpasS/o2kXcRQWs6/qIPwUVUIpUpSX3/v6YSUxCyiTkpvF9v1555d4z7Tm5N/PMnDlzJpZMJhEREYkiXtUBiIhIzaGkISIikSlpiIhIZEoaIiISmZKGiIhEll7VAaRQJtAJWAzkVHEsIiI1RRrQAvgI2FB4Ym1OGp2A96o6CBGRGqoLMLNwYW1OGosBVqz4lURC96KIiEQRj8do0mQbCPehhdXmpJEDkEgklTRERMquyGZ9XQgXEZHIlDRERCSylDZPmdnNQE8gCTzs7mPM7FhgDFAPeMrdh4XztgUmAFnADOBCd882s5bAFGA7wIE+7r52S+JKJpOsWLGUjRvXh6FJbZCWlk6DBo2pV2+bqg5FpNZKWdIwsyOAo4EDgDrAl2b2JjAROAL4AZhmZie5+ysEiaG/u88ys4eBAcA4YCww1t2fNLPhwHBgyJbEtnbtKmKxGNtvvzOxmE62aoNkMsmmTRtZuXIpgBKHSIqkbI/p7u8CR7l7NsFZQjrQGJjn7t+F5VOAM8xsV6Ceu88KF58cltcBDgem5i/f0tjWrVtLw4aNlTBqkVgsRkZGJo0bN2ft2pVVHY5IrZXS5il332RmNwFXAs8AO1KwG9diYOcSypsBq8MEk788sqZNG2xWtmRJkszMDGKxWFlWJTVAWlpdVq5M0Lx5w6oORaRWSnmXW3e/wcxGAi8DrSl4ESEGJAjOeKKUE5ZHtnz52s263CYSCXJykkWsWmqDRCLB0qVrqjoMkRopHo8VebCdK5XXNNoAdd39E3f/zcyeI7gonr/v7w7AIuBHgtvWC5cvARqZWZq754TzLEpVzDXJXXeN5pNPPgZg/vxvadFiRzIz6wIwfvzEvNelWbx4EX379uL113Xz/NasUVY9MjKL3h1s3JDNqtXrKjmiiqX6VVz9UnmmsTtwk5l1JjikPxUYD4w2sz2B74DewER3X2Bm683sMHd/HzgbeCVs3noP6AU8DvQFXklhzDXGpZdelfe6Z8/u3HDDLbRps08VRiQ1WUZmOrcOnVrktOtG9KzkaCqe6ldxUpY03H26mR0EfExwdvFs2ANqKfAsUBeYzu8XufsAD5lZFjAXuCcsHwg8YmbDgO+Bs1IVc20wefIE3njjX6SlpbHLLi257LKradq0GUuW/Mztt/+dn35aRDKZ5KSTutG7d9+qDldEaphUXwi/EbixUNmbwIFFzPspcFAR5QuAI1MSYC0zbdpLzJr1bx566FHq1avHww+PZ8SImxgz5l5uvnk4nTsfzpln3snatWv5618HsN1227PvvvtXddgiUoOoz2ktMmvWv+natTv16tUD4IwzzmLOnA9Zu3Ytn3/+Kaed9icAGjRoQNeu3Zg1699VGa6I1EBKGrVIIpFToBtxMpkgJycnfF24B1mS7OxsRETKQkmjFjn44D8wbdpLrFsX9JSYOvUp2rZtT4MGDdh33/147rmnAVi7di2vvjqNTp0OrspwRaQGqs1Do291unU7lSVLfmbAgHNIJhPstNMuXH/93wC4/vpbGDNmJNOnv8ymTZs47rgT6dq1Oz/9VOSQ+SIiRVLSqAWmTn0573X//hfSv/+Fm83TosWOjB59d5HlukejdCX1g4fa0ddfJAolDZEISuoHD7Wjr79IFEoaUiHS4jFi8eLH8komkuToCYoiNZ6ShlSIWDzG4oUrip3eYqcmoKQhUuOp95SIiESmpCEiIpEpaYiISGRKGiIiEpkuhAMNs+pSN7NOha93/YZNrFm9vsR5LrroPE4//U8ce+wJeWXr1q3j9NO70bRpUy677Grat++YN23EiBtp164DXbt2p2fP7tSpU4cnnngub3p2djY9epzIoYd2ZujQG8sc85dffsE777zFwIEXl3lZEan9lDSAupl16H31Pyp8vY+P6sMaSk4aJ598Cq+99mqBpPHuu2/Rvn1HVq4svjdSrg0bNvB///cNe+yxJwBz5nxE8ODD8pk//ztWrPil3MuLSO2m5qkqdvTRx/H555+yevWqvLJ//Ws6J598SqTljzjiaN55582892+++RpHHnlM3vtBg85n7Ni7ueCCc+nVqwf/+c/7APzyy3KuvfYK/vKXP9O/f18++ugD1qxZw4QJDzBz5gweeeThCqqhiNQmShpVrH79+nTpcgRvvfUGAMuWLeX77xdw0EGHADBy5C3069c772fmzBkFlj/qqGN49923ANi0aRPffPM1++yzb4F5Nm3KZvz4SQwefDkPPTQOgLvvvp2TTz6FiROn8Pe/j2H06FtJS4vTv/+FdO58OOecc16qqy4iNZCap6qBrl27M2HCA/TocTqvvfYKJ5zQlbS0NACGDBm22TWN/Jo3345ttmnA/PnfsXDhj3TqdMhm6z/44EMB2H33PVizZjUAs2d/yIIFC5gwYTwQXAtZuPDHVFRPRGoRJY1qoG3b9ixfvoyff/6Jf/3rFW69dXSZlj/qqGN55503+eGH7+nVqzfffDOvwPSMjAwAYrFY3nM1cnIS3HPPOLKyGgGwbNkymjRpwrx5X1dAjUSktlLzVDVx4okn8+ijE8nKymKnnXYu07JHH30sb731OgsWzKd16zaRlunQoSPPPfcMAN999y19+/Ziw4b1pKWl5T24SUSkMJ1pEHSNfXxUn5SsN6rcLrTXXnt9mbfTrFlzGjRoSLt2HSIvc9llVzNq1AjOOedMkskkw4ffTP3627D33vsyceKDjBt3LxddNLjMsYhI7aakAaxZvb7UrrGp1rz5drz77gcFyu6778HN5st/70X+52iMHTsh73XXrt3p2rX7Zuto0WLHvGWaNWvOqFF3bbb+li135amnXihPFURkK6DmKRERiUxJQ0REIlPSEBGRyJQ0REQkMiUNERGJLKW9p8zsBuBP4dtp7n61mU0COgO/huU3ufvzZtYWmABkATOAC90928xaAlOA7QAH+rj72lTGLSIiRUvZmYaZHQscD7QD2gIdzOyPQEfgcHdvG/48Hy4yBRjk7q0JhmkdEJaPBca6extgNjC8omNt0iiD5s0bVvhPk0YZ5YpnxIgbmT795QJlDz88nocfDob8GDTofObOnV3sMp07d6QohcvvuGMkgwadz2+//VauOEVk65PKM43FwBXuvhHAzP4HtAx/JprZTsDzwE3ALkA9d58VLjsZuMnMJgCHAz3ylb8LDKnIQNMzMpkzqn9FrhKADldPADZW+Horwl133c7338/n9tvvoW7dulUdjojUEClLGu7+39zXZrYXQTNVF+BIYCCwCvgncB7wBUGSybUY2BloBqx29+xC5ZE1bdpgs7IlS+Kkp1fO5ZzStjNkyBWccEJXjj46GM78nHN6s9122xOPxwosG4/H8tYXi8VISytYh1gsVmCZ4rabnh7n7rvH8P3387njjrsrNWFU1t88Ho/TvHnDStlWflWxzcpSm+sGql9ZpPyOcDPbF5gGXOXuDvwx37R7gb7Al0Ay32IxIEHQfJa/nLA8suXL15JIFFxFIpEgO7tMqym30rZz/PFdee21Vzj88KP44Yfv2bBhIw0aNOTBB8fx5JOP5833yy/LOfXU08jOTpBMJsnJKViHZDJJIpHMKytuu/fccxdPPDGFe+55gPT0jAr7O0RJCJX1N08kEixduqZC1xnln66it1mZSqtfTa4bqH5lqV88HivyYDtveuQ1lYOZHQa8CVzj7o+Y2f5mdnq+WWLAJuBHoEW+8h2ARcASoJGZpYXlLcLyWuMPf+jMF198zm+//cobb/yLE044CYD+/S9k8uTH835OPfW0vGXi8c0/tmQySSxW+hP75s//lmHDbuK2225m7Vr1JxCRsknlhfBdgBeA3u7+ZFgcA+4ysyZmVgc4H3je3RcA68MkA3A28Iq7bwLeA3qF5X2BV1IVc1WoU6cOhx3WhZkzZ/DWW69z3HEnlrpMw4YNN9vhr1ixgoYNs0pddsSI0Zx44snst98BjB59a7njFpGtUyrPNK4E6gJjzOwTM/sE+ANwG/A+QZPUJ+7+RDh/H+BOM/sKaADcE5YPBM43sy8JrokMS2HMVeKEE7ry5JNTaNSoMTvs0KLU+Tt0OIhXX51GdnZwqef77+fj/j/22++AUpetU6cOAJdfPoTPP/+UadNe2rLgRWSrksoL4ZcAlxQzeWwR838KHFRE+QKCi+e11gEHtGXt2rX06NEz0vynnPJHFi78kX79ehOPx8jMzOSGG/5G48aN8+Y57rguea+3374FU6Y8XWAdDRs2ZOjQG7n22ivZf/8Dadly1wqpi4jUbhoaHcjeuCHsHlvx643q6adfzHudf/jzXOedd0He6/T0dAYPvozBgy8rcl0zZ86OVN6hQydee+3dyDGKiChpACtWbaS63k8hIlKdaOwpERGJTElDREQiU9IQEZHIlDRERCQyJQ0REYlMSUNERCJTl1sgq1EmmRnle/ZFSTZs3MjqVaXfq/Hrr2t54IH7+eSTOaSlpdOwYUMGDbqMrKwsBg++gKlTCz5bo3PnjsycOZu5c2czZMhl7LTTLgWmn3vuAI444ig6d+7Innu2BoKxqdauXcPBBx/KFVdcQ1paWonTr732Clq2bMWgQZfmrfeFF55l+vSXGTfuYdLS0hCRrY+SBpCZkUG/ScXdvF5+k8+9Gyg5aSQSCa688hLat+/IpEmPk56ezty5s7nyyosZPfruUrdhtjf33fdg8TFM/n2k3F9/XcvZZ/fiww9nceihh5U4/aqrrqNfv7M4/vgTad26DcuWLWXSpAe5997xShgiWzE1T1WxuXNn8/PPP3HeeReQnh7k8PbtO3LdddeTSORU6LZWrlzJhg3rycpqVOr05s2346KLBjNy5AgSiQR33jmKPn360bJlqwqNSURqFp1pVLGvv3b22qv1ZsOdH3poZxYvXsSyZUvp1693scu7/2+z6XffPZZGjRoD0K9fb7Kzs1m58hd23XU3LrnkKvbdd7+8eUua3q1bD95883Vuvnk4q1ev5owzzqygWotITaWkUcXi8RgZGZnFTm/WrHmBJiQo+KzvqM1TTz31D6ZP/ydduhxRpulDhgyjZ8/uPPPMS5Ge1yEitZuap6pYmzb78PXXX5FMFny64Pjx9zN3btEDD5ZHr159aNq0KfffX/R1kuKm5w7V3qLFjhUWi4jUXEoaVezAA9vRpMm2TJz4IDk5wTWMDz74D9Onv0SrVrtX6LYGDbqMadNe5Jtv5pVruoiImqcIusYGPZ0qfr2licVi/P3vY7j33jvo27cX6enpNGrUmNGj7470JL6irmkcc8zxnH12v83m3X33PTjxxJO57747ueuuzR5pUup0ERElDQjvpYj+7IuK1rhxY4YP/1uR0wrfowG/PxejffuOvP76e8Wut6jnagwZMizy9JLmE5Gtk5qnREQkMiUNERGJTElDREQiU9IQEZHIlDRERCQyJQ0REYlMXW6Bxg0zqFO3+KE8ymvT+g2sXFP6vRqLFy/irLNOy7uZL5lM8Ouvv3LSSd0477wLAPj222/o2/dMbrllJEceeUzesoMGnc/SpUuoV69+Xln37j04/fQ/AXDeeWfTtGkzRo26M2/69Okv8/HHcxg69Ma8srlzZzNx4oMlDkkiIqKkAdSpm8n0vudW+Hq7PjoJIiQN2HyMqWXLlnLmmX/kmGOOp1Wr3Zg27SWOOupYXnzxuQJJA4J7K9q371h4lXzzzTwyMurwzTdf8/PPP7H99jtsWYVEZKuX0qRhZjcAfwrfTnP3q83sWGAMUA94yt2HhfO2BSYAWcAM4EJ3zzazlsAUYDvAgT7uvjaVcVcHy5YtI5lMUr9+fbKzs3nttVe5//6HuOiiv7Bw4Y/stNPOpa5j+vSX6NjxYFavXsXLL79A//4XVkLkIlKbpSxphMnheKAdkAReNbOzgJHAEcAPwDQzO8ndXyFIDP3dfZaZPQwMAMYBY4Gx7v6kmQ0HhgNDUhV3VckdAn3jxg2sWrWSNm325dZbb2e77bZnxox32GGHHWjZcle6dDmSF198joEDL85bduTIW/Kap+rXr8/YsRPyEs29945n9erV3HDDtfTr1z/vmR0zZ84oMPzIunW/0bz5dpVbaRGpcVJ5prEYuMLdNwKY2f+A1sA8d/8uLJsCnGFmXwL13H1WuOxk4CYzmwAcDvTIV/4utTBp5DZPJRIJ7rvvTubP/45OnQ4GgjOGY489AYBjjjmOm24azoABF1GnTh2g6Oap999/j6ZNm7HbbruTTCaJx+O8//57HHHEUQB07nx4kdc0RERKkrKk4e7/zX1tZnsRNFPdS5BMci0GdgZ2LKa8GbDa3bMLldda8XicgQMv4dxze/PEE4/RtWt3Zs36N+5f8cwzT5JMJlmzZjXvvvtWXiIpyvTpL/Hzzz/Rs2d3IHiU64svPpeXNEREyiPlF8LNbF9gGnAVkE1wtpErBiQIuv4mI5QTlkfWtGmDzcqWLImTnl45vY2jbCctLV5g3vT0DC6++DKGDh1CIpFDx44Hcddd9+XN/9BDD/Dii89x4oknEYvFSEsrWJ/ly5cze/aHPPPMi2y3XdDktHDhj/zpT6fx88+LiMdjxGKxAsukpcU3K6tolfU3j8fjNG/esFK2lV9VbLOy1Oa6gepXFqm+EH4Y8CxwaXhN4gigRb5ZdgAWAT8WU74EaGRmae6eE86zqCwxLF++lkSiYN5JJBJkZ5cp95RblO3k5CQ2m7dTp0PZd9/9mTbtZQYNuqzAtB49zmDKlEf4v//7lmQySU5OwfpMm/ZPDjnkD2y7bbO88u2335HOnbvw3HPPsuuurUgmkwWWyclJbFZWFlESQmX9zROJBEuXrqnQdUb5p6vobVam0upXk+sGql9Z6hePx4o82M6VygvhuwAvAL3c/a2w+INgku0JfAf0Bia6+wIzW29mh7n7+8DZwCvuvsnM3gN6AY8DfYFXKjrWTes3BN1jU7DeKFq02LHIIdDvvPP+Iudv0qQJb775PkCR91X07n12kcuNGDE673XXrt0LTGvfvmOR3XZFRPJL5ZnGlUBdYIyZ5ZY9APQjOPuoC0wHpobT+gAPmVkWMBe4JywfCDxiZsOA74GzKjrQlWs2Rr6fQkRka5bKC+GXAJcUM/nAIub/FDioiPIFwJEVGpyIiJSLxp4SEZHIlDRERCQyJQ0REYlMSUNERCJT0hARkcg0NDrQKKseGZkV/6fYuCGbVavXlTrfV199yQsvPEtOTg5z5nxEVlYjADZt2shpp53B6af3KvJ5F4sXL2Lw4AuKvMcj19y5sxky5DJ22mkXYjHYsGEDu+++B9dddwP162+z5ZUUka2KkgaQkZnOrUOnlj5jGV03omek+dq02YdrrtmHESNupH//C/NuvPvll+X06tWDDh0264lcJmZ7F0g2w4ZdzaOPTuLCCwdt0XpFZOujpFEN5J5FtGixY4HybbdtSsuWrZg//9u8s4/irF+/npEjb+Gbb74mHo9z5pl/5qSTuhU5b7t2Hfjgg1lFThMRKUmkpGFmO7n7wkJl+7j7l6kJSwDmzfuahQt/wGxvFi9ehPv/CjwDIzt7U97riRPH06hRIx577GlWrlzJgAHnsNdettk6161bx8yZM2jXrkOl1EFEapcSk4aZbRu+nG5mRxKMPgtQB3gOaJO60LZOEyY8wNNPP0EikUNmZl2uumooLVrsyOLFizZrZsq9pgEwZ85srrlmOACNGzemS5fD+fjjOeyxx54Fkk12djYdOnSkV68+lV85EanxSjvTeAI4Lny9PF95DvBMSiLayuW/plEWyWSi0HvIyQkeQ1I42YiIlFeJXW7d/QR3jwOT3T2e76eOu/cuaVmpXO3bd2LatBcBWLlyJe+99w7t2mnUWhGpWFEvhF9vZmPdfaAFQ9aOBC5w959TGFul2bghO3JPp7Kut7Kce25/7rhjJH379iKRSNC3718wa8PcubMrLQYRqf2iJo3JwEvh6wXAO8AkoGvFh1T5otxLkUpRnmVR1Dz5n8OxzTYNuP76v5Vr3SIiUUW9I7yZu98D4O7r3f0uCj5pT0REtgJRk0a6meXdRGBm2/N7TyoREdlKRG2eGgN8YmavAkngWOCqlEUlIiLVUqQzDXefSJAoPgZmAye4++OpDExERKqfsoxyWxf4DXgQaJiacEREpDqLlDTMrB9Bb6mrgUbAi2Y2IIVxiYhINRT1msbFwKHAu+6+xMw6AK8CD6UsskrUKCuDjMzMCl/vxg0bWLV6Y6nzVceh0Xv27M69947fbBDFshp+/eX87eYxW7SOokyf/jIffzyHoUNvrPB1i0jxoiaNHHdfHdzXB+7+g5lV3p1rKZaRmcmYay+o8PVeftt4oPSkUZuHRv/iv5+mfBsiUnmiJo1fzKwtQc8pzKwP8EuqgtraVNeh0SdOfJBvvvmajIxMrrrqOvbccy9++WU5o0ffys8//0w8HueCC/5Kp04H89FHH3DnnWOIxWI02KYBl182jKefeQyAq6/5K4899jjduh3HEUccxRdffEb9+vW5/vpbaNFiR3r27M4+++zHvHnO2LETmDXr3zzzzBMkEknM2nD55UPIzMzk1Ven8cgjD7PNNg3YYYcdqFevfsS/sIhUlKhJ4xJgKrCHmS0G1gGnpiwqAap+aPSdd96FoUNv5D//mcmIETcwadLj3H337Zx88il07nwEy5YtY+DA85g8+XEmTXqYCy+4lL32bMPzLzzJt9/Oo/95g5g2/XlG/f1+AFauXMF++x3AVVddx9SpT3LXXaMZOfJOAA455A/cfPNtfPvt//Hyyy8wbtxEMjMzeeCB+3jiicfo1u1Uxo27h0mTHicrqxFXX32pkoZIFYiaNOoDBwKtgTTA3X1TyYtIeVSnodG7d+8BwKGHdubmm69nzZo1zJ79IQsWLGDChPF561u48Ee6dDmckSNv4KCDDuOgg/5A2wM3H7okIyOTE088GYCTTurG+PH3503bZ5/9APj449n8+OMPXHDBueH6N9G6dRs+//xT9tvvALbdtikAxx9/EnPmfBT9DysiFSJq0viHu+8N/C+VwUj1Gho9LS0t3/qSpKenk5OT4J57xuU1ly1btowmTZqw9957Y3u1ZfacWTz62EPMm+ec0bNgYorHY8RiwUACiUSywPozw44IOTkJjj76WC69NLh39Lfffgs7CHxIMll0bCJSeaLep/GZmfU2s5Zmtm3uT2kLmVmWmX1hZq3C95PMbJ6ZfRL+/DEsb2tms83sazObYGbpYXlLM5thZl+Z2Ytm1qC8Fa3tUjE0+muvvQrAu+++TatWu1GvXj06dOjIc88Fj1L57rtv6du3Fxs2rOcvf+nLunW/0b3b6XTvdjrffjsPgHg8Tk5ODhBcd5k5cwYA06e/xMEH/2GzbbZr14EZM95hxYpfSCaT3HHHbTz99OMccEBb/vvfz1i6dAmJRIK33np9i+omIuUT9UzjVOCMQmVJgqaqIpnZwQRdclvnK+4IHO7uiwvNPgXo7+6zzOxhYAAwDhgLjHX3J81sODAcGBIx5q1KKoZG/+GHBfTr15v69evndW297LKrGTVqBOeccybJZJLhw2+mfv1tuOiiv3LH7aOIp6VRt249/nrRFQAc1OkwLrt8AFP+8SQAb7/9Bg89NJamTZszbNiNm21zr71ac+65A7j44gtJJpPsuWdr/vznfmRmZnLppVdx6aUDqVu3Hq1a7VbueolI+cWS+c/5i2FmHdx9TllWbGYTgEeAx4AjgSXAYuDfwE7A88BNwC7AW+6+R7hcl7D8BIKnBW7r7tlmtgvBfSK7RwyhFfDd8uVrSSQK1vGnnxawww675r2v6vs0aoP09DiLF64odnqLnZpwyCHtmTkz9c/3KPz5VoTmzRty69CpxU6/bkRPli5dU6HbrEwl1a+m1w1Uv7LULx6P0bRpA4DdgPmFp0c905gC7B15q4C79wfIvbcD2AF4CxgIrAL+CZwHfEGQTHItBnYGmgGr3T27UHmFC3bsW8fOPb/Bgy9gzZrNv0w9epxGjx4V/1AqEan5oiaNz8ysNzATWJtb6O6R79Vw92+BP+a+N7N7gb7Al4T3f4RiQILgekvh06AEZRRmzAKWLImTnl6WYbdqp3HjKveG/lmz5lbKduLxOM2bV/7waFWxzcpSm+sGql9ZpOyaRmFmtj/Q2t2fDYtiwCbgRwo+0GkHYBFBc1YjM0tz95xwnkVRt5erqOapRCJBdnaZ84+UIEoSrqy/eSKRqPDmhij/dDW5iaO0+tXkuoHqV87mqSJFShruXjfyFosXA+4ys7cIzlbOBx5x9wVmtt7MDnP394GzgVfcfZOZvQf0Ah4nOCt5pQLiAIIupLndP6X2CLoe63MVSZVIScPM4sCVwElAHeA14NZ81xtK5e6fmdltwPvhOp519yfCyX2Ah8wsC5gL3BOWDwQeMbNhwPfAWVG3V5L09Ax+/XU122yTpcRRSySTSXJyslmzZgUZGRVxjCMiRYnaPHUbwR3hdxNcazgfuB24tLQF3b1VvtdjCbrRFp7nU2CzUfncfQFBz6sK1aRJc1asWMratSsretVbrXg8zqpVvxY7PRlbTSKR2uapeDyNevUa0KBByeN0iUj5RU0aJwIdc4cOMbNpQI0dvjQtLZ1mzVqUPqNEVtu7pIpIIGoXonj+sabcfQPBRWwREdmKRD3T+MTM7gTuI+g1NRj4LGVRiYhItRT1TOOvQBOCu7k/ILjxbnCqghIRkeqpxDMNM8sgGD/qBXfvF5ZNA3KA1SmPTkREqpXSzjRuBrIIusnmGgA0Bm5MTUgiIlJdlZY0ugG93X1JboG7LyK40e6PxS4lIiK1UmlJY6O7rytc6O6rgQ2pCUlERKqr0pJGjpltNqhJWFYnNSGJiEh1VVrSeAKYYGbb5BaErycAzxa7lIiI1Eql3adxF/AA8JOZ/ZcgyewN/IPgIrmIiGxFSkwa7p4AzjezEUAHgudZfFDE41pFRGQrEHVo9AXAghTHIiIi1ZweXyciIpEpaYiISGRKGiIiEpmShoiIRKakISIikSlpiIhIZEoaIiISmZKGiIhEpqQhIiKRKWmIiEhkShoiIhKZkoaIiESmpCEiIpFFGuW2vMwsC/g30M3d55vZscAYoB7wlLsPC+drS/BgpyxgBnChu2ebWUtgCrAd4EAfd1+byphFRKR4KTvTMLODgZlA6/B9PWAicCrBg5w6mdlJ4exTgEHu3hqIAQPC8rHAWHdvA8wGhqcqXhERKV0qm6cGAH8FFoXvDwLmuft37p5NkCjOMLNdgXruPiucb3JYXgc4HJiavzyF8YqISClS1jzl7v0BzCy3aEcg/xP/FgM7l1DeDFgdJpj85SIiUkVSek2jkDiQzPc+RvD42KjlhOVl0rRpg7IuIinSvHnDqg4hpWpz/Wpz3UD1K4vKTBo/Ai3yvd+BoOmquPIlQCMzS3P3nHCeRZTR8uVrSSQK5x6paFG+lEuXrqmESFJja69fTa4bqH5lqV88HivxYLsyu9x+AJiZ7WlmaUBv4JXw+ePrzeywcL6zw/JNwHtAr7C8L/BKJcYrIiKFVFrScPf1QD/gWeBL4Ct+v8jdB7jTzL4CGgD3hOUDgfPN7EugCzCssuIVEZHNpbx5yt1b5Xv9JnBgEfN8StC7qnD5AuDIFIYnIiJloDvCRUQkMiUNERGJTElDREQiU9IQEZHIlDRERCQyJQ0REYlMSUNERCJT0hARkciUNEREJDIlDRERiUxJQ0REIlPSEBGRyJQ0REQkMiUNERGJTElDREQiU9IQEZHIlDRERCQyJQ0REYlMSUNERCJT0hARkciUNEREJDIlDRERiUxJQ0REIlPSEBGRyJQ0REQkMiUNERGJLL0qNmpmbwPbAZvCoguAhsAYoB7wlLsPC+dtC0wAsoAZwIXunl3ZMYuISBWcaZhZDGgNHOjubd29LfAZMBE4Fdgb6GRmJ4WLTAEGuXtrIAYMqOyYRUQkUBVnGhb+fs3MmgIPAZ8D89z9OwAzmwKcYWZfAvXcfVa4zGTgJmBc5YYsIiJQNUmjCfAmMBioA7wDjAQW55tnMbAzsGMx5ZE1bdpgC0KVitS8ecOqDiGlanP9anPdQPUri0pPGu7+H+A/ue/N7GHgZmBmvtliQIKg+SxZRHlky5evJZFIlj6jbJEoX8qlS9dUQiSpsbXXrybXDVS/stQvHo+VeLBdFdc0OpvZMfmKYsB8oEW+sh2ARcCPxZSLiEgVqIout42B0WZW18waAucA1wFmZnuaWRrQG3jF3RcA683ssHDZs4FXqiBmERGhCpKGu/8TmAZ8DMwBJoZNVv2AZ4Evga+AqeEifYA7zewroAFwT2XHLCIigSq5T8PdhwPDC5W9CRxYxLyfAgdVUmgiIlIC3REuIiKRKWmIiEhkShoiIhKZkoaIiESmpCEiIpEpaYiISGRKGiIiEpmShoiIRKakISIikSlpiIhIZEoaIiISmZKGiIhEpqQhIiKRKWmIiEhkShoiIhKZkoaIiESmpCEiIpEpaYiISGRKGiIiEpmShoiIRJZe1QGIyNajYVZd6mbWKXJazqaNpNXJKHbZ7I0bWLFqY6pCqxBbUr+N2RvJSC96es7GjaRlFL9sZVLSEKlGStrpAKzfsIk1q9dXYkQVq25mHXpf/Y8ipz0+qg9zRvUvdtn9Lx9H8+YNi5xWXXaqW1K/DldPoN+kS4qcNvncu5ne99xil+366KSyBboFlDRqoZJ2PDV9p7M171Qh2PGsoXLrl71pU7E7a4CNGzawanXqzwAy0uukZKdaXeqXKiXVrzx1U9KoRI2y6pGRWfSffOOGbFatXlch2yntaKeydzpQcV/c6rhThYr/x6xO0uvUYcy1FxQ7/fLbxgOqX3VVUv3KUzcljUqUkZnOrUOnFjntuhE9KzmaylXRX9zyyGqUSWYJTRgbNm5k9aoN5Vp3da/fltRNJD8lDclT2k61urQbl1dmRkaxzRsAD/e5vcRmiuqupPrV9LpJ9VEjkoaZ9QaGAXWAu9z9/ioOqcZKZJfcflvSTrWkduPKvBBXktLqV5K0jDq1tn4l1Q2qT/2k+qv2ScPMdgJGAB2ADcC/zextd/+yaiOrmeLpdYrtwdHh6gmVHE3FU/1EUqvaJw3gWOAtd/8FwMymAj2Bm0tZLg0gHo9tNqFBg0wyt6Cv+Ko1myIFXpRGjesXO62oWMurWZNtip2WkdW0+OUabFvieus1K37ZkuoGkNW4+GXLUveS6gaqX3FKqhuUXL+S6gYVV7+S6gaqX0nKW7/Cdcv3Pq2o+WPJZLLEQKqamV0LbOPuw8L3/YGD3P38UhbtDLyX6vhERGqpLsDMwoU14UwjDuTPbDEgEWG5jwgqvRjISUFcIiK1URrQgmAfupmakDR+JNj559oBWBRhuQ0UkSVFRKRU/1fchJqQNN4AbjSz5sCvwOlAaU1TIiKSAtV+lFt3XwgMBd4GPgEed/cPqzQoEZGtVLW/EC4iItVHtT/TEBGR6kNJQ0REIlPSEBGRyJQ0REQksprQ5TblzKwncC3B3yMOPOruoyto3TsCE9y9a0Wsb0uZWTowBPgzwU2TacAjwG3uXqN7RZhZK+BroPC4ZN3d/YdyrrMjcKG79zezd4Ab3f2dLYmzIphZA2AkcAJBV/TVBLG9Wcpy/YAj3b1fqmPcEmaWBdwGHAFkAyuAK4AsgnoeWXXRlU2h72USyCC41+xcgnvJjnT3+RHWcyTVoO5bfdIIB0S8A2jv7svDf8Z3zczd/aUtXb+7LwKqRcIIjQW2Bw5195XhP+fzwCqgNowevMjd21bUytx9NlD8MzqrgJnFgJcJuqDv4+4bzawdMM3MeleHpLYlzCwOTCfoZt/W3bPN7CjgFWBglQZXfgW+l2Z2B1AhB6aVbatPGkAzgiHX6wPL3X2tmZ0DrDez+cDTwHHhvH9x94/NbE9gHNAU+A0YHJbvCkwCtgvL+xMcAb7j7q3MbHtgPLALwVAo17r7G2Z2DDCK4ChkBXCWuy+r6Iqa2c4EZxg7uftKAHdfbWZ/BfYtIb4bgZ2AvYBdCc6cRhQ+as09Ege+Af4BbBOu52J3n1XR9YnKzPYD7gUaEHw2t7n7A2G9WgKtgeYEoykfAxwMfAqcSXCkW+DozsweA2a4+0Ph+3eAIe7+QSVV6QiCz+Ho3LPD8Pt3CzA8rNcvwL5AL2B/gkcLrAYWAGvDuDsBdxJ895cBF7j7d2F98pZ3908qqV65jiL4XG5w9wSAu79tZucSfIYAmNkRBJ9ZfaAxcJm7vxg+SuFqguGDviP4zjejGn0nCRLibeHr68OkXx/oCywH3gJauXsiPMMYQnBmCfz+v+bu74RnMrn7mMkEZ57tCf4m1wFnAwcCL7j7FVsa+FZ/TcPdPwVeBL41sw/NbCSQ5u7fhLP86u7tgOsJmnEIf1/t7u0J7k5/MiwfCzzr7vsR7DyHFdrc3cBEd+8AnAKMN7OG4XwXuntH4HWCDzwVDgK+dPcV+Qvd/St3f7aE+AAOAI4n2KFeY2aNS9jOecA/w/pcTzB4ZGXZ0cw+yfdzFUHyvsXdOxHskPIf4e0PHEnwOU4i+Mfcj+AzOKCYbUwk+EckPFBoXokJA6ATMLuI5sQZ4TSAz9zdgCUEBySHA4cCDQHMLAOYAPQOv8d3AA/lW9dn7m5VkDAA2gGf5CaMXO4+naA+uQYD/cP4+wO3hOW3AMeH3+PvgDZU7XeyADOrQzBS93/Coi/Dfcy9wJXhvuc7gu8lBIlkchk2saO7Hwr8neA7fSHQFhhgZo22NH6daQDuflF4lHY8QRvxLDPrE05+MJznZTN7JDxa7wRMMrPcVTQws6YER4BnhfNPB6aHRwG5jgXamFnusO51gD2Al4DnzewF4EV3fz01NQXyDf4YXssZRnBdYz2wWzHxAbzt7huBJWb2C1DSl+8N4LncJhPgvoqtQok2a54yszTgxHDE5P3Jd7QKvB42fywAFuc+p8XMFgJNitnGOwTJqRVB8ni0QmtQuiRF/+9m8Pvnm5vE/gD8291/BjCzKQRnU60Jv3v5vsdZ+dZVmUmwsAREetD7n4FuZnYGcAi/f64vA++b2fMEB3GfmNk2VN13EsKDmfB1JvAhcA3BPueFsPy/BMMkQXhgYmazCD6vgQR1jOKV8PcC4At3XwIQ/t82IWiKLret/kzDzE42s17uvtDdJ7n7mcDFBEcmEFyEyxUn3MG6e9vcH4Kj71+ATfnWGzOzfQptLo2gSSH/cp+7+50ERxXfAKPMbGiFVzQwG9gnvI6Bu08N4+hO0DxTZHzhsvn/iZMEow3n/s5VJ1zv+8A+wL8ImkdeTlF9onoa+CPBhcjCf9v8D+/OJoLwCP8RggOEXsBjFRBjWXwAdAyPWPM7lOAzBlgX/i78GeXWMQ34Nt9n3YGCR9/rqDqzgfbhtZs8ZnYrBevyHsHZ8xyCZqoYgLtfQrDzXQFMMbM/V4Pv5KJ8+4y93f2c3GcE8ftnkv+zeoagWbwnMN3dCyfR/PMW/h6U+TtdFlt90iC49nBb7hlB+EVtC3wcTj8zLP8j8D93XwDMM7M/h+XHETQLEP4+M3x9LOFZSj5vEV7ICxPKF0B9M/sAaOjudxG0MaekecrdvyfYwT2S27wU9qbqRtD+W2R8JaxyGbB3mCB3I2zOMbNRwJ/d/RFgUKrqUwbHAde7+4vASZB39rElJhOc9n8fdnaoNO7+HsFR6V25icPMOhCcNf6t0OwzgUPNbKfwAnOvsPwrYFszyx1B+i/A4ykPPpr3CJqhbsj9nMzsBILeRtuF77clOFu6nuDI+lQgzczSzWwesMzdbyM4C2xXDb+TJXL33wjqdStFN00tI7jmBNCjcqIKbPVJw93fBm4C/mlmTvDPlMPv/3yHhaeVVwLnhGV9gP5m9hnBxaxe4dHnIOD0cP6b2Hw03sHAIeFyTxF8idcQXKyabGZzCM5wrklFXUMDgfeBt8M45hEcZZ5UQnzFeQP4AXCC6yG5Q9HfC/QM/w7PE7TJVqUbgZlm9iXBMPvzCZriys2DLrzfU7a25op0GsHw/1+E9bqb4PN6J/9MYbPUYILP6kOCi+G4+wbgDOCO8PM+h9/PrqtU+L90CkHz2RdhfEMIeiH+HM7zC/AwQfL8H8G1mvoETT/XA6+b2WyCJp2RVL/vZBRPAquLuV42ChhoZnOBepUZlAYsLEHYeypSH2rZeoRnoy2Ad4H9wh2wSIUJz7BGAEvcfUxVx5OfLoSLlN3pBF2uL1LCkBSZTdAEdUpVB1KYzjRERCSyrf6ahoiIRKekISIikSlpiIhIZEoaIiISmXpPiZTCzO4hGLsJgruKv+P3O6YPdfeU3T0dDj7YzN0HlTDPfKCnByPyiqSUkoZIKdz94tzX4Q66j3bQsrVS0hAph3AAvHEEw8U3BdYAvQnuEp8N3O/uY83sPOBSgnG8DiC4kzeT4ObA19090l3YZnY8cIe77x++b0xwxrN7xdVKpHS6piFSPicBK939UHdvDXwEDAqbqs4EbjazrgR39fYMxxK6hGAMrIMJmrlOCceMiuJ1gtGUO4bvzwKmFR7mXiTVlDREysHdpxKMFzbYzO4mGKW4QTjtc8LxzICr3N3Dxc4BGpvZdQTPXqlHwWHaS9pekmC47H5h0bkUfP6FSKVQ0hApBzO7iGDAvN8IRod9goLDdu9LMLhe/mcgzCAYdO8r4GZgYaFlSjMROMPM2gKN3f3d8sYvUl5KGiLlcwIw2d0fJhjltzvBMyows9OAowmfdmhmPcJrEJ0IHgv7HLAzsGfuMlG4+0KCkWrHEzx1T6TSKWmIlM/twAXhsN3vAXOBPc1sF+AB4Gx3X0rQJPUgQTPUbcBcM/uCYPj79wkSR1k8RPA41EdKm1EkFTRgoYiIRKYutyJVyMwaEpypFGWNu3cpZppIldCZhoiIRKZrGiIiEpmShoiIRKakISIikSlpiIhIZEoaIiIS2f8D9j4OIcO620YAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set(rc={'figure.figsize':(15,10)})\n",
    "g = sns.barplot(data=results,\n",
    "                x='Taxa_lvl', y='Correct', hue='Tool',\n",
    "                ci = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3oAAAJPCAYAAAAqk6bSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAABI8ElEQVR4nO3dd3RU1d7G8WfSCS0YEkK5FkABKVKliHQJBkIJURAkBgFBmqLSQwSvSBMjoPQmCgoqUoQLKIhKUwFFVBCRCwJCSOgJpM68f3Cd10iUSTKTsvP9rOVyTtvndyabmTw5+5xjsdlsNgEAAAAAjOGW1wUAAAAAAJyLoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYTzyuoCcuHgxUVYrjwEEAAAAULi4uVlUqlTRv11eoIOe1Woj6AEAAADAXzB0EwAAAAAMQ9ADAAAAAMMU6KGbmUlPT9PFi3FKS0vJ61LgRG5u7ipSpJiKFSspi8WS1+UAAAAA+ZpxQe/ixTj5+PiqaNEgAoEhbDab0tPTdPXqJV28GKfbbgvM65IAAACAfM24oZtpaSkqWrQEIc8gFotFHh6e8vPzV0pKUl6XAwAAAOR7xgU9SYQ8Q1ksbpK4yyoAAABwK0YGPQAAAAAozIy7Rs8Er78+Td99960k6fjxYypbtpy8vX0kSfPmLba/vpUzZ35XREQ3ffLJly6rFQAAAED+49Kgl5CQoO7du2vu3LmqUKFChmWHDh3S2LFjlZiYqPr162vChAny8CB3StKzzw63vw4PD9WLL76sqlXvzcOKAAAAABQkLhu6eeDAAT322GM6fvx4psuHDx+u6Ohobd68WTabTatWrXJVKcZYunShHn/8ET3xRHdFRY3Q+fPxkqRz52I1YsQwRUR0U69ej2rFimV5XCkAAACAvOSyoLdq1Sq9+OKLCgy8+Vb4p0+fVlJSkmrXri1JCgsL06ZNm1xVihE2bFinPXt2acGCZXrrrfd0112VNHHiBEnSSy+NU9269bRs2UrNmbNYmzf/R59+ujmPKwYAAACQV1wW9CZOnKj69etnuuzcuXMKCAiwTwcEBCg2NtZVpRhhz55dCgkJVZEiRSRJjzzymPbt+1oJCQk6ePCAwsIelSQVK1ZMISEdtGfPrrwsFwAAAEAeypOL4qxWa4ZHINhstmw9EsHfv9hN886dc5OHh1k3E3V3d5PNZpW7+/8fm7u7lJ6eLg8Pi2w2mzw8Mh631Zoud/cb0ya9H25ubgoIKJ7XZQAAAAD5Wp4EvaCgIMXFxdmn4+PjMx3ieSvnzyfIas34XDWr1aq0NGuOa8xP0tOtuv/+xlq/fq1atWqrIkWK6L333lXt2nXl41NU1avX0KpV76l798eVkJCgjRs/1qOPPqb09Bvvg0nvh9VqVVzc1bwuAwAAAMhTbm6WTE98/SFPgl758uXl7e2tffv2qV69elq7dq2aNWuWF6UUGB06dNK5c7Hq1+8J2WxWlS//L0VH/1uSFB39sl57bYo2blyv1NRUPfRQO4WEhOrs2TN5XDUAAACAvGCx2Wy2W6+Wfa1atdKyZctUoUIF9evXT0OHDlXNmjV1+PBhRUVFKSEhQdWrV9ekSZPk5eWVpbYzO6N39uwJBQXd4cxDQD7CzxcAAAC49Rk9lwc9VyLoFT78fAEAAIBbBz1z7tIBAAAAAJBE0AMAAAAA4xD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMkyfP0ctNxUv4yMfb0+ntJiWn6uqVJKe3CwAAAAA5ZXzQ8/H2VI8Ry53e7oqpPXVV/xz0nn66j7p2fVRt2gTb512/fl1du3aQv7+/hg0bobp169uXTZw4XnXq1FNISKjCw0Pl6empd99dbV+elpamzp3bqXHjpho7dnyWa/7ppx+0ffs2DRw4NMvbAgCQ20qWKCIvb+f8qpKSnKbLV647pS3kP/QVZEVh6S/GB7281L59R23ZsilD0Pv8822qW7e+Ll26eMvtk5OT9euvR1WpUmVJ0r5930iyZLue48f/q4sXL2R7ewAAcpOXt4deGfuBU9oaMzHcKe0gf6KvICsKS3/hGj0XatXqIR08eEBXrly2z9u8eaPat+/o0PbNm7fS9u1b7dNbt25Rixat7dODBz+l2bNnqH//3urWrbN2794pSbpw4bxGj35eTz75uPr2jdA333ylq1evauHCudqx4wu99dYiJx0hAAAAgPyIoOdCvr6+evDB5tq27VNJUnx8nH777YTuv7+RJGnKlJcVGdnD/t+OHV9k2L5ly9b6/PNtkqTU1FQdPXpE995bPcM6qalpmjdviYYMeU4LFsyRJM2Y8arat++oxYvf0eTJr2natFfk7u6mvn0HqGnTZnriiT6uPnQAAAAAeYihmy4WEhKqhQvnqnPnrtqy5T8KDg6Ru7u7JGnkyKibrtH7s4CAQBUtWkzHj/9Xp0+fUoMGjW5qv2HDxpKkihUr6erVK5KkvXu/1okTJ7Rw4TxJN67tO336lCsODwAAAEA+RNBzsdq16+r8+XjFxp7V5s3/0SuvTMvS9i1bttH27Vt18uRv6tath44e/SXDci8vL0mSxWKRzWaTJKWnWzVz5hyVKFFSkhQfH69SpUrpl1+OOOGIAAAAAOR3DN3MBe3atdeyZYtVokQJlS9fIUvbtmrVRtu2faITJ47rnnuqOrRNvXr1tXr1+5Kk//73mCIiuik5OUnu7u5KT0/Pcv0AAAAAChbjz+glJadqxdSeLmnXUX88LmH06Ogs76d06QAVK1ZcderUc3ibYcNGaOrUiXriie6y2WwaN+4l+foWVbVq1bV48XzNmTNLTz89JMu1AAAAACgYLLY/xvsVQOfPJ8hqzVj+2bMnFBR0Rx5VBFfj5wsAhUdAQHGn3gI9Lu6qU9pC/kNfMZ9fcS95+ng7rT0T+oubm0X+/sX+drnxZ/QAAAAAFGyePt7aGNHbKW2FLFvilHbyO67RAwAAAADDcEYPAJBrSpYoIi9v53z1pCSn6fKV605pCwAA0xD0AAC5xsvbw6nXRQAA/l/xEj7y8fbMcTvpqSly9/RyQkVSWkqyLl5OcUpbyBqCHgAAQCFQqqSXPLxyfjMLfnHPv3y8PdVjxPIct7Niak/tm9rXCRVJ9UYslER/yQsEPQA5xnA8AMj/PLy8nfLLO7+4AwWD8UHPWX+9+qvs/jVr4sTxqlOnnkJCQu3zFi2aJ0nq06e/Bg9+Sk8++ZTq1q2f6TZNm9bXjh17b2r3r/OnT5+i//73V02d+rp8fX2zXCeQFQzHAwDXcNZQPACFj/FBz1l/vfqr/PzXrNdff1W//XZcr746Uz4+PnldDgAAyCZnDcWTbgzHA1B48HgFFxozZri2b99qn37yycd19eoVl+5z1qwYnTjxX02dGkPIAwAAAAop48/o5aXg4BB98sl/1KJFa508+ZtSUlJUvHgJLVw4V6tWvWtf78KF8+rUKSzH+5s9e6ZWrlyumTPnytubkAcAAAAUVpzRc6EmTZrqhx8O6tq1RH366WYFBz8sSerbd4CWLl1h/+/PIc/N7eYfic1mk8ViueX+jh8/pqioCZo06SUlJCQ470AAAAAAFCgEPRfy9PTUAw88qB07vtC2bZ/ooYfa3XKb4sWL3xTSLl68qOLFS9xy24kTp6ldu/aqUaOWpk17Jdt1AwAAACjYCHouFhwcovfee0clS/opKKjsLdevV+9+bdq0QWlpaZKk3347rp9/PqQaNWrdcltPzxt35XruuZE6ePCANmxYl7PiAQAAABRIXKPnYrVq1VZCQoI6d3bslvEdO3bR6dOnFBnZQ25uFnl7e+vFF/8tPz8/+zoPPfSg/XWZMmX1zjurMrRRvHhxjR07XqNHv6CaNe/T7bff4ZRjAQCgoEpLTVVAQPEct5OSnKzLV/LnXbfhHM7qKxL9pTDIz58txge9tJTk/z0KwfntOmrVqrX212PHjr9peZ8+/e2vPTw8NGTIMA0ZMizTtjJ7hl5m8+vVa6AtWz53uEYAAEzm4emp10b3v/WKt/DcpHnKr49XgnM4q69I9JfCID9/thgf9G481Jx/YAAAAAAKD67RAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxh/180SJb3l7eXl9HaTU1J05bLjj1gAAAAAgNxifNDz9vJS5JJnnN7u0t4zJN066CUmJmju3Df13Xf75O7uoeLFi2vw4GEqUaKEhgzprw8+WJ9h/aZN62vHjr3av3+vRo4cpvLl/5Vhee/e/dS8eUs1bVpflSvfI0my2WxKSLiqhg0b6/nnR8nd3f0fl48e/bxuv/1ODR78rL3dNWs+1MaN6zVnziK5u7vn7M0BAAAAkKeMD3p5yWq16oUXnlHduvW1ZMkKeXh4aP/+vXrhhaGaNm3GLbevUqWa3nhj/t8uX7p0hf11YmKCevXqpq+/3qPGjR/4x+XDh49RZORjatu2ne65p6ri4+O0ZMl8zZo1j5AHAAAAGIBr9Fxo//69io09qz59+svD40amrlu3vsaMiZbVmu7UfV26dEnJyUkqUaLkLZcHBATq6aeHaMqUibJarYqJmaqePSN1++13OrUmAAAAAHmDM3oudOTIz7r77nvk5pYxTzdu3FRnzvyu+Pg4RUb2+Nvtf/750E3LZ8yYrZIl/SRJkZE9lJaWpkuXLuiOO+7SM88MV/XqNezr/tPyDh06a+vWT/TSS+N05coVPfJIdycdNQAAAIC8RtBzITc3i7y8vP92eenSARmGV0o3rtH7g6NDN1euXK6NGz/Wgw82z9LykSOjFB4eqvffXyeLxeLYQQEAAADI9xi66UJVq96rI0cOy2azZZg/b96b2r9/r9P2061bT/n7++vNNzO/7u/vlgcFlZUklS1bzmm1AAAAOMqvuJcCAorn+D8AN+OMngvdd18dlSp1mxYvnq/IyL5yd3fXV1/t1saN6/TKK9Oduq/Bg4epT5/H1blzV1WufHeWlwMAAOQ2Tx9vbYzoneN2QpYtcUI1gFmMD3rJKSn/exSC89u9FYvFosmTX9OsWdMVEdFNHh4eKlnST9OmzVDx4iVuuX1m1+i1bt1WvXpF3rRuxYqV1K5de73xRoxef312lpcDAAAAMIfxQe/GQ83z7sHmfn5+Gjfu35ku++sz9CRpx44bQzrr1q2vTz758m/b/WO9Pxs5Msrh5f+0HgAAAICCjWv0AAAAAMAwBD0AAAAAMIzxQzcBAGZKS0112t32UpKTdfnKra+9BgCgoCDoAQAKJA9PT702ur9T2npu0jxJBD0AgDkYugkAAAAAhiHoAQAAAIBhjB+66VfcS54+3k5vNzUpWZeu3nqYz5kzv+uxx8J0550VJUk2m1WJiYl6+OEO6tPnxpCjY8eOKiKiu15+eYpatGht33bw4KcUF3dORYr42ueFhnZW166PSpL69Oklf//Smjo1xr5848b1+vbbfRo7drx93v79e7V48Xy98cb8HB0zAAAAgILB+KDn6eOtjRG9nd5uyLIlkgNBT5JKlw7Q0qUr7NPx8XHq3r2LWrduqzvvvEsbNqxTy5ZttHbt6gxBT7rx7Lu6devf1ObRo7/Iy8tTR48eUWzsWZUpE5SzAwIAAABgDOODXn4UHx8vm80mX19fpaWlacuWTXrzzQV6+ukndfr0KZUvX+GWbWzcuE716zfUlSuXtX79GvXtOyAXKgcAAAAcl5LmvDskI2sIerkgPj5OkZE9lJKSrMuXL6lq1ep65ZVXFRhYRl98sV1BQUG6/fY79OCDLbR27WoNHDjUvu2UKS/bh276+vpq9uyF9nA4a9Y8XblyRS++OFqRkX3l4XHjx7ljxxeKjOxhb+P69WsKCAjM3YMGAABAoefl4anIJc/kuJ2lvWc4oZrChaCXC/4Yumm1WvXGGzE6fvy/atCgoaQbZ+batAmWJLVu/ZAmTBinfv2elqenp6TMh27u3Pml/P1L6667Kspms8nNzU07d36p5s1bSpKaNm2W6TV6AAAAAAoHgl4ucnNz08CBz6h37x569923FRISqj17dunnnw/r/fffk81m09WrV/T559vs4S8zGzeuU2zsWYWHh0qSEhMTtHbtanvQAwoyZz0EmwdgAwCAwoygl8s8PDw0aNCzGjdulNLS0lSv3v2aPn2mffmiRfO0Zs2Hfxv0Llw4r2+++VorV35kH455+vQp9ejRVadPn8qVYwBcyVkPweYB2AAAoDDjOXp5oFGjJqpRo6b+85+P1aVLeIZlYWGP6tChH3XixPFMt920aaMaN26S4Zq78uUrqGnTZlq37iNXlg0AAACggDD+jF5qUvKNRyG4oF1HlC1bTh98sP6m+TExb2a6fqlSpbR1605JyvS5dz169Mp0u4kTp9lfh4SEZlhWt279TB/RAADIG8VL+MjH2zPH7SQlp+rqlSQnVAQAMI3xQe/S1RSHn3cHAEBu8PH2VI8Ry3PczoqpPXVVBD0AwM0YugkAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYx/q6bJUsUkZe38w8zJTlNl69cd3q7AAAAAJBTxgc9L28PvTL2A6e3O2Zi+K1XknT48E9as+ZDpaena9++b1SiRElJUmpqisLCHlHXrt20f/9eLV48P8Nz886c+V1DhvTP9Bl8f9i/f69Gjhym8uX/JYtFSk5OVsWKlTRmzIvy9S2aswMEkCXOei6axLPRAABAzhkf9PJa1ar3atSoezVx4nj17TvA/jDzCxfOq1u3zqpX7/4ctV+lSrUMATEqaoSWLVuiAQMG56hdAFnjrOeiSTwbDQAA5BxBz8X+OFtXtmy5DPNvu81ft99+p44fP2Y/y/d3kpKSNGXKyzp69Ijc3NzUvfvjevjhDpmuW6dOPX311R6n1V/YOessDWdoAAAAkJsIennkl1+O6PTpk6pSpZrOnPldP/98SJGRPezL09JS7a8XL56nkiVL6u23V+nSpUvq1+8J3X13lZvavH79unbs+EJ16tTLlWMoDJx1loYzNAAAAMhNBL1ctHDhXK1a9a6s1nR5e/to+PCxKlu2nM6c+f2mIZh/XKMnSfv27dWoUeMkSX5+fnrwwWb69tt9qlSpcoaAmJaWpnr16qtbt565f3AAAAAA8g2CXi768zV6WWGzWf8yLaWnp0m6+Ro9AAAAAOA5egVA3boNtGHDWknSpUuX9OWX21WnTv08rQkAAABA/mX8Gb2U5DSHH4WQ1XZzS+/efTV9+hRFRHST1WpVRMSTqlKlqvbv35trNQAAAAAoOIwPenn9UPO6deurbt1/PvuW2Tply5azP0OvaNFiio7+d7baBgAAAFD4MHQTAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMY/3iFkiW85OXt7fR2U5KTdflKyi3XO3z4J61Z86HS09O1b983KlGipCQpNTVFYWGPqGvXbtq/f68WL56vN96Yb9/uzJnfNWRIf/sjFjKzf/9ejRw5TOXL/0sWi5ScnKyKFStpzJgX5etb9G+3Cw8P1axZ81S2bLksHPHNhgzpr1mz5uWojcxs3Lhe3367T2PHjnd62wAAAEBhYHzQ8/L21muj+zu93ecmzZN066BXteq9GjXqXk2cOF59+w5QSEioJOnChfPq1q2z6tW7P0d1VKlSLUNAjIoaoWXLlmjAgME5atcR3367z+X7AAAAAJB1xge9vPbH2bq/nj277TZ/3X77nTp+/Jj9LN/fSUpK0pQpL+vo0SNyc3NT9+6P6+GHO2S6bp069fTVV3tuWdfixfN19OgReXl5a/jwMapc+W5duHBe06a9otjYWLm5ual//0Fq0KCh9u79WrNnz5TFYlHx4sU1fvwrWrp0gSSpX78ntGDBW+rQ4SE1b95SP/zwvXx9fRUd/bLKli2n8PBQ3XtvDf3yy8+aPXuh9uzZpffff1dWq01VqlTVc8+NlLe3tzZt2qC33lqkokWLKSgoSEWK+Dr4DgMAAAD4K67RyyO//HJEp0+fVJUq1SRJP/98SJGRPez/DR/+jH3dxYvnqWTJknr77VWaMWOuFi9eoKNHf7mpzevXr2vHji9Uo0bNW+6/QoV/acmSFYqM7KOJE1+UJM2Y8arat++oxYvf0eTJr2natFd07Vqi3nprkYYPH61Fi95WgwYNdeTIYT377HBJ0oIFb0mSLl26qBo1aumtt95T69Zt9frr0+z7atSoid59d7UuXryo9evXaM6cxVq6dIVKlbpN7777tuLj4zRnzky9+eYCzZ27WNeuXcv+GwsAAACAM3q5aeHCuVq16l1Zreny9vbR8OFjVbZsOZ058/tNQzD/uEZPkvbt26tRo8ZJkvz8/PTgg8307bf7VKlSZXtAlKS0tDTVq1df3br1vGUtoaGdJUmNGzfVSy9F6+rVq9q792udOHFCCxfOs7d3+vQpNW3aTGPGDNeDDzbXgw82V4MGjW5qz8vLW+3atZckPfxwB82b96Z92b331pAkffvtXp06dVL9+/f+X/upuueeqjp48IBq1Kil227zlyS1bfuw9u37xvE3FgAAAEAGBL1c9Odr9LLCZrP+ZVpKT0+TdPM1eo5yd3f/U3s2eXh4KD3dqpkz59iHksbHx6tUqVK6++4qeuCBZtq160vNnj1TLVr8qCee6JOhPTc3iywWiyTJarVlaN/7fzfDSU+3qlWrNvazgdeuXfvfTWq+ls2WeW0AAAAAso6hmwVA3boNtGHDWknSpUuX9OWX21WnTv0ctbllyyZJ0ueff6Y777xLRYoUUb169bV69fuSpP/+95giIropOTlJ/fo9oWvXEvXooz306KM9dOTIYUk3Alla2o3AmZSUpB07vpAkbdy4Tg0bNrlpn3Xq1NMXX2zXxYsXZLPZNH36JK1atUK1atXWjz9+r7i4c7Jardq27ZMcHRsAAABQ2HFGrwDo3buvpk+fooiIbrJarYqIeFJVqlTV/v17s93myZMnFBnZQ76+vvbHGAwbNkJTp07UE090l81m07hxL8nXt6j69x+kiRMnyN3dXb6+vho5MkqS1LRpM0VG9tCiRW9Lkj777FMtWDBb/v4Biooaf9M+7777HvXu3U9Dhw6QzWZT5cr36PHHI+Xt7a1nnx2uZ58dKB+fIrrzzruyfVwAAAAACkHQS0lO/t+jEJzfriPq1q2vunX/+exbZuuULVvO/gy9okWLKTr639lqOzN/92y+0qUDNHXq6zfNr1//fr311rs3zZ84cVqG6XHjXrrlvkJDO9uvD/yzli3bqGXLNv9QNQAAAABHGR/0bjzU/NbPuzPNkCH9dfXq1Zvmd+4cps6dw/OgIgAAAAC5xfigV1jNmuX8s5j/ZMeO7A8jBQAAAOBc3IwFAAAAAAxjZNCz/fle/TDGjcdMWPK6DAAAACDfMy7oeXh4KTHxCmHPIDabTWlpqbp0KV5eXj55XQ4AAACQ77n0Gr3169drzpw5SktL0xNPPKGePXtmWP7jjz8qOjpaqampKlu2rKZNm6YSJUrkaJ+lSgXo4sU4JSRcylE7yF/c3NxVpEgxFStWMq9LAQAAAPI9lwW92NhYxcTEaPXq1fLy8lL37t3VsGFDVa5c2b7OxIkTNXToUDVv3lyTJ0/WokWLNGzYsBzt193dQ6VLl81p+QAAAABQYLls6OauXbvUqFEj+fn5ydfXV8HBwdq0aVOGdaxWqxITEyVJ169fl48Pw/IAAAAAIKdcFvTOnTungIAA+3RgYKBiY2MzrDNq1ChFRUWpadOm2rVrl7p37+6qcgAAAACg0HDZ0E2r1SqL5f/vkGiz2TJMJyUlaezYsVq6dKlq1aqlJUuWaOTIkZo/f77D+/D3L+bUmgFXCggontclFDoF+T0vyLUXVAX1PS+odRdkvOe8B1nBewVHObuvuCzoBQUFae/e/3+IdlxcnAIDA+3TR44ckbe3t2rVqiVJ6tatm2bMmJGlfZw/nyCrlbtrwnWc+Q8uLu6q09rKb/Lrl1huvufOfg9M7S/5ta9IBbe/mNpXpPzbXwpqX3EmZ70H+fX4nIn+Akdlta+4uVn+8cSXy4ZuNmnSRLt379aFCxd0/fp1bdmyRc2aNbMvv+OOO3T27FkdO3ZMkrR161bVrFnTVeUAAAAAQKHhsjN6ZcqU0bBhwxQREaHU1FSFh4erVq1a6tevn4YOHaqaNWtq0qRJevbZZ2Wz2eTv769XXnnFVeUAAAAAQKHh0ufohYaGKjQ0NMO8BQsW2F83b95czZs3d2UJAAAAAFDouGzoJgAAAAAgbxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwjEdeFwAAyP/8invJ08c7r8tAAUBfAYD8gaAHALglTx9vbYzoneN2QpYtcUI1yM+c1Vck+gsA5ARDNwEAAADAMAQ9AAAAADAMQQ8AAAAADMM1ekAhxQ0TAAAAzEXQAwopbpgAAABgLoIeAAAAHJaSlqqAgOJ5XQaAWyDoAQAAwGFeHp6KXPKMU9pa2nuGU9oBcDNuxgIAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGMYjrwsAAGRkTUtVQEDxHLeTlpKsi5dTnFAR8itn9RWJ/gIApiHoAUA+4+bhqX1T++a4nXojFkriF3eTOauvSPQXADANQzcBAAAAwDAEPQAAAAAwDEM3gVzAdTQAAADITQQ9IBdwHQ0AAAByE0M3AQAAAMAwBD0AAAAAMAxBDwAAAAAM49Kgt379eoWEhKht27Zavnz5TcuPHTumXr16qWPHjurTp48uX77synIAAAAAoFBwWdCLjY1VTEyMVqxYoTVr1mjlypU6evSofbnNZtPTTz+tfv36ad26dapWrZrmz5/vqnIAAAAAoNBwWdDbtWuXGjVqJD8/P/n6+io4OFibNm2yL//xxx/l6+urZs2aSZIGDBignj17uqocAAAAACg0XPZ4hXPnzikgIMA+HRgYqO+//94+/dtvv6l06dIaM2aMDh06pIoVK2rcuHFZ2oe/fzGn1QsUJM56Jp/peJ94D7KC94r3wFG8T8gK+gsc5ey+4rKgZ7VaZbFY7NM2my3DdFpamr7++mu98847qlmzpl5//XVNnjxZkydPdngf588nyGq1ObVu4M/y64dzXNzVHLeRX4/NmZzxPjkqv76fznoP8uvxORP9hc8WR9FXkBX0Fzgqq33Fzc3yjye+XDZ0MygoSHFxcfbpuLg4BQYG2qcDAgJ0xx13qGbNmpKkDh06ZDjjBwAAAADIHpcFvSZNmmj37t26cOGCrl+/ri1bttivx5OkOnXq6MKFCzp8+LAkadu2bapevbqrygEAAACAQsNlQzfLlCmjYcOGKSIiQqmpqQoPD1etWrXUr18/DR06VDVr1tSbb76pqKgoXb9+XUFBQZo6daqrygEAAACAQsNlQU+SQkNDFRoammHeggUL7K/vu+8+ffDBB64sAQAAAAAKHZc+MB0AAAAAkPsIegAAAABgGIIeAAAAABjGoaA3ZMgQ7dq1y9W1AAAAAACcwKGg99BDD2n27NkKDg7WokWLdOnSJReXBQAAAADILoeCXseOHfXOO+9o9uzZOn/+vMLDwzV8+HAecA4AAAAA+ZDD1+hZrVadOHFCx48fV3p6uvz9/TV+/HjNnDnTlfUBAAAAALLIoefoxcTEaPXq1frXv/6lHj16aMaMGfL09NS1a9fUsmVLDR061NV1AgAAAAAc5FDQu3DhghYsWKCqVatmmO/r66vp06e7pDAAAAAAQPY4NHRz0KBBeu+99yRJx44d08CBAxUXFydJatq0qeuqAwAAAABkmUNBb9SoUapYsaIkqXz58rr//vs1ZswYlxYGAAAAAMgeh4LexYsXFRERIUny9vZWZGSk/YweAAAAACB/cSjopaenKzY21j4dHx8vm83msqIAAAAAANnn0M1YIiMj1blzZz344IOyWCzatWuXRowY4eraAAAAAADZ4FDQCw8PV40aNbRnzx65u7urT58+uueee1xdGwAAAAAgGxwKepIUFBSk4OBg2Ww2paena+fOnXrggQdcWRsAAAAAIBscCnozZszQ/Pnzb2zg4aGUlBRVrlxZ69evd2lxAAAAAICsc+hmLGvXrtVnn32m4OBgbd68WZMmTVLlypVdXRsAAAAAIBscCnq33XabAgMDVbFiRR0+fFidO3fWkSNHXF0bAAAAACAbHAp6Hh4e+u2331SxYkXt3btXaWlpSk5OdnVtAAAAAIBscCjoDRgwQOPGjVOLFi30ySefqEWLFmrUqJGrawMAAAAAZINDN2NJS0vTW2+9JUlas2aNTpw4oSpVqri0MAAAAABA9jh0Ri8mJsb+ukiRIqpataosFovLigIAAAAAZJ9DZ/TuuecezZkzR/Xr15evr699fvXq1V1WGAAAAAAgexwKegcOHNCBAwf0/vvv2+dZLBZt3brVZYUBAAAAALLHoaC3bds2V9cBAAAAAHASh4LekiVLMp3fu3dvpxYDAAAAAMg5h4Lenx+OnpKSom+++UaNGzd2WVEAAAAAgOxzKOhNmjQpw3RsbKzGjh3rkoIAAAAAADnj0OMV/qpMmTI6ffq0s2sBAAAAADhBlq/Rs9ls+uGHH+Tv7++yogAAAAAA2Zfla/QkqWzZshoxYoRLCgIAAAAA5IzD1+h98803atCggS5duqS9e/cqKCjI1bUBAAAAALLBoWv0YmJiNHPmTElSUlKS5s+fr9mzZ7u0MAAAAABA9jgU9LZu3arFixdLkoKCgvTOO+9o48aNLi0MAAAAAJA9DgW91NRUeXp62qc9PT1lsVhcVhQAAAAAIPscukavbt26ev755xUeHi6LxaI1a9bovvvuc3VtAAAAAIBscCjojRs3TjNnztSkSZPk4eGhJk2aaNCgQa6uDQAAAACQDQ4FPV9fX7Vu3VqjRo2y33WzSJEirq4NAAAAAJAN3HUTAAAAAAzDXTcBAAAAwDDcdRMAAAAADMNdNwEAAADAMA6d0Rs3bpxKly6tSZMmaerUqfL399fYsWNdXRsAAAAAIBscvuvm6NGjXV0LAAAAAMAJHAp63377rebPn69r167JZrPJarXq1KlT2r59u4vLAwAAAABklUNDN6OiolSnTh0lJCQoNDRUxYoVU9u2bV1dGwAAAAAgGxw6o2exWPTUU0/p4sWLqlixokJDQ9W1a1dX1wYAAAAAyAaHzugVLVpUknT77bfrl19+kY+Pj9zcHNoUAAAAAJDLHDqjV6tWLT377LN65pln1L9/fx0/flweHg5tCgAAAADIZQ6dlhszZowiIyN11113acyYMbJarZo+fbok6fjx466sDwAAAACQRQ5fo1e7dm1JUosWLdSiRQv7smHDhumjjz5yRW0AAAAAgGzI8YV2NpvNGXUAAAAAAJwkx0HPYrE4ow4AAAAAgJNw60wAAAAAMAxBDwAAAAAMQ9ADAAAAAMM4HPQSExMlSSkpKTp//rx9/p133un0ogAAAAAA2edQ0Nu4caO6dOkiSTp9+rTat2+vbdu2SZJiYmJcVx0AAAAAIMscCnpz587VsmXLJEl33XWXPvroI82aNculhQEAAAAAssehB6ZbrVYFBQXZp8uWLSur1eqyogAAOZeSlqqAgOJ5XQYKCPoLAJjFoaB322236b333lN4eLgsFos++ugjlS5d2tW1AQBywMvDU5FLnnFKW0t7z3BKO8i/nNVf6CsAkD84NHTzpZde0qpVq3TfffepVq1aWrVqlcaPH+/i0gAAAAAA2eHQGb2jR49q9erVunz5stzd3VWsWDFX1wUAAAAAyCaHzuj9cWfNkiVLEvIAAAAAIJ9z6IzePffcozlz5qh+/fry9fW1z69evbrLCgMAAAAAZI9DQe/AgQM6cOCA3n//ffs8i8WirVu3uqwwAAAAAED2OBT0/ng4OgAAAAAg/3PoGr3ExES99NJLeuKJJ3Tp0iVFR0crMTHR1bUBAAAAALLBoaD38ssvq3jx4jp//ry8vb2VkJCg6OhoV9cGAAAAAMgGh4LeoUOHNGzYMHl4eKhIkSJ69dVXdejQIVfXBgAAAADIBoeCnptbxtXS09NvmgcAAAAAyB8cuhlLgwYNNG3aNCUlJenLL7/UO++8o4YNG7q6NgCZSElLVUBA8bwuAwAAAPmYQ0HvhRde0Pz582Wz2fTvf/9bDz/8sAYOHOjq2gBkwsvDU5FLnslxO0t7z3BCNQAAAMiPHAp6J0+e1LZt23Ty5ElJ0r59+3T+/HmVK1fOpcUBAAAAALLOoQvtRo8erUceeUQHDhzQd999p+DgYI0dO9bVtQEAAAAAssGhoHf9+nV1795dnp6e8vLyUq9evRQfH+/q2gAAAAAA2eBQ0KtYsaL2799vnz5y5IgqVKjgsqIAAAAAANnn0DV6v//+u3r16qUqVarIw8NDP/30kwICAhQaGipJWr9+vUuLBAAAAAA4zuG7bgIAAAAACgaHgt7999/v6joAAAAAAE7i0DV6AAAAAICCg6AHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYVwa9NavX6+QkBC1bdtWy5cv/9v1tm/frlatWrmyFAAAAAAoNDxc1XBsbKxiYmK0evVqeXl5qXv37mrYsKEqV66cYb34+HhNmTLFVWUAAAAAQKHjsjN6u3btUqNGjeTn5ydfX18FBwdr06ZNN60XFRWlwYMHu6oMAAAAACh0XBb0zp07p4CAAPt0YGCgYmNjM6yzbNky3XvvvbrvvvtcVQYAAAAAFDouG7pptVplsVjs0zabLcP0kSNHtGXLFi1dulRnz57N1j78/YvluE4A5goIKJ7XJaAAob/AUfQVZAX9BY5ydl9xWdALCgrS3r177dNxcXEKDAy0T2/atElxcXHq2rWrUlNTde7cOfXo0UMrVqxweB/nzyfIarU5tW7gz/hwLtji4q7m2r7oKwUf/QWOoq8gK+gvcFRW+4qbm+UfT3y5bOhmkyZNtHv3bl24cEHXr1/Xli1b1KxZM/vyoUOHavPmzVq7dq3mz5+vwMDALIU8AAAAAEDmXBb0ypQpo2HDhikiIkKdO3dWhw4dVKtWLfXr108HDx501W4BAAAAoNBz2dBNSQoNDVVoaGiGeQsWLLhpvQoVKmjbtm2uLAUAAAAACg2XPjAdAAAAAJD7CHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGI+8LiC3FS/hIx9vT6e0lZ6aIndPrxy3k5aSrIuXU5xQEQAAAAAUwqDn4+2pHiOWO6WtFVN7at/Uvjlup96IhZIIegAAAACcg6GbAAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYlwa99evXKyQkRG3bttXy5ctvWv7pp5+qU6dO6tixowYOHKjLly+7shwAAAAAKBRcFvRiY2MVExOjFStWaM2aNVq5cqWOHj1qX56QkKDx48dr/vz5WrdunapUqaJZs2a5qhwAAAAAKDRcFvR27dqlRo0ayc/PT76+vgoODtamTZvsy1NTU/Xiiy+qTJkykqQqVarozJkzrioHAAAAAAoNlwW9c+fOKSAgwD4dGBio2NhY+3SpUqX00EMPSZKSkpI0f/58tWnTxlXlAAAAAECh4eGqhq1WqywWi33aZrNlmP7D1atXNWjQIFWtWlVdunTJ0j78/YvluM78IiCgeF6XABiHf1fICvoLHEVfQVbQX+AoZ/cVlwW9oKAg7d271z4dFxenwMDADOucO3dOffr0UaNGjTRmzJgs7+P8+QRZrbYsbZNf/7HFxV3N6xKQifzaX+CY3Px3RV8p+OgvcBR9BVlBf4GjstpX3Nws/3jiy2VDN5s0aaLdu3frwoULun79urZs2aJmzZrZl6enp2vAgAF6+OGHNXbs2EzP9gEAAAAAss5lZ/TKlCmjYcOGKSIiQqmpqQoPD1etWrXUr18/DR06VGfPntVPP/2k9PR0bd68WZJUo0YNTZw40VUlAQAAAECh4LKgJ0mhoaEKDQ3NMG/BggWSpJo1a+rw4cOu3H2BkZKW6rRT7ckpKbpyOdkpbQEAAAAomFwa9OAYLw9PRS55xiltLe09QxJBDwAAACjMXHaNHgAAAAAgbxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMAxBDwAAAAAM45HXBSB/KlmiiLy8ndM9UpLTdPnKdae0BQAAAODWCHrIlJe3h14Z+4FT2hozMdwp7QAAAABwDEM3AQAAAMAwBD0AAAAAMAxBDwAAAAAMQ9ADAAAAAMMQ9AAAAADAMAQ9AAAAADAMQQ8AAAAADEPQAwAAAADDEPQAAAAAwDAEPQAAAAAwDEEPAAAAAAxD0AMAAAAAw3jkdQEwX1pqqgICijulrZTkZF2+kuKUtgAAAABTEfTgch6ennptdH+ntPXcpHmSCHoAAADAP2HoJgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYgh4AAAAAGIagBwAAAACGIegBAAAAgGEIegAAAABgGIIeAAAAABiGoAcAAAAAhiHoAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIYh6AEAAACAYQh6AAAAAGAYlwa99evXKyQkRG3bttXy5ctvWn7o0CGFhYUpODhYY8eOVVpamivLAQAAAIBCwWVBLzY2VjExMVqxYoXWrFmjlStX6ujRoxnWGT58uKKjo7V582bZbDatWrXKVeUAAAAAQKHh4aqGd+3apUaNGsnPz0+SFBwcrE2bNmnw4MGSpNOnTyspKUm1a9eWJIWFhWnmzJnq0aOHw/twc7Nkq7bSpYpma7vMeJXwd0o7pYvd5pR2pOy/L39V0s/XKe1IUgk/57xPkvOOz1HO6i/O6iuS8/pLkdLOqyk/9peC2lek/PnZ4qz+kh/7ilRw+wufLY7js4XPFkfx2cJnS1bk1WfLrda32Gw2W04K+jvz5s3TtWvXNGzYMEnS+++/r++//17//ve/JUnffvutpk6dqnfffVeSdOLECT311FPavHmzK8oBAAAAgELDZUM3rVarLJb/T5k2my3D9K2WAwAAAACyx2VBLygoSHFxcfbpuLg4BQYG/u3y+Pj4DMsBAAAAANnjsqDXpEkT7d69WxcuXND169e1ZcsWNWvWzL68fPny8vb21r59+yRJa9euzbAcAAAAAJA9LrtGT7rxeIV58+YpNTVV4eHh6tevn/r166ehQ4eqZs2aOnz4sKKiopSQkKDq1atr0qRJ8vLyclU5AAAAAFAouDToAQAAAAByn0sfmA4AAAAAyH0EPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD08olNmzYpLCxMHTt2VGhoqBYuXOi0tmNjY9WvXz+ntYf8JS0tTXPmzNHDDz+skJAQBQcHa+7cueI+S/izU6dOqUaNGurUqVOG/86cOZPtNg8ePKixY8dKknr16qWvvvrKWeUiH0lMTNSECRP00EMPqWPHjurRo4d27959y+1Wr16tUaNG5UKFyE8SEhI0YcIEdejQQZ06dVKvXr30448/6quvvlKvXr3yujzkkT9/B3Xu3Fnt27dX7969dfbsWbVq1UqnTp1yqB36UdZ45HUBuBHEpkyZotWrV6tUqVJKTExUr169dNddd6l169Y5br9MmTJasGCBEypFfjRhwgTFx8dr5cqVKlGihBISEjRo0CAVL15cPXv2zOvykI8EBgZq7dq1TmuvZs2aqlmzptPaQ/5js9k0YMAAVatWTRs2bJCXl5d++uknPfXUU5o+fboaNmyY1yUiH7FarerXr58aNmyoNWvWyMPDQ3v27FG/fv304osv5nV5yGN//Q6aPHmypk6dmocVmY8zevnAxYsXlZqaqqSkJElS0aJFNXnyZFWuXFmtWrXS1KlT1blzZ3Xu3Fk//fSTJOnEiRPq3bu3unTposcee8w+//Tp04qIiFCHDh0UHh6uw4cP69SpU2rVqpUkKT4+XgMHDlRYWJi6du2qXbt2SZJ2796tsLAwhYWFqXfv3rpw4UIevBPIqrNnz2rdunWaPHmySpQoIUkqVqyYoqOjVbp06b/9ec+aNUtRUVHq1auXWrVqpTlz5ki6+S/wf5ylOXv2rB5//HGFhYUpPDxc3333Xa4fK1zjyJEj6tWrl7p27aqWLVvq3XfflXSjj4wePVqPPfaYgoODtWbNGo0cOVLt2rXTs88+K5vNlulfVocPH65Vq1bZp3v16qUDBw7k6jHBeb7++mv9/vvvGj16tP05t/fee6+efvppzZ49W7169dLgwYMVHBysQ4cOac2aNQoODlbXrl21fft2ezvff/+9HnvsMXXp0kVPPvmkTp48KUk3bY+C7auvvtKZM2c0dOhQeXjcOJfQqFEjTZo0Senp6fb1vv76a3t/aN26tT799FNJN56/3KlTJ4WFhWno0KFKTk7m+8dgDRs21C+//CJJevPNN9W5c2cFBwfrwIEDOnHihFq0aCGr1SrpRt/q27dvhu3/PJLkz7/rjho1ShMmTFC3bt308MMP65NPPtHgwYPVpk0bTZ48ORePMO9xRi8fqFq1qlq3bq02bdqoWrVqatiwoUJDQ3XHHXdIknx9fbVmzRpt27ZNI0eO1Pr16zVy5EhFR0fr3nvv1dGjRzVo0CBt3rxZEyZMUHBwsHr27KnPP/9cc+bM0fDhw+37mjhxorp27arWrVvr3Llz6tGjh9asWaPZs2dr/PjxqlWrlhYsWKCffvpJTZs2zau3BA76/vvvValSJZUsWTLD/EqVKqlSpUoaNmxYpj9vSfr555+1fPlyXb16VW3atPnHs38ffPCBWrRoob59++qLL77Qvn37VLt2bRceGVzh3Llz6tSpk306NDRUsbGxGjhwoBo3bqyTJ0+qY8eOeuyxxyTdCIErV67U/v379cQTT2j9+vW68847FRISop9//jnTfXTt2lWzZs3So48+qtOnT+vChQu67777cuX44HwHDx5UjRo1ZLFYMsxv0KCBpk+frurVq6tKlSp64403FBsbq1dffVVr1qyRn5+f+vfvL19fX6WkpCgqKkpz585VuXLl9OWXX2rcuHFaunSpJNm3R8H3008/qWrVqnJzy3geoXnz5hmGdr/zzjt6+eWXValSJe3evVuvvPKK2rRpo9dff12rVq2Sv7+/pkyZomPHjmnr1q18/xgoNTVVmzdvVu3atbVz505VrlxZkyZN0jvvvKNFixZp5syZqlChgr766is1btxYa9asUVhYmMPtnzt3TitXrtRHH32k0aNHa/PmzfL29lazZs3so54KA4JePjFhwgQNHDhQO3bs0I4dO/Too4/q1VdflSQ9+uijkqRWrVpp1KhROnv2rH744QeNHj3avv21a9d08eJFffPNN3rttdck3fhgbd68eYZxz7t27dKxY8c0c+ZMSTeu7zp58qRat25t/2tH69at9cADD+TWoSOH/vwL2KZNmzRnzhxZrVZ5eXnp1KlTmf68pRt/SfPy8pK/v7/8/Px09erVv91H48aNNWTIEB06dEjNmzfX448/7tqDgktkNnQzPT1dX375pebNm6cjR47o2rVr9mUPPPCAPDw8VK5cOQUEBKhy5cqSbgwHv3z5cqb7aNiwocaNG6dTp05p7dq1GYIlCh6LxZLhTMwfUlNT7Z89tWrVkiR9++23qlOnjkqXLi3pxh8S9uzZo+PHj+vkyZN6+umn7dsnJCTYX/+xPQo+Nzc3eXt733K9adOm6bPPPtOmTZt04MABJSYmSpJatmypxx57TG3atFFwcLCqVauma9eu8f1jiD//sTElJUW1atXS888/r507d6pNmzaSpMqVK2vz5s2SbvzhcN26dapdu7b27Nmj8ePHO3xGt1mzZpKkcuXK6e6775a/v78kyc/PT5cvXyboIfds375d165dU0hIiLp27aquXbtq1apV+uCDDyTJPvxBujH+PT09XV5eXhl+YTt79qz8/PwyrGuz2fTrr7/Kx8cnw/ZvvfWW/Pz8JN34R+fv769q1aqpZcuW+uyzzzRt2jR9//33Gb6UkT/VqFFDv/76qxISElSsWDG1a9dO7dq106lTpxQREfG3P+9PP/00w5exxWKRzWaz//8PqampkqR69eppw4YN2r59uzZu3KiPPvpIS5YsydVjhWs8++yzKlGihFq2bKmQkBB9/PHH9mWenp7213/+bPknFotFnTt31oYNG/Sf//xHixYtcnrNyD333Xef3n77baWmpmboD999951q1Kghq9Vq/4756+fHH33GarWqQoUK9u+s9PR0xcfH29f783cUCrYaNWpoxYoV9u+TP7z22mtq0qSJfbpHjx5q2LChGjZsqMaNG+uFF16QJEVFRenw4cP6/PPPNXz4cA0ePFidOnXi+8cQ/3SduLu7u6SMf7xu166dYmJitHnzZjVr1izTPyL88ZmTlpaWYX52vr9MxDV6+YCPj4+mT59uP/Nms9l06NAhVatWTZK0YcMGSdInn3yiSpUqqXz58rrzzjvt/1h27txpH3ZXv359+/q7du3SuHHjMuyrUaNGWrFihSTp6NGjCg0N1fXr1/XII48oMTFRkZGRioyMtF/zh/ytXLly6tixo0aOHKkrV65IuvFht337drm5uf3tz/vvlCpVSr/++qtsNptOnjxpH543depUrVu3Tl26dFF0dDT9wyA7d+7U0KFD1aZNG33xxReSlOkZnKwICwvTe++9p7Jly6pMmTLOKBN5pH79+qpcubJeeeUV+x9+fvjhB82ZM0cDBw7MsG69evX03XffKTY2VlarVRs3bpQkVaxYUZcvX9bevXslSR9++KH9F3uYpX79+vL399cbb7xh/xz58ssvtXr1avu1/5cuXdLx48f1zDPPqFmzZtq6davS09OVlpamtm3bqlSpUurfv786deqkQ4cO8f1TiBUpUkTNmjXTa6+9lumwzVKlSuno0aOSZL/OExkV3oibjzRq1EiDBw/WgAED7F+kDz74oAYNGqT169dr//79+uCDD1SkSBH7RaTTpk3T+PHjtXDhQnl6eiomJkYWi0XR0dGKiorSihUrVKRIEb388ssZ9hUVFaXo6GiFhoZKuvELfLFixfTcc89p1KhR8vDwkK+v703bIf8aP368lixZooiICKWnpysxMVENGzbUggUL5Ovrm+nP++80adJEH374odq1a6e77rpL9erVk3Tjgufnn39eq1evlru7u6ZMmZIrxwbXGzJkiHr06CFvb29VrVpV5cuXd/g213+nbNmyKlu2rLp06eKkKpGX3njjDcXExKhDhw5yd3dXyZIlNW3aNDVs2DDDtXWlS5dWVFSUIiMjVaRIEftQXy8vL82YMUMTJ05UcnKyihUrxmeIoSwWi2bPnq1JkyapQ4cO8vDwUKlSpTR//nz75QF+fn4KDw9X+/bt5eHhoUaNGikpKUkpKSkaOnSonnzySXl7e8vf31+TJ09WSkoK3z+FWPv27bV///5Mr/Xu27evRo0apQ8//NApd6k3kcXGw7bytVatWmnZsmWqUKFCXpcCALdks9l07tw59erVSx9//LH9To0AAGRFenq6YmJi5O/vr969e+d1OQUSQzcBAE6zefNmderUSc899xwhDwCQbV27dtWPP/5ovxM0so4zegAAAABgGM7oAQAAAIBhCHoAAAAAYBiCHgAAAAAYhqAHAAAAAIbhOXoAAOO9/PLL+uabbyRJv/76q8qXLy8fHx9J0sqVK+2vXWHWrFm6ePGioqOj/3adVq1aacaMGapZs6bL6gAAFC4EPQCA8aKiouyvW7VqpVdffZVQBQAwGkM3AQCF0rVr1zRixAh169ZNwcHBCgsL07Fjx5SUlKT27dtr+fLlkqT3339foaGhun79ur777jv17NlTjzzyiFq0aKExY8Y4vL8dO3YoNDTUPn3lyhU1aNBAly9fdvqxAQBA0AMAFEpffPGFSpQooZUrV2rz5s2qUaOGli9fLh8fH7322muaOXOmPv/8c73++uuaMWOGihQpomXLlmno0KF6//33tWHDBm3btk0//PCDQ/t74IEHlJiYqIMHD0qSPv74YzVv3lwlS5Z05WECAAophm4CAAqldu3a6V//+pfefvttnThxQl9//bXq1KkjSapSpYoGDx6s/v37a/LkyapYsaIkafLkyfriiy80d+5cHTt2TMnJybp27ZpD+7NYLOratas++ugj1axZU6tXr9aIESNcdnwAgMKNM3oAgEJpxYoVGjt2rHx8fBQaGqoOHTrIZrPZl//yyy8qXbq0Dhw4YJ/3+OOP6/PPP1fFihU1aNAgBQYGZtjmVsLDw7Vp0yYdOnRIV69e1f333+/UYwIA4A8EPQBAobRjxw516dJFjzzyiO666y5t27ZN6enpkqQtW7boq6++0rp167Rz5059+umnunLlig4ePKgXXnhBbdu21dmzZ/Xbb7/JarU6vM8yZcqoVq1aio6OVnh4uKsODQAAgh4AoHB68skntXLlSoWGhqpnz56qXr26fvvtN505c0Yvvviipk6dqttuu02TJ0/WuHHjdO3aNT311FPq0qWLOnTooPnz56tu3bo6ceJElvb7yCOP6NChQ+rSpYuLjgwAAMliy8qYEwAAAABAvsfNWAAAyIGEhAT17Nkz02VFixbVihUrcrkiAAA4owcAAAAAxuEaPQAAAAAwDEEPAAAAAAxD0AMAAAAAwxD0AAAAAMAwBD0AAAAAMMz/AWopoCKdhF2+AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1080x720 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set(rc={'figure.figsize':(15,10)})\n",
    "g = sns.barplot(data=results,\n",
    "                x='Taxa_lvl', y='perc_accuracy', hue='Tool',\n",
    "                ci = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA4MAAAJPCAYAAADGygGBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAA8u0lEQVR4nO3dd5xUhb3///fCUiIotkW8xHjtScRewAYJREEBIVijQbFrLNEbUaKo2FFRYwleErERjRJRFEIwGoyaiy3qvUZFY1cQcUGQJsjuzu8Pv+4viBpUhkXO8/l4+IA5c+bMZ3bGZV97zpmpKJVKpQAAAFAojRp6AAAAAJY/MQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKqLKhByi3mTPnpa7ORykCAADF0qhRRdZYo8XnXr/Sx2BdXUkMAgAAfIrDRAEAAApIDAIAABTQSn+YKAAAFFVtbU1mzqxOTc1HDT0KZVZZ2TRrrFGVxo2XPvHEIAAArKRmzqxO8+arpEWLNqmoqGjocSiTUqmUefNmZ+bM6qy99rpLfTuHiQIAwEqqpuajtGixmhBcyVVUVKRFi9W+9B5gMQgAACsxIVgMX+V5FoMAAAAF5JxBAACgQU2d+k4OOKB3Ntxw4yRJqVSXb31rley330/SpcvuX3jbG2/8bTbeeJPsttsPlvlcp5xyfM4558Ksvvrqy3zbKwIxCAAANLhmzZrlpptuq7/87rtT8/OfH5fGjRvlBz/o8rm3e+qpJ/Of/7lBWWZ68snHy7LdFYUYBAAAVjht2qybI444NrfdNiIbbrhxrrjiksyfPz8zZkzPJptsmnPPvThjx96Tl16alKFDr06jRo2zwQYbfuZ6zZo1y/Dhw/Lwww+msrJJWrVqlTPOGJS11147b7zxeq66akg++OCD1NXVZd99D0iPHr1y0UXnJklOOumYXHbZVVlnnTYN/BVZ9sQgAACwQtp4403y2muvZMyY0dlzzx7p2nWv1NTU5IgjfppHH/1b9tln/zz44APZZ5/906nTD/PrX1/1met973ubZ+TI2zJmzP1p2rRpfv/73+WFF57LzjvvmoEDT89ZZ52XzTb7bubOnZtjjz0s//mfG+aMM87JuHFjcvXVwxwmCgAAsDxVVFSkefPmOe64E/Pkk4/n1ltvzttvv5Xp06vz4YcfLrH+561XVdU6G2+8aQ4//Kfp0GHndOiwc7bffse8/vpreeedybn44vPqt7Fw4cK8/PJLaddui+X5UBuEGAQAAFZIL774QjbccOMMGnRmamtr0rnz7tlpp10zbdq7KZVKS6z/ees1atQo1177m7z44gv5+9+fyDXXXJH27XdK167d06JFy8XOVXz//Rlp0aLl8nyYDcZHSwAAACuct956MzfddH0OPPCneeKJR9Ov31Hp0mWPJMkLLzyXurraJEnjxpWpqalJks9d7+WX/5m+fQ/I+utvkL59D8v++x+USZNeyHe+s36aNWuW++4blySZNu3dHHLIAXnppUn/b9uN67e9MrJnEAAAaHALFy5Mv34HJUkaNapI06bNcswxJ2TnnXfN0UcfnzPO6J9vfat5WrRoma233jaTJ09Okuy6a8cMG/br1NTUfO56PXr0TufOP8qRR/bNt761Spo1a5aTTz41TZo0ycUXX56rrhqS2267JTU1NTnyyGOz5ZZbJ0l+8IMuOeGEo3PRRZfWf+zFyqSi9Fn7V1ciM2bMTV3dSv0QAQDgM7377ptp02b9hh6D5eTTz3ejRhVZa63PP+TVYaIAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggHzOIAAAFMSqqzVP82ZNlvl2FyxclDmzFyzz7VJeYhAAgBXWGq2aprJps4Yeo2xqPlqYmR98tNzur3mzJjnotFuX+XZvu/TgzMnSxeCDDz6QESNuSm1tbUqlunTr1j0HHXRIhg8flu233zFbbbXNl77/4cOHJUmOOOKYJa4bNWpkxowZnVKplIqKihxwwEHZc88eSZKLLjo3hx9+dNq0WXep7mfq1Hdy4onH5M47x3zpGVdEYhAAgBVWZdNmeerSIxt6jLLZ7rTrkyy/GGxo1dXv5dprf5UbbvhdWrVaPfPnz88JJxyd73xn/TzzzFPZZpvtlun9Pf/8cxk7dnSGDbshzZo1z8yZ7+eII/pm4403zSabbJqnn/57DjvsqGV6n98kzhkEAACWi1mzZqWmpiYLFny8F3GVVVbJwIGD8vLL/8xLL03KJZdckFdffSXPPPNUjjvuiBx++MHZb79eeeSRvyZJLrxwUIYMuThHHNE3Bx7YJ+PH/7F+25MmPZ9jjz08++zTo35P4fvvT0+plPr7W2ONNXPBBZdkjTXWyIgRN2X69Or07//zfPDBrOy7b89MnfpOkuTpp/+eE044Oknyz3++mMMPPziHH35wbrzxt0mSefPmpnv3Lpk3b26Sj/cY/vSn+5X7y7fMiUEAAGC52GSTTbPbbp2y//69ctRRh2To0KtTW1uXww47Kptt9r2cfvrAbLTRxhk16o4MGHBWbrjh1gwYMDC//e119duYMmVyhg27MVdffV1+/eurMmPG9CTJ+++/n6uv/u8MH/67/P73IzJ//rx06LBL1l133fTq1S0nnHB0hg8fllatVs/aa1elb99+WXvtqlx22VVp1Wr1z535ggvOyXHHnZgbbrg1//EfbZMkLVq0zE477ZoHH/xLkmT8+D+mW7fu5fvClYkYBAAAlptTT/1l7rxzTHr33jfTpk3NMccclocemrDYOmeddX5ee+2V3HTT9bn99t/lww8/rL9ur716prKyMq1br5Mtttgqzz77v0mSDh12TtOmTbP66qunVavVM3v27DRp0iQXX3x5fve7P6Rz593z0ksv5tBDD8xzz/1jqWadNWtWpk+fnh126JAk9ecaJkn37nvnvvvGJUnuv3/8NzIGnTMIAAAsFxMn/i0ffjg/Xbrske7d90737nvn3nvvztix9yy23vHHH5Vtt90u22yzXbbbboece+7A+usaN/7/E6ZUqqu/3Lhx4/rlFRUVKZVK+dOfxqaqqnW2337HfPvb66VPn/0ybNivc99949Ku3RaL3ecnt0mS2tqa/7cs9cs+fd9bb71tqqur89BDE7Luum2z9tpVX/fLs9zZMwgAACwXzZs3z3//96/rz80rlUp5+eV/ZpNNNkvjxpWpra3N7Nkf5O2338wRRxybDh12ySOPPJS6urr6bUyYcH9KpVLefXdqXnjhuWy11dafe391dXUZNuzazJo1K0myaNGivPHGa9l0082SfByQtbW1SZJWrVbP66+/liR55JGH6pe1adMmEyf+LcnHewA/UVFRkT337J5f/WpI9trr/99j+E1izyAAABTEgoWLctulB5dlu0tj2223z+GHH5XTTjs5NTUf731r336n9Ot3ZO68844MGXJxBg48Nz169ErfvvunsrIy2267QxYsWFB/qOjChQtyxBF9s2jRR+nf/8wvPN+ve/e988EHs3LccYenUaOP94N16bJHevTolSTZeefdcuqpP88VV1yTI444OldeeVluvPG32XHHDvXbOOus83Pxxefmt78dms0333Kx7f/oR13z+9//Lrvt9oOl/VKtUCpK/7rfcyU0Y8bc1NWt1A8RAGClVVW16kr/0RLV1XPKtv13330zbdqsX7btL28XXjgo22yzXfbaq2dDj5K6urqMHj0qb731Rk4+uX9Dj5Nkyee7UaOKrLVWy89d355BAACAL+nMM/tn2rR3c/nl1zb0KF+ZGAQAAL4RzjxzUEOPUO/iiy9v6BG+Nm8gAwAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAUBBrtGqaqqpVl/l/a7RqulT3P3XqO9l11+3z5JOPLbZ833171n8Q/WfdZt99l81HSdx7792LfXD8c8/9Iz//+XE59NCfpG/f/TNkyOAsXLjgC7cxfPiwDB8+LEmy667bf6U5hg8flv/7v2eSJIMHn58XX3zhK23n6/JuogAAUBCVTZuV5XMbtzvt+iQfLd0MlZW55JILc8stt2eVVVos81m+yD/+8X/ZZpvtkiSvvPJyzjzz1Fx00eXZfPN2qampya9+dVkuvfTCnHXW+WWd45lnnqqfY8CAs8p6X19EDAIAAMvN2mtXZYcd2ueaa36V008/s355bW1tLrnkgrz22qt5//33s/HGG2fQoAuTJAsXLsw55/wyr732alZddbVcfPGQtGq1ekaNuiPjx4/LggUfpkmTJhk06MJ85zv/mX337ZmuXffKE088mg8/XJCBA8/NnDmz87e/PZynnnoya621dv7853Hp3r1XNt+8XZKPI/W4407Mk08+niR5//0ZGTz4/Eyb9m4aN26co48+Ph067PyZj2n+/Pm54opL8tprr6auri4HH3xIdt+9WxYuXJgrrrgkzz77v6msrEy/fkfmo48+yksvTcoll1yQiy4akiuvvDSHH350tt12+9xyyw3585//lEaNGmWHHTrkZz87Ke+9Ny1nnHFqNtxwo/zzny9lzTXXyvnnD85qq7X62s+Fw0QBAIDl6oQTTs4TTzy62OGi06dXp7KySYYNuzF33HF35syZk0cf/Z8kyaxZM3PAAQdnxIiRWXPNNfPAA3/OvHlz8/DDD+Xaa4dlxIiR2Xnn3TJq1Mj67bVq1Sq//e0t6d27T0aMuCE77NA+u+7aMUceeWzat98p//znS9lss+8uNleLFi3zgx90SZJceeVl2Xbb7XPzzbfn/PMvycUXn5f335/xmY/n5puHZ7PNvpcbbvhdfv3r3+SWW27IlCmTM2rUHfnwww9z66135le/Gpobb7w+P/pR12y22fdy+ukDs9FGG9dv49FH/yd/+9vDuf76EbnhhlszZcrbGT16VJKP92J+8vhbtmyZP//5T8vkebBnEAAAWK5atGiZ008fWH+4aJKss06b/PjH+2bUqJF56603Mnny2/nwww+TfLw38fvf/3gP3gYbbJQPPpiVFi1aZtCgC/LAA3/O22+/lccfn5hNNtms/j7at/94L96GG26chx56cIkZGjVqlKZNP/9cx6effjKnnz4wSdK27bfz/e+3ywsvPPeZ6/79709k4cIF+eMf702SLFiwIK+//lr+93+fzt57/ziNGjXKWmutnd/9buRn3j5JnnrqyfzoR13TvHnzJEn37nvnT3/6Y3beedesscaa2XTT79Y/ntmzZ3/udr4MewYBAIDlbscdO9QfLpokf/vbwznvvLPSvHnz7LXX3tlqq21SKpWSJI0bN17stqVSKdOmvZtjjjksc+fOSYcOO2fPPXvWr59ksdD71+Wf2Gyz7+XFFycttmzevLk57bRTsmjRotTVffo2pdTW1n7mY6mrq81ZZ52fm266LTfddFuGDbsxHTrsnMaNK5NU1K83efLbWbRo0Wduo1Sq+9TlpLa2ZonH8nmP56sQgwAAQIP45HDRGTOm5+9/fzydO/8o3bvvnZYtW+aZZ55KXd1nx1eSvPjiC/n2t9fLAQccnO997/t5+OEHv3D95OOo/CToDjjg4Nx99531e/tqampy7bW/SsuWLdOkSZNst932GTt2dJJkypTJ+cc//i+bb77lZ2532213yOjRdyZJpk+fnkMP/UmmTXs3W2+9TSZMuD+lUikzZ76fE044OosWfZTGjSuXCMttt90hDzxwXxYuXJCampqMG3dvtt32q71b6dJymCgAANAgPjlc9L/+64QcffTxOffcM/PAA/elsrJJtthiy7zzzjvZbrvPvu0OO3TI3XffmZ/+dL+USqVsvfW2ee21V7/w/rbffscMGzY0LVu2zA9/+KOcddZ5ueqqy7NgwYLU1CzK9tvvmJNPPjVJcvLJ/XPppRdm3LgxqaioyOmnD8zaa6/9mds9/PCjcvnll6Rv3/1TV1eXn/3spLRt++38+Mf75Ve/uiz9+v0kSXLKKf2zyiot0r79Thky5OIMHHhu/TZ22WW3vPzySzniiENSW1uTHXfskH32OSDV1e99ha/s0qkoLat9jCuoGTPmfsYuXgAAvgmqqlYty0chrCi2O+36VFfPKdv23333zbRps3795TVaNU1l02bL/H5qPlqYmR8s3UdLUD6ffr4bNarIWmu1/Nz17RkEAICC+DjYRBsfc84gAABAAYlBAACAAhKDAACwElvJ3yKE/+erPM9iEAAAVlKVlU0zb95sQbiSK5VKmTdvdiorm/77lf+FN5ABAICV1BprVGXmzOrMnTuroUehzCorm2aNNaq+3G3KNAsAANDAGjeuzNprr9vQY7CCcpgoAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFVPYYvOSSSzJgwIAkycSJE9OzZ8/sscceufLKK+vXmTRpUvr06ZOuXbvmzDPPTE1NTZLknXfeycEHH5xu3brluOOOy7x588o9LgAAQCGUNQYfffTR3H333UmSBQsW5IwzzsjQoUMzbty4PPfcc3nooYeSJP3798/ZZ5+d++67L6VSKSNHjkySnHvuuTnooIMyfvz4tGvXLkOHDi3nuAAAAIVRthicNWtWrrzyyhx77LFJkmeffTbrr79+1ltvvVRWVqZnz54ZP358pkyZkgULFmTrrbdOkvTp0yfjx4/PokWL8uSTT6Zr166LLQcAAODrqyzXhs8+++yccsopmTp1apLkvffeS1VVVf31rVu3zrRp05ZYXlVVlWnTpmXmzJlp2bJlKisrF1v+Za21Vsuv+UgAAKB8qqpWbegRKKiyxOAf/vCHrLvuutlpp51y1113JUnq6upSUVFRv06pVEpFRcXnLv/kz3/16ctLY8aMuamrK33FRwIAQEMqQihVV89p6BFYSTVqVPGFO8fKEoPjxo1LdXV1evXqlQ8++CDz58/PlClT0rhx4/p1qqur07p167Rp0ybV1dX1y6dPn57WrVtnzTXXzJw5c1JbW5vGjRvXrw8AAMDXV5ZzBm+88caMHTs299xzT0466aR07tw5119/fV5//fW8+eabqa2tzdixY9OxY8e0bds2zZo1y1NPPZUkueeee9KxY8c0adIk22+/fcaNG5ckGT16dDp27FiOcQEAAAqnbOcMflqzZs0yePDgnHjiiVm4cGE6deqUbt26JUmGDBmSgQMHZu7cudl8881zyCGHJEnOOeecDBgwINddd13WXXfdXHHFFctrXAAAgJVaRalUWqlPqHPOIADAN1dV1ap56tIjG3qMstnutOudM0jZ/LtzBsv+ofMAAACseMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAIqawxeddVV2WuvvdK9e/fceOONSZKJEyemZ8+e2WOPPXLllVfWrztp0qT06dMnXbt2zZlnnpmampokyTvvvJODDz443bp1y3HHHZd58+aVc2QAAIBCKFsMPvHEE3nsscdy7733ZtSoURkxYkRefPHFnHHGGRk6dGjGjRuX5557Lg899FCSpH///jn77LNz3333pVQqZeTIkUmSc889NwcddFDGjx+fdu3aZejQoeUaGQAAoDDKFoM77rhjbrnlllRWVmbGjBmpra3N7Nmzs/7662e99dZLZWVlevbsmfHjx2fKlClZsGBBtt566yRJnz59Mn78+CxatChPPvlkunbtuthyAAAAvp7Kcm68SZMmufrqq3PDDTekW7duee+991JVVVV/fevWrTNt2rQllldVVWXatGmZOXNmWrZsmcrKysWWfxlrrdVy2TwYAAAog6qqVRt6BAqqrDGYJCeddFKOOuqoHHvssXnjjTdSUVFRf12pVEpFRUXq6uo+c/knf/6rT1/+d2bMmJu6utLXexAAADSIIoRSdfWchh6BlVSjRhVfuHOsbIeJvvrqq5k0aVKS5Fvf+lb22GOPPP7446murq5fp7q6Oq1bt06bNm0WWz59+vS0bt06a665ZubMmZPa2trF1gcAAODrKVsMTp48OQMHDsxHH32Ujz76KH/5y19y4IEH5vXXX8+bb76Z2trajB07Nh07dkzbtm3TrFmzPPXUU0mSe+65Jx07dkyTJk2y/fbbZ9y4cUmS0aNHp2PHjuUaGQAAoDDKdphop06d8uyzz6Z3795p3Lhx9thjj3Tv3j1rrrlmTjzxxCxcuDCdOnVKt27dkiRDhgzJwIEDM3fu3Gy++eY55JBDkiTnnHNOBgwYkOuuuy7rrrturrjiinKNDAAAUBgVpVJppT6hzjmDAADfXFVVq+apS49s6DHKZrvTrnfOIGXTYOcMAgAAsOISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACigpYrBadOmLbHslVdeWebDAAAAsHx8YQzOmjUrs2bNylFHHZUPPvig/vL06dNzwgknLK8ZAQAAWMYqv+jKX/ziF/mf//mfJEn79u3rlzdu3DjdunUr72QAAACUzRfG4PDhw5Mkv/zlL3PxxRcvl4EAAAAov6U6Z/DnP/95Bg0alCR57bXX8rOf/SzTp08v51wAAACU0VLF4IABA7LhhhsmSdq2bZsdd9wxv/zlL8s6GAAAAOWzVDE4c+bMHHLIIUmSZs2apV+/fqmuri7rYAAAAJTPUsVgbW3tYh8vMX369JRKpbINBQAAQHl94RvIfKJfv37p3bt3dtttt1RUVGTixIk57bTTyj0bAAAAZbJUMbjvvvumXbt2eeyxx9K4ceMcccQR2XTTTcs9GwAAAGWyVIeJJslHH32U5s2b54ADDsi8efPKORMAAABltlQxeNddd+WXv/xlrr/++syZMyc/+9nPMnLkyHLPBgAAQJksVQyOGDEid9xxR1q2bJm11lord911V26++eZyzwYAAECZLFUMNmrUKC1btqy/vO6666Zx48ZlGwoAAIDyWqoYXH311TNp0qRUVFQkSe699960atWqrIMBAABQPkv1bqJnnHFGfv7zn+ett97KrrvummbNmmXo0KHlng0AAIAyWaoYXLBgQe6555688cYbqa2tzQYbbJAmTZqUezYAAADKZKkOEz311FPTuHHjbLTRRtl0002FIAAAwDfcUsXgZpttljFjxuSdd97JrFmz6v8DAADgm2mpDhP9y1/+kvHjxy+2rKKiIpMmTSrLUAAAAJTXUsXg73//+7Rr167cswAAALCcLNVhov379y/3HAAAACxHzhkEAAAoIOcMAgAAFNBSxeA//vGPcs8BAADAcrRUMVhXV5fhw4fn4YcfTk1NTXbZZZcce+yxqaxcqpsDAACwglmqcwYvv/zyPPbYYzn00ENz2GGH5Zlnnskll1xS7tkAAAAok6XatffII49k1KhRadKkSZLkBz/4Qfbee++yDgYAAED5LNWewVKpVB+CSdK0adPFLgMAAPDNslQx+N3vfjcXXXRR3nrrrbz99tu56KKLsummm5Z7NgAAAMpkqWLwnHPOyezZs3PggQdmv/32y8yZM3PWWWeVezYAAADK5Atj8KOPPsrpp5+eRx99NIMHD87EiROz5ZZbpnHjxmnZsuXymhEAAIBl7Atj8Oqrr87cuXOz7bbb1i87//zzM3v27FxzzTVlHw4AAIDy+MIY/Otf/5rLL788a621Vv2yddZZJ5deemkeeOCBsg8HAABAeXxhDDZp0iTNmzdfYnnLli3TtGnTsg0FAABAeX1hDDZq1Chz585dYvncuXNTU1NTtqEAAAAory+MwR49emTgwIGZP39+/bL58+dn4MCB2WOPPco+HAAAAOXxhTF46KGHZtVVV80uu+yS/fffP/vuu2922WWXrLbaajn++OOX14wAAAAsY5VfdGWjRo1y/vnn59hjj83zzz+fRo0aZcstt0zr1q2X13wAAACUwRfG4Cfatm2btm3blnsWAAAAlpMvPEwUAACAlZMYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACKmsMXnvttenevXu6d++eSy+9NEkyceLE9OzZM3vssUeuvPLK+nUnTZqUPn36pGvXrjnzzDNTU1OTJHnnnXdy8MEHp1u3bjnuuOMyb968co4MAABQCGWLwYkTJ+Zvf/tb7r777owePTrPP/98xo4dmzPOOCNDhw7NuHHj8txzz+Whhx5KkvTv3z9nn3127rvvvpRKpYwcOTJJcu655+aggw7K+PHj065duwwdOrRcIwMAABRG2WKwqqoqAwYMSNOmTdOkSZNstNFGeeONN7L++utnvfXWS2VlZXr27Jnx48dnypQpWbBgQbbeeuskSZ8+fTJ+/PgsWrQoTz75ZLp27brYcgAAAL6eynJteJNNNqn/+xtvvJE//elP+elPf5qqqqr65a1bt860adPy3nvvLba8qqoq06ZNy8yZM9OyZctUVlYutvzLWGutll/zkQAAQPlUVa3a0CNQUGWLwU+8/PLLOeaYY3LaaaelcePGeeONN+qvK5VKqaioSF1dXSoqKpZY/smf/+rTl/+dGTPmpq6u9LUeAwAADaMIoVRdPaehR2Al1ahRxRfuHCvrG8g89dRT6devX37xi1/kxz/+cdq0aZPq6ur666urq9O6desllk+fPj2tW7fOmmuumTlz5qS2tnax9QEAAPh6yhaDU6dOzfHHH58hQ4ake/fuSZKtttoqr7/+et58883U1tZm7Nix6dixY9q2bZtmzZrlqaeeSpLcc8896dixY5o0aZLtt98+48aNS5KMHj06HTt2LNfIAAAAhVG2w0SHDx+ehQsXZvDgwfXLDjzwwAwePDgnnnhiFi5cmE6dOqVbt25JkiFDhmTgwIGZO3duNt988xxyyCFJknPOOScDBgzIddddl3XXXTdXXHFFuUYGAAAojIpSqbRSn1DnnEEAgG+uqqpV89SlRzb0GGWz3WnXO2eQsmnQcwYBAABYMYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAVU1hicO3duevTokcmTJydJJk6cmJ49e2aPPfbIlVdeWb/epEmT0qdPn3Tt2jVnnnlmampqkiTvvPNODj744HTr1i3HHXdc5s2bV85xAQAACqNsMfh///d/+clPfpI33ngjSbJgwYKcccYZGTp0aMaNG5fnnnsuDz30UJKkf//+Ofvss3PfffelVCpl5MiRSZJzzz03Bx10UMaPH5927dpl6NCh5RoXAACgUMoWgyNHjsw555yT1q1bJ0meffbZrL/++llvvfVSWVmZnj17Zvz48ZkyZUoWLFiQrbfeOknSp0+fjB8/PosWLcqTTz6Zrl27LrYcAACAr6+yXBu+8MILF7v83nvvpaqqqv5y69atM23atCWWV1VVZdq0aZk5c2ZatmyZysrKxZZ/WWut1fIrPgIAACi/qqpVG3oECqpsMfhpdXV1qaioqL9cKpVSUVHxucs/+fNfffry0pgxY27q6kpffXAAABpMEUKpunpOQ4/ASqpRo4ov3Dm23N5NtE2bNqmurq6/XF1dndatWy+xfPr06WndunXWXHPNzJkzJ7W1tYutDwAAwNe33GJwq622yuuvv54333wztbW1GTt2bDp27Ji2bdumWbNmeeqpp5Ik99xzTzp27JgmTZpk++23z7hx45Iko0ePTseOHZfXuAAAACu15XaYaLNmzTJ48OCceOKJWbhwYTp16pRu3bolSYYMGZKBAwdm7ty52XzzzXPIIYckSc4555wMGDAg1113XdZdd91cccUVy2tcAACAlVpFqVRaqU+oc84gAMA3V1XVqnnq0iMbeoyy2e60650zSNmsMOcMAgAAsOIQgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUUGVDDwAAwFe36mrN07xZk4YeA/gGEoMAAN9gzZs1yUGn3drQY5TNbZce3NAjwErLYaIAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQGIQAACggMQgAABAAYlBAACAAhKDAAAABSQGAQAACkgMAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAApIDAIAABSQGAQAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAFJAYBAAAKSAwCAAAUkBgEAAAoIDEIAABQQJUNPQDL3xqtmqayabOGHqNsaj5amJkffNTQYwAAwApNDBZQZdNmeerSIxt6jLLZ7rTrk4hBAAD4Ig4TBQAAKCB7BgFYJhyCztLyWgHKwfeWL08MArBMOASdpeW1ApSD7y1fnsNEAQAACkgMAgAAFJDDRIHP5dh7AICVlxgEPpdj7wEAVl4OEwUAACggMQgAAFBAYhAAAKCAxCAAAEABiUEAAIACEoMAAAAF5KMlPsOqqzVP82ZNGnoMviG8XoBy8L0FgHL7RsTgmDFjct1116WmpiaHHnpoDj744LLeX/NmTXLQabeW9T4a0m2XlvfrVzQr8+vFawUaju8tQDn4RRP/aoWPwWnTpuXKK6/MXXfdlaZNm+bAAw9M+/bts/HGGzf0aABfin+AAWhoftHEv1rhY3DixInp0KFDVl999SRJ165dM378+JxwwglLdftGjSq+0v2uvUaLr3S7b4qmq63V0COU1Vd93r+qlfn14rWy7DRv1iQnXTx6ud3f8nb1L3t7vSxjvrd8c3mtLFteL8vWyvx68Vr5cutXlEql0tcZqNyGDRuW+fPn55RTTkmS/OEPf8izzz6b888/v4EnAwAA+OZa4d9NtK6uLhUV/3/RlkqlxS4DAADw5a3wMdimTZtUV1fXX66urk7r1q0bcCIAAIBvvhU+Bnfeeec8+uijef/99/Phhx/mz3/+czp27NjQYwEAAHyjrfBvILPOOuvklFNOySGHHJJFixZl3333zZZbbtnQYwEAAHyjrfBvIAMAAMCyt8IfJgoAAMCyJwYBAAAKSAwCAAAUkBgEAAAoIDH4DfCTn/wkf/zjHxdbNn/+/LRv3z49evTI448/vth1AwYMyF133ZUk6dy5c7p27brY9TU1NenQoUMGDBjwleZ59tlnc9lll32l29Kw/vW18Ylrrrkm11xzTZKkb9++X/h62myzzT5zu59efu655+anP/1p5s2bt6xGZxmaO3duzj333PTo0SO9evVK37598/zzz2fy5Mnp3LnzEut/8vw+/vjj2WabbdKrV6/F/rv//vvr1/tk2d57750f/vCHOfvss1NbW/tvrz/mmGNyySWXLHa/t99+e/bff//627NimDx5ctq1a1f/XPbs2TOdO3fO1VdfXb/OP//5z2y22Wa57777Frtt3759s/vuuy/2+rn11lvrr+/Tp0+OPfbYxW5z1113LfHv1eOPP56+ffuW4dFRLv/4xz9y5plnZsCAAfnBD35Q//zvueee+d3vfpfks5/Xz/u+9K/+9XtT7969061bt5x00kmZO3du2R5Pka2Iz2Xnzp0zefLkr/fAkrJ9X/ms72MrihX+oyVI9tlnn4wZMybdu3evX/bnP/857du3z8yZM//t7RcsWJCXXnqp/ge6Rx99NBUVFV95nldeeSUzZsz4yrdn5XbBBRfktddey29/+9t861vfauhx+JS6urocddRRad++fUaPHp3Kyso89thjOeqoo/Kb3/zm396+Xbt2GTFixOdef88999T/fe7cuenRo0f+9re/pVOnTl94/XnnnZe99947PXv2zPe///1MmzYt1157bW655ZY0btz4azxiyqF169aLPZfTpk1L165d071792y00UYZNWpUunXrljvuuGOJX0hecMEFad++/RLbfPHFF9O0adO8+OKLmTp1atZdd92yPw6Wny222CJbbLFFBgwYkJNOOil9+vRJkkyfPj277757dtppp6+1/U9/bzrppJMybNiw/OIXv/ha22VJK/Nz+cQTT5T9PlY09gx+A+y55555+umnM2vWrPpl9957b/bZZ5+luv0ee+yx2G9nx40bt9g/zn379s2ll16aAw44ILvvvnseeuihJB//T/2zn/0sffr0yT777JOJEydm9uzZufrqqzNhwoRcd911y+YBUhYnnHDCYs97nz598sEHH5T1PgcPHpzXXnstw4YNE4IrqMcffzxTp07NSSedlMrKj38f2KFDh1x88cWpq6tbpvc1c+bMfPjhh1l99dX/7fXrrLNOTj311AwcODB1dXW54IILctRRR2XDDTdcpjNRHtXV1SmVSmnRokUWLVqUMWPG5OSTT87zzz+ft956a6m2cdddd2WXXXZJly5dMnLkyDJPzPL2eXtz11577WywwQZ55ZVX/u02Pvzww/ziF79Ijx490rNnz4wePfpz191xxx3z8ssvf52R+Rwr6nN57bXXpnfv3jnggAPy4osvJvnsn2WTj3eM9OnTJ3369Mlhhx2W999/PxdccEGSZL/99kuS7LTTTjn77LPTs2fPHHjggfV7Hjt37pyTTz45Xbt2zYwZMzJ69Oj8+Mc/Tq9evXLGGWdk4cKFSZLRo0ena9eu2WefffLXv/71387fUOwZ/AZo0aJFunTpkvHjx+fAAw/MtGnT8vrrr2fXXXfN9ddfn4EDB2aVVVapX3/q1KnZcccd6y9369YtgwYNykknnZSPPvooL774Yvr27bvYbz8WLVqUO+64IxMmTMhVV12VTp065cILL8w+++yTLl265L333stBBx2U0aNH56STTsoTTzyR4447brl+HfhyevXqlTFjxqRr16554403snDhwrRq1SpXX311br755vr1pk+fngMPPPBr399ll12WG2+8MTfffHOaN2/+tbdHebzwwgv57ne/m0aNFv9dYKdOnTJ58uS899576dWr1+fe/rnnnlvi+ptuuilrrLFGko9fdzU1NZkxY0Y22mijDBw4MFtttVX9ul90/X777Zc//elP6d+/f2bNmpVDDjlkWT1slrFPXicLFy7MzJkzs8UWW+Taa69NmzZt8sADD+Q//uM/ssEGG+RHP/pR7rjjjvTv37/+tv/6b1aLFi1y22231QfkiBEjMmvWrJxyyik5/vjj639hMWHChMVed/Pnz0+bNm2W74OmLF588cW89dZb2XzzzTNlypQlvscsWrSo/u/XXHNN1lhjjYwdOzbvv/9+9ttvv3z3u99dYpvz58/PhAkTFvtZiPJr6Ody/fXXz+DBg/PQQw9lwIABGT169Of+LDt06NAMGjQoW265ZX7729/mhRdeyMCBAzNixIj84Q9/SJK8//772WabbXLeeedlxIgRueCCC/Lf//3fSZKOHTvmV7/6VV5++eWMHDkyt99+e5o1a5bLL788w4cPzz777JMhQ4Zk9OjRWX311XPMMccs9rP6ikQMfkP06dMnV111VQ488MCMGTMme++9d/2hU58+5ObTxySvs846admyZV599dW89dZb2WWXXZbY/m677ZYk2WSTTer3QE6cODGvvfZa/XkgNTU1efvtt8vx8CiDTp065bzzzsvcuXMzduzY7L333nn99dcXO6QjSf35gkk+8/DhUqm0RDh8lldeeSWXXHJJzjjjjNxzzz1ZddVVl80DYZlq1KhRmjVr9rnXf/rwv2Txc0KX9jDRm266KXfddVe6dOnypa4///zz07lz5/zlL3/5WoezU16fvE7q6uoyePDgvPrqq/X/towaNSo9evRIkuy111459dRT8/Of/zxNmzZN8tmHif71r39NVVVVNt544/rvOQ8++GB23333JB//Jn7w4MH16z/++OO59tprl8dDpQw++aVkXV1dmjdvnvPOOy/f/va3M2XKlCW+x0yePLn+F0OPPfZYLrrooiTJmmuumS5duuSJJ57IZptttlh4fPLeCIcddtjyf3AFsyI9l5/s0evUqVP69++f2bNnf+7Psl26dMkJJ5yQH/3oR+nSpctn/mzcrFmz9O7dO0ny4x//OFdccUX9dZ/8EvPxxx/Pm2++mf333z/Jx8H7/e9/P88880y22WabrL322kmSnj175rHHHlv6L+xyJAa/IXbYYYdUV1dn6tSpuffee7/0P4LdunXL+PHj8+abb6Zfv371u88/8ckPh//6w1ddXV1uvvnm+kO83nvvvay11lqZNGnS13swLBdNmzbND3/4w0yYMCHjx4/PsGHDFgu/z9KqVavMmTNnsWUzZszIaqut9m/v75prrknTpk3zyCOP5JxzzlnsmyYrjnbt2uW2225LqVRa7P/3K664Iuuvv/4yu59+/frlkUceyaWXXppBgwYt9fVt27ZNknz7299eZrNQPo0aNcppp52W3r17Z/jw4enTp08eeeSRPP/887nllltSKpUye/bs3H///Yud9/5po0aNytSpU+vfXGLu3Lm5/fbb62OQlcunfym5tEql0hKXP3mDqX/3iyrKY0V6Lv/1/PJSqZTKysrP/Vn2e9/7Xn74wx/mwQcfzGWXXZZnn312iSPeGjVqVP/vZF1d3WLb/+Tn5tra2uy5554ZOHBgkmTevHmpra3No48+uthj/OQohxWRcwa/QXr37p3rrrsurVq1yne+850vddtPYvDVV1/N97///aW6TYcOHXLbbbcl+XivT8+ePfPhhx+mcePGqamp+dLzs/z16tUrN954Y1ZfffX6H7K/SIcOHTJ69Oj65/e1117L888/n6233vrf3vaT3/qfc845efrppzNq1KivNTvlsf3222ettdbKtddeW/8P7yOPPJK77rorG2+88TK9rwEDBuTOO+9c4pdPS3s93wyVlZU57bTTMnTo0IwcOTIdOnTIww8/nAkTJuTBBx/Msccem9tvv/1zbz99+vRMnDgxY8eOzYQJEzJhwoSMHj06jz32mKNRWEyHDh1y5513Jvn4EL6//OUvDgX9hirHczlmzJgkyf3335+NNtooq6yyyuf+LLvffvtl3rx56devX/r165cXXnghSRb7GffDDz/MhAkTknx8TnPHjh2XuM/27dvn/vvvz4wZM1IqlTJo0KDcfPPN2W677fK///u/mTZtWurq6jJu3Liv9djKacXNVJbQp0+fdO7cORdeeOGXvu0666yTVVdd9Uv9jzZw4MD6E2eT5NJLL03Lli2z5ZZb5tprr82QIUNy6qmnfulZWH622267zJkzJz/5yU+Wav0DDjggb7/9dnr16lV/OOGQIUOy5ppr1q+zzTbb1P/9P/7jP5b42JPVVlstF198cY4//vhsu+222WCDDZbNg2GZqKioyNChQ3PxxRenR48eqayszBprrJHf/OY3S7UH+LPOGezevXuOPvroJdbdZJNN0rt371xyySW58cYbv/T1fHN07Ngx22yzTe6+++4lTlU4+OCDc/311+fVV1/9zNvec8896dSpU9ZZZ536Zeutt146d+6cO+64w5sIUe/444/PoEGD0rNnz9TW1ubYY4/N5ptvvsRHIrHiK8dz+cYbb6RXr15p0aJF/WHln/ez7H/9139lwIABqayszCqrrFL/5jFdunRJr1696j9Sa/z48bnyyivTunXrJT7+KEm++93v5oQTTsihhx6aurq6fO9738vRRx+dZs2aZeDAgenXr1++9a1vLfNfti5LFaVP76cFAAAosM022ywvvfRSQ49RdvYMAgAAK7S+fftm9uzZSyw/8MADl/oIKJZkzyAAAEABeQMZAACAAhKDAAAABSQGAQAACsgbyADAl3TBBRfkySefTJK8+uqradu2bZo3b54kueOOO+r//u9Mnjw5PXv2zDPPPFO2WQHg84hBAPiSBg4cWP/3zp07Z8iQIdliiy0acCIA+PLEIAAsI7/+9a/zxz/+MY0bN84GG2yQs846K1VVVXn33XczaNCgTJkyJaVSKb17986RRx7Z0OMCUHDOGQSAZWDUqFF55JFHcuedd2bMmDHZZJNNMmDAgCTJqaeemvbt22fMmDH5/e9/n3vvvTd//OMfG3hiAIpODALAMvDwww+nT58+WWWVVZIkhxxySB577LHMmTMnTz/9dA4++OAkyaqrrpo+ffrk4YcfbshxAUAMAsCyUFdXl4qKisUu19TUJElKpdIS635yHQA0FDEIAMvAbrvtllGjRmX+/PlJkhEjRmSHHXbIqquumq222iq33nprkmTOnDkZPXp0dt5554YcFwC8gQwALAv77rtvpk6dmv322y91dXVZf/31M2TIkCTJkCFDct555+Wuu+7KRx99lJ49e6ZPnz6ZMmVKA08NQJFVlD597AoAAAArPYeJAgAAFJAYBAAAKCAxCAAAUEBiEAAAoIDEIAAAQAGJQQAAgAISgwAAAAUkBgEAAAro/wNwYte6FWAX7QAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1080x720 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set(rc={'figure.figsize':(15,10)})\n",
    "g = sns.barplot(data=results,\n",
    "                x='Tool', y='Correct', hue='Dataset',\n",
    "                ci = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3oAAAJPCAYAAAAqk6bSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAABC+ElEQVR4nO3deYBdg93/8c8kk5kIiQRZqNZSJbXE3qYpEaoispGklpAFjSgajceapdZYqyoqJRpbUVsJIUVRbR+JXe2KWqNkkSD7MjO/P/zM0xQ1ItckJ6/XP51z77nnfO9IXe+cc88pq6mpqQkAAACF0aC+BwAAAGD5EnoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYMrre4AvY9asuamudhtAAABg1dKgQVlatFj9M59fqUOvurpG6AEAAPwHp24CAAAUjNADAAAomJX61M1PU1W1JLNmTc+SJYvqe5SVRnl5RVq0aJmGDQv3xwEAAFZJhfsv+1mzpqdx4yZZffU2KSsrq+9xVng1NTWZO/fDzJo1Peuss259jwMAACwHhTt1c8mSRVl99WYir47Kysqy+urNHAEFAIACKVzoJRF5X5DfFwAAFEshQw8AAGBVJvQAAAAKpnAXY1lefvWr8/L3vz+ZJHn99Vez7rrrpbKycZLk0ksvr/25FMaNuzQffPB+jjnmhM9cp0+f7jnjjHPStu3mJZsDAABYOQm9z/Cznx1X+3OfPt1z8slniCoAAGCl4NTNL2D+/Pk5/fSfZ/Dgg7P//r1yyCEH5c03X8/ChQty0EH75pZbbkqS3HHH+PTvv18WLFiQZ599JkceOSiDBg1Ir15dc9ZZp9V5f4888lD699+vdnn27NnZc89d8+GHHy739wYAABSH0PsCHnrowTRt2jSXXnpFrr/+lnz725vnD3+4MZWVjXPqqWdm3LhLMnny/2bs2N/kjDPOSePGjXPTTb/PoYcOzmWXXZVrrrkpDz7417z44gt12t+OO3438+fPz4svPp8kuffeu9Ohw05p1qxZKd8mAACwknPq5hew6667Z7311s/NN1+fKVOm5MknH8+WW7ZLknzzm5vk4IMH5fjjh2b48FPyjW9smCQZMeLUTJ78YK6++vK88cbrWbhwYebPn1en/ZWVlaVr1x6ZOHFC2rbdPBMn3p4jjji6VG8PAAAoCEf0voBbb705Z599Who3bpwf/nDP7L5759TU1NQ+/9prr2attdbKc889W/vYkUcOyuTJD2aDDTbMwQcPyjrrtFzqNZ+na9ce+fOf78vLL/8js2fPybbbbr9c3xMAAFA8Qu8LeOSRyenSpXu6dds73/jGBnnwwb+luroqSfKXv9yfxx9/LFdeeX0effSh/PWvD2T27Nl58cXn85Of/DS77LJbpk2bmrffnpLq6uo677Nly1b59re3yLnnnpnu3XuW6q0BAAAFIvS+gP3375fbb78lAwbsnyOP/HE23bRtpkyZkqlT3815552VkSNPS4sWLTJ8+Ck599xRmT9/Xg46aGAOPfSg9Ou3b6655qpstdXWmTLlrS+03x499s7LL/8jXbp0K9E7AwAAiqSs5oucR7iCee+9OamuXnr8d999I23abFBPE628/N4AAGDl0aBBWdZee43PfN7FWOrBvHlzc8QRgz71uSZNmmTMmN9+xRMBAABFIvTqQZMmq+fKK6+r7zEAAICCKul39ObMmZNu3bplypQpn3juhRdeSK9evdK5c+cMHz48S5YsKeUoAAAAq4yShd5TTz2VAw44IK+//vqnPn/cccfl5z//ee6+++7U1NTkxhtvLNUoAAAAq5SShd6NN96Yk08+Oa1atfrEc2+//XYWLFiQbbbZJknSq1ev3HXXXaUaBQAAYJVSsu/ojRo16jOfmzZtWlq2bFm73LJly0ydOrUkczRt1jiNKxst9+0uWLg4sz9csNy3CwAA8GXVy8VYqqurU1ZWVrtcU1Oz1HJdfdrlRKdNa5Dy8v87UNm4slH6Hn/tsg36X1x37oGZX77oc9e7//57c9VVl6eqqio1NTXp0qVrDjpowHKZYfr06TnzzNNywQUXfeltNWjQIC1bNl0OU8GKYcniqpQ3aljfY5TMksWLUt6oor7HKJklixenvNHy/0s6AFhV1EvotWnTJtOnT69dnjFjxqee4vl5Pu0+etXV1VmypPpLz1gXn7ef6dOn5cILf5nLL78ma67ZPPPmzctRRx2W9df/RnbaaZcvvf8WLdbOeedduFzeb3V1daZPn/2ltwMripYtm+bM4TfX9xglM2xUn/zypMH1PUbJHHPWpf6dBAD/xQp5H72vfe1rqayszOOPP57tt98+t912Wzp27Fgfo5TU+++/nyVLlmTBggVZc82P7pE3YsQpqaioTJ8+3bPbbrvn0UcfTpKcdNLPs+mmbTNlylv5xS/OyocffpDKysYZOvS4bLpp27z77js588xTM2vWzDRu3DgnnDAyq6++en7608G5+eYJmTnzvZx33pmZOnVqGjRokMGDj8yOO343jz32SMaMGZ2ysrI0bdo0p5xyZpo3b16/vxgAAKCkSnp7hf80aNCgPPPMM0mSX/ziFznrrLOy5557Zt68eenfv/9XOcpX4lvf2jQ777xL9t23ZwYN6p8xY0anqqo666//9SRJ48ar5Yorrsuhhx6eUaNOSZKMGnVyjjhiSC6//Nocf/zwnHzysCTJ+eefnV122S2/+92NOeSQw3LVVeOW2teFF/4iXbv2yOWXX5Ozz/5lzjvvzMybNzdXXTUuxx13UsaN+1123PG7eemlF7/S3wEAAPDVK/kRvfvvv7/258suu6z257Zt2+bmm4t7WtXHjj32pAwYcGgeeeShPPLI5AwefHBOPvn0JEmPHr2SJDvt1DGjRp2SadOm5oUXns+ZZ55W+/r58+fngw/ez9///kROOeWjC9x873s75Xvf2ynvvPOv2vUee+yRvPHGG/ntby9NkixZsiRvvz0lO+3UMcOGHZedd94lO++8S3bcsf1X9dYBAIB6Ui+nbq4qJk3638yfPy8/+MEe6dq1R7p27ZHbb781d9xxW5KkYcP/u1BETU11qqurU1FRmSuvvK728WnTpqZZszXTsGH5v61bk9dffy2NGzeufayqqjqjR/8mzZqtmeSj7z22aNEi3/rWZvn+9ztm0qS/ZcyY0enU6bkMGHBoqd86AABQj77SUzdXNY0bN84ll1xce+StpqYmL7/8Ur71rc2SJPfdd3eS5C9/+XM22GCjtGmzbtZf/+u5++6JSZJHH30oRx55WJJkm222zb333pMkeeyxh3PuuUvfvmL77XfILbfclCR57bVX07//flm4cEEGDRqQefPmZt99+2bfffs6dRMAAFYBhT+it2Dh4lx37oEl2e7n2W67HXLIIYNy/PE/y5IlS5Ik3/3u9zJw4I9zzz1/zDPPPJU77rg9q63WOMOHn5IkOfnkM3LeeWfmuuuuTnl5o5x22pkpKyvL0KHH55xzzsitt978/y/GMmKpfQ0denzOPXdUBgzYPzU1NRk58rQ0abJ6Bg8+MqNGnZqGDRumSZMmn3gdAABQPGU1NTU1n7/aiunTbq/w7rtvpE2bDepporrr06d7Lrro0qy77nr1PUqSlef3BnXl9gorN7dXAID/7vNur+DUTQAAgIIp/KmbK6qbb55Q3yMAAAAF5YgeAABAwQg9AACAgnHqJl9KizUrUl5RWd9jlMyiJYtSUV5R32OURNWiRWlYUcz3BgCwqhN6fCnlFZV5/Nwf1/cYJbP98b/NwCuOru8xSuLKgy/MxP4H1/cYJbPX1VfU9wgAAPWm8KFXqiNOSxYtzKwPFn3+ekuW5Nprr8o99/wxZWVlqaqqSpcu3dKv38EpKytb7nMBAAAUPvRKdcRp++N/m+TzQ+/888/JrFnv5ZJLrkjTpk0zd+6cDBt2XFZffY307r3vcp8LAACg8KFXn6ZNm5p77pmYW2/9Y5o2bZokWX31NXLMMSfktdf+mZkz38t5552ZqVOnpkGDBhk8+MjsuON3M27cpZkxY3reeuvNTJ36brp165kBAw7NxIkT8uSTj2f48FOSJEcddVgOOeSwrL/+13PaaSMzf/78NGhQlqOPPi5bbrlVPb5zAACgPgm9Enrhheey4YYbp1mzZks9vsEGG2aDDTbMySeflK5de2SnnXbJjBkzcsQRh+bKK69LkrzyyssZM+a3mTNndvbdd+/06vXZR//uuOO2dOiwU/r27Z+HHpqUp5/+u9ADAIBVmNArsX//Ht6f/3xvrrrq8lRXV6WiojLvvPN23njjjfz2t5cm+ej7fG+/PSVJst12O6RRo0Zp0WKtNGvWLHPnzvnMfeyww3cyfPjxeemlf6RDh52cEgoAAKs4oVdCm222eV5//dXMnTsnq6++Rnbddffsuuvueeedf+WnPx2cqqrqjB79mzRrtmaSZMaMGWnRokX++tcHUvFvl70vKytLTU1N7f9+rKpqSZKkXbttcs01N2bSpP/Nfffdk4kTJ+RXvxrz1b5ZAABgheGG6SXUpk2bdO68V84445TMnj07yUdH7SZN+lsaNGiQ7bffIbfcclOS5LXXXk3//vtl4cIFn7m9NddsnjfeeC01NTX517/eziuvvJIkGTPmwtx99x/TpUu3DB16Ql566R+lf3MAAMAKyxG9Evuf/zkx119/bYYMGZzq6urMmzcv2267fX7xi9Fp0qRJzj13VAYM2D81NTUZOfK0NGmy+mdua4cdvpM777wtBxzQOxtssEHatdsmSdK793459dQRmThxQho0aJARI079it4dAACwIip86C1ZtPD/3wph+W+3Lho0aJC+ffulb99+n/r8uef+6hOPHXro4KWWb755Qu3Po0ad96nbGTNm+b9HAABg5VT40Pvopuaff787AACAovAdPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACiYwl91s9malamsqFju2124aFE+/KBut1gAAAD4KhU+9CorKjLwiqOX+3avPPjCJJ8feu+8868ccECvbLjhxks9fs45v0zr1m2Wad8vvvh8xo//Q048cWSOOuqwHHLIYdluux2WaVsAAEDxFD70VgTrrNMyV1553XLbXtu2m+fEEzdfbtsDAACKRejVk1dffSUXXHBe5s+fn1mzZqZfv4HZe+8+GTfu0kyd+m7eeuvNvP/+rPTvf0gef/zRPP/8s9lkk01z6qln5sknH8/ll4/Nr389tnZ7p58+MltvvV169NgnSXLUUYflJz8Zki222LK+3iIAAFBPhN5XYMaM6Rk4sG/t8h577Jnp06dnwIBDs8MO38nbb0/JwIF9s/fefZIkr776z1xyyeV55pmncvTRP8lVV12fr3/9GznooB/llVde/tR9dO3aM+PGXZoePfbJu+++k/fff1/kAQDAKkrofQU+7dTNqqqqPPzw5Pzud1fkn/98JfPnz6t9bscdv5vy8vK0abNu1l57nWy00ca125k9+8NP3ce2226fGTOm5513/pW7756YPffcq3RvCAAAWKG5vUI9+fnPT8xf//rnbLjhRjnssCOWeq68/P/6u2HDhnXaXllZWbp06ZZ777079913T/bcs+tynRcAAFh5CL168uijj+THPz48O+/cKQ89NCnJR0f5vowuXbpl/Pg/pHXrNllnnZbLY0wAAGAlVPhTNxcuWvT/b4Ww/Lf7ZRxyyKD85Cc/TmVlRb75zW9l3XXXyzvv/OtLbbN16zZp3bpNunTp/qW2AwAArNwKH3of3dS8/m5svu666+Xmmyd84vH99z8o++9/0CceP/TQwZ/52n+/yubH9837+LGampq8996MzJz5XnbeeZflNj8AALDycepmQTzwwH0ZOPCADB58ZCoqKup7HAAAoB4V/ojeqmLXXXfPrrvuXt9jAAAAKwBH9AAAAApG6AEAABSM0AMAACgY39EDAAC+kDWbVaSisrK+xyiZRQsX5oMPv9zt1Opb4UOvedOKNGq8/P8QLl6wMO/P/vx/+PPmzctvfnNRHnlkcho3Xi2rr756DjnksOyww3f+6+smTpyQJ598PMOHn7KcJgYAgOWjorIyvzxp8OevuJI65qxLkwi9FVqjxpWZ2P/g5b7dva6+Ivmc0KupqckJJwzNt761aa655qY0atQoL730Yo477mc5+eQzau+FBwAAsDwVPvTq05NPPp533303o0dfkrKysiTJppu2zYABh+aqq8bl8svHplmzNfPaa//MaaedlX/+85VcddW4rL76GmnTpk1WW61JkuSFF57L6NG/zMKFC7Lmms1z3HHDst56X8tRRx221Ou/9a3N6vPtAgAAKwgXYymhF198Pm3bfrs28j62zTbb5oUXnk+SfPObm+T3v78lLVqsld/8ZnQuvviyXHLJ5Zk3b16SZPHixTn77DNy8smjcvnl12b//Q/KOeeMqt3Wx68XeQAAwMcc0SupslRVVX3i0cWLl+Tj9tt88y2TJM8881S23LJd1lpr7STJHnt0yeOPP5q33noj//rXlJx44jG1r587d27tzx+/HgCAFcuazVZLRaX/3KZ++JNXQltssWVuvvn6LFmyJOXl//erfvbZp9O27eapqqpK5f+/WlFZWVlqav7vtQ0bNkySVFVVZ731vpYrr7zu/y9XZdasmbXrVRb4akcAACuzisrynDn85voeoySGjepT3yPwOZy6WUJbb71tNtrom7nwwvOzZMmSJMmLL76Qq68elwEDDl1q3Xbttslzzz2d6dOnpbq6Ovff/6ckyQYbbJgPP/wwTz31ZJLkzjtvzymnDP9q3wgAALBScUSvxM4889yMHTsm/frtm4YNG6Zp02YZOfL0bLfdDrn88rG166211tr52c+Oy89+dkQaN14tG264UZKkoqIip59+di688BdZtGhRmjRZPSNGnFpfbwcAAFgJFD70Fi9Y+NGtEEqw3bqorGycn/70mPz0p8d84rlf/3rsUsu77rp7dt1190+st+WW7XLZZVd/7usBAACSVSD03p+96HPvdwcAAFAkvqMHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABVP4q26u2Wy1VFQu/7e5aOGSfPDh/OW+XQAAgC+r8KFXUVmeM4ffvNy3O2xUnzqtN3funFxyycX5+98fT8OG5WnatGmOOmpo5s6dk8svH+teeAAAwHLn1M0Sqq6uzrHHHp1mzZrliiuuy5VXXpeDDx6UY48dkg8//KC+xwMAAAqq8Ef06tMTTzyWqVPfzaGHDk6DBh819Xbb7ZBhw36eefP+77TPJ598PGPHjsnChQsye/acDBkyNDvv3Cn33HNXrrvu6jRo0CDrrbdeRo48PR988H5OO21k5s+fnwYNynL00cdlyy23qq+3CAAArICEXgm99NI/8q1vbVobeR/73vd2yhNPPFa7/Ic/3JATTxyZDTbYMI8//mguvPAX2XnnTrnsst9k7Ngr0qLFWrn44gvz5puv529/+0s6dNgpffv2z0MPTcrTT/9d6AEAAEsReiXUoEFZKioqP3e9kSNPz6RJf8uf/3xvnnvumcyf/9HRvu9/f+f85CeHpmPHTtlll93yrW9tlvnz52f48OPz0kv/SIcOO6V3731L/TYAAICVjO/olVDbtpvnpZdeTE1NzVKPX3rpxUstH3nkoLzwwnPZbLO26d//kNr1f/azY3PGGeemadNmOf30kbn77olp126bXHPNjfnud7+X++67JyecMPQrez8AAMDKQeiV0NZbb5sWLdbK5ZePTVVVVZLk4YcnZ+LE2zNr1swkyYcffpC33nojhx56eNq3/37+9re/pLq6OkuWLMn++++T5s2bp1+/g7Pnnl3z0kv/yJgxF+buu/+YLl26ZejQE/LSS/+oz7cIAACsgAp/6uaihUvqfCuEL7rdz1NWVpazz/5lLrro/PTvv1/Ky8uz5prNc955F2bOnDlJkmbN1ky3bj3Tr9++KS8vz3bb7ZgFCxZk8eLFOfTQwfnZz45MZWVlWrRokeHDT8miRYty6qkjMnHihDRo0CAjRpy63N8bAACwcit86NX3Tc2bN2+ekSNP/9TntttuhyTJT396TH7602NqHz/22BOTJD/84Z754Q/3/MTrxoz5bQkmBQAAisKpmwAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBgCn97hTWbVaSisnK5b3fRwoX54MNFn7veO+/8Kwcc0CsbbrhxysqSxYuXZJ111smwYSfniCN+nIsuujTrrrve527niScey+WXj82vfz12eYwPAAAUWOFDr6KyMr88afBy3+4xZ12a5PNDL0nWWadlrrzyutrliy66IBdffOFynwkAACBx6ma92G67HfLaa/9MklxxxWU5+OC+OeCAXnnuuWczZcpb6dWra6qrq5N8dCTvf/5nyFKvP+qow/LEE48l+eiIYZ8+3ZMko0adkvPPPyeDBx+cAw/sk7/85c8ZNuy47Ltvz1x00QVf4TsEAADqk9D7ii1ZsiQPPHBfttiiXZJkww03zhVXXJc+ffbL73//u6y//tez3npfy5NPPp4kueuuO7PXXt3qvP0ZM6bn0kuvyEEHDcxZZ52a4447KVdeeV0mTBifOXPmlOQ9AQAAKxah9xWYMWN6Bg7sm4ED+2bAgP1TU1OTn/zkqCRJx46dkiQbbfTNfPDB+0mSrl175O67J2bBggV5/PFHs9NOu9R5X+3bd0iStG7dJhtt9M20aLFWmjRZPc2aNcvs2R8u1/cFAACsmAr/Hb0VwX9+R+/fNWzYsPbnmpqaJMmuu+6esWPH5M9/vjff+973U/kfF5MpKyur/XnJkiVLPdeoUaNP3TYAALDqcERvBdS4ceO0b98hY8eOSZcu3T/x/JprNq/9jt/f/vbAVzobAACw4hN6K6gf/GCPrL766tliiy0/8dyBB/bPrbfenEMOOTALFy6sh+kAAIAVWeFP3Vy0cOH/vxXC8t9uXay77nq5+eYJn/rcvz++3XY7ZLvtdkiSVFVV5dFHH063bj0/9flvf3uLXHPNTbXPHXzwoCTJ8OGnfOr6/7kvAACg2Aofeh/d1Lxu97tbUfz4x/2y5prNc845v6zvUQAAgJVQ4UNvZXTFFZ9+4RYAAIC68B09AACAgilk6H18mwLqxu8LAACKpXChV15ekblzPxQvdVRTU5O5cz9MeXlFfY8CAAAsJ4X7jl6LFi0za9b0zJnzfn2PstIoL69IixYt63sMAABgOSlc6DVsWJ511lm3vscAAACoN4U7dRMAAGBVJ/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwZQ09CZMmJC99tore+yxR6699tpPPP/cc8+ld+/e6dGjRwYPHpwPP/ywlOMAAACsEkoWelOnTs0FF1yQ6667LuPHj88NN9yQV155Zal1Ro0alSFDhuT222/PRhttlHHjxpVqHAAAgFVGyUJv0qRJad++fZo3b54mTZqkc+fOueuuu5Zap7q6OnPnzk2SzJ8/P40bNy7VOAAAAKuM8lJteNq0aWnZsmXtcqtWrfL0008vtc6JJ56YQw45JGeeeWZWW2213HjjjV9oH2uvvcZymRWAFU/Llk3rewQAVmEr++dQyUKvuro6ZWVltcs1NTVLLS9YsCDDhw/PlVdemXbt2uWKK67ICSeckLFjx9Z5H++9NyfV1TXLdW6+mJX9/wDAimv69Nn1PQLAl+K/k1ZuK/rnUIMGZf/1wFfJTt1s06ZNpk+fXrs8ffr0tGrVqnb5pZdeSmVlZdq1a5ck2W+//fLII4+UahwAAIBVRslCr0OHDpk8eXJmzpyZ+fPn55577knHjh1rn99ggw3y7rvv5tVXX02S3Hfffdlqq61KNQ4AAMAqo2SnbrZu3TpDhw5N//79s3jx4vTp0yft2rXLoEGDMmTIkGy11VY566yz8rOf/Sw1NTVZe+21c+aZZ5ZqHAAAgFVGyUIvSbp3757u3bsv9dhll11W+/Muu+ySXXbZpZQjAAAArHJKesN0AAAAvnpCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAVT0qtuAgDAZ2netCKNGlfW9xhQSEIPAIB60ahxZSb2P7i+xyiZva6+or5HYBXm1E0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKJjy+h4AAIBP12zNylRWVNT3GMBKSOgBAKygKisqMvCKo+t7jJK58uAL63sEKCynbgIAABSMI3ol1rRZ4zSubFTfYwAAAKsQoVdijSsbpe/x19b3GCVz3bkH1vcIAKzCWqxZkfKKyvoeA2CFI/QAgJVWeUVlHj/3x/U9Rslsf/xv63sEYCXlO3oAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACiYkobehAkTstdee2WPPfbItdde+4nnX3311fTr1y89evTIoYcemg8++KCU4wAAAKwSShZ6U6dOzQUXXJDrrrsu48ePzw033JBXXnml9vmampr85Cc/yaBBg3L77bfn29/+dsaOHVuqcQAAAFYZJQu9SZMmpX379mnevHmaNGmSzp0756677qp9/rnnnkuTJk3SsWPHJMnhhx+eAw88sFTjAAAArDJKFnrTpk1Ly5Yta5dbtWqVqVOn1i6/+eabWWeddTJs2LDss88+Ofnkk9OkSZNSjQMAALDKKC/Vhqurq1NWVla7XFNTs9TykiVL8sgjj+Saa67JVlttlV/96lc5++yzc/bZZ9d5H2uvvcZynRmAFUfLlk3rewQAVmEr++dQyUKvTZs2eeyxx2qXp0+fnlatWtUut2zZMhtssEG22mqrJEm3bt0yZMiQL7SP996bk+rqmuUzcIms7H9AAOrL9Omz63sEVgI+Z4FSWdE/hxo0KPuvB77qdOrmT3/600yaNOkL7bhDhw6ZPHlyZs6cmfnz5+eee+6p/T5ekmy77baZOXNmXnzxxSTJ/fffny222OIL7QMAAIBPqlPo/fCHP8yYMWPSuXPnjBs3Lu+///7nvqZ169YZOnRo+vfvn7333jvdunVLu3btMmjQoDzzzDNp3LhxLr744owYMSJdu3bNww8/nBNPPPHLvh8AAIBVXp1O3ezRo0d69OiRf/7zn/nDH/6QPn36ZNttt02/fv3Srl27z3xd9+7d071796Ueu+yyy2p/3nrrrXPzzTcv4+gAAAB8mjpfdbO6ujpvvPFGXn/99VRVVWXttdfOKaecktGjR5dyPgAAAL6gOh3Ru+CCC3LLLbfk61//evr27ZsLL7wwjRo1yrx587Lrrrt+4YuoAAAAUDp1Cr2ZM2fmsssuS9u2bZd6vEmTJjn//PNLMhgAAADLpk6nbh555JG5/vrrkySvvvpqjjjiiEyfPj1JstNOO5VuOgAAAL6wOoXeiSeemI033jhJ8rWvfS3f+c53MmzYsJIOBgAAwLKpU+jNmjUr/fv3T5JUVlZm4MCBtUf0AAAAWLHUKfSqqqoyderU2uUZM2akpqamZEMBAACw7Op0MZaBAwdm7733zs4775yysrJMmjQpxx9/fKlnAwAAYBnUKfT69OmTLbfcMg899FAaNmyYQw89NJtuummpZwMAAGAZ1Cn0kqRNmzbp3LlzampqUlVVlQcffDDf//73SzkbAAAAy6BOoXfhhRdm7NixH72gvDyLFi3KJptskgkTJpR0OAAAAL64Ol2M5bbbbsuf//zndO7cOXfffXfOOuusbLLJJqWeDQAAgGVQp9Bba6210qpVq2y88cZ58cUXs/fee+ell14q9WwAAAAsgzqFXnl5ed58881svPHGeeyxx7JkyZIsXLiw1LMBAACwDOoUeocffnhGjhyZTp065U9/+lM6deqU9u3bl3o2AAAAlkGdLsayZMmSXHXVVUmS8ePH54033shmm21W0sEAAABYNnU6onfBBRfU/rzaaqulbdu2KSsrK9lQAAAALLs6HdHbdNNN85vf/CY77LBDmjRpUvv4FltsUbLBAAAAWDZ1Cr2nnnoqTz31VG666abax8rKynLfffeVbDAAAACWTZ1C7/777y/1HAAAACwndQq9K6644lMfP/jgg5frMAAAAHx5dQq9f785+qJFi/Loo4/me9/7XsmGAgAAYNnVKfTOOuuspZanTp2a4cOHl2QgAAAAvpw63V7hP7Vu3Tpvv/328p4FAACA5eALf0evpqYmzz77bNZee+2SDQUAAMCy+8Lf0UuSddddN8cff3xJBgIAAODLqfN39B599NHsuOOOef/99/PYY4+lTZs2pZ4NAACAZVCn7+hdcMEFGT16dJJkwYIFGTt2bMaMGVPSwQAAAFg2dQq9++67L5dffnmSpE2bNrnmmmsyceLEkg4GAADAsqlT6C1evDiNGjWqXW7UqFHKyspKNhQAAADLrk7f0dtuu+3yP//zP+nTp0/Kysoyfvz4bL311qWeDQAAgGVQp9AbOXJkRo8enbPOOivl5eXp0KFDjjzyyFLPBgAAwDKoU+g1adIkP/jBD3LiiSfWXnVztdVWK/VsAAAALANX3QQAACgYV90EAAAoGFfdBAAAKBhX3QQAACiYOh3RGzlyZNZZZ52cddZZOffcc7P22mtn+PDhpZ4NAACAZVDnq26edNJJpZ4FAACA5aBOoffkk09m7NixmTdvXmpqalJdXZ0pU6bkgQceKPF4AAAAfFF1OnVzxIgR2XbbbTNnzpx07949a6yxRvbYY49SzwYAAMAyqNMRvbKyshx22GGZNWtWNt5443Tv3j29e/cu9WwAAAAsgzod0Vt99dWTJN/4xjfy8ssvp3HjxmnQoE4vBQAA4CtWpyN67dq1y89+9rMcffTRGTx4cF5//fWUl9fppQAAAHzF6nRYbtiwYRk4cGA22mijDBs2LNXV1Tn//POTJK+//nop5wMAAOALqvN39LbZZpskSadOndKpU6fa54YOHZpbb721FLMBAACwDL70F+1qamqWxxwAAAAsJ1869MrKypbHHAAAACwnLp0JAABQMEIPAACgYIQeAABAwdQ59ObOnZskWbRoUd57773axzfccMPlPhQAAADLrk6hN3HixOyzzz5Jkrfffjtdu3bN/fffnyS54IILSjcdAAAAX1idQu+SSy7J1VdfnSTZaKONcuutt+aiiy4q6WAAAAAsmzqFXnV1ddq0aVO7vO6666a6urpkQwEAALDs6hR6a621Vq6//vosWbIkVVVVufnmm7POOuuUejYAAACWQZ1C77TTTsuNN96YrbfeOu3atcuNN96YU045pcSjAQAAsCzK67LSK6+8kltuuSUffPBBGjZsmDXWWKPUcwEAALCM6nRE7+Mra6655poiDwAAYAVXpyN6m266aX7zm99khx12SJMmTWof32KLLUo2GAAAAMumTqH31FNP5amnnspNN91U+1hZWVnuu+++kg0GAADAsqlT6H18c3QAAABWfHX6jt7cuXNz2mmnZcCAAXn//ffz85//PHPnzi31bAAAACyDOoXeGWeckaZNm+a9995LZWVl5syZk5///Oelng0AAIBlUKfQe+GFFzJ06NCUl5dntdVWyy9+8Yu88MILpZ4NAACAZVCn0GvQYOnVqqqqPvEYAAAAK4Y6XYxlxx13zHnnnZcFCxbkb3/7W6655pp897vfLfVsAAAALIM6HZY79thj06RJk9TU1OT0009P27Ztc/zxx5d6NgAAAJZBnY7ovfXWW7n//vvz1ltvJUkef/zxvPfee1lvvfVKOhwAAABfXJ2O6J100kn50Y9+lKeeeip///vf07lz5wwfPrzUswEAALAM6hR68+fPz/77759GjRqloqIi/fr1y4wZM0o9GwAAAMugTqG38cYb54knnqhdfumll7L++uuXbCgAAACWXZ2+o/evf/0r/fr1y2abbZby8vI8//zzadmyZbp3754kmTBhQkmHBAAAoO7qFHrHHntsqecAAABgOalT6H3nO98p9RwAAAAsJ3X6jh4AAAArD6EHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBlDT0JkyYkL322it77LFHrr322s9c74EHHshuu+1WylEAAABWGeWl2vDUqVNzwQUX5JZbbklFRUX233//fPe7380mm2yy1HozZszIOeecU6oxAAAAVjklO6I3adKktG/fPs2bN0+TJk3SuXPn3HXXXZ9Yb8SIETnqqKNKNQYAAMAqp2RH9KZNm5aWLVvWLrdq1SpPP/30UutcffXV2XzzzbP11lsv0z7WXnuNLzUjACuuli2b1vcIAKzCVvbPoZKFXnV1dcrKymqXa2pqllp+6aWXcs899+TKK6/Mu+++u0z7eO+9OamurvnSs5bSyv4HBKC+TJ8+u75HYCXgcxYolRX9c6hBg7L/euCrZKdutmnTJtOnT69dnj59elq1alW7fNddd2X69Onp3bt3DjvssEybNi19+/Yt1TgAAACrjJKFXocOHTJ58uTMnDkz8+fPzz333JOOHTvWPj9kyJDcfffdue222zJ27Ni0atUq1113XanGAQAAWGWULPRat26doUOHpn///tl7773TrVu3tGvXLoMGDcozzzxTqt0CAACs8kr2Hb0k6d69e7p3777UY5dddtkn1lt//fVz//33l3IUAACAVUZJb5gOAADAV0/oAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwZQ09CZMmJC99tore+yxR6699tpPPH/vvfemZ8+e6dGjR4444oh88MEHpRwHAABglVCy0Js6dWouuOCCXHfddRk/fnxuuOGGvPLKK7XPz5kzJ6ecckrGjh2b22+/PZtttlkuuuiiUo0DAACwyihZ6E2aNCnt27dP8+bN06RJk3Tu3Dl33XVX7fOLFy/OySefnNatWydJNttss7zzzjulGgcAAGCVUbLQmzZtWlq2bFm73KpVq0ydOrV2uUWLFvnhD3+YJFmwYEHGjh2b3XffvVTjAAAArDLKS7Xh6urqlJWV1S7X1NQstfyx2bNn58gjj0zbtm2zzz77fKF9rL32Gl96TgBWTC1bNq3vEQBYha3sn0MlC702bdrkscceq12ePn16WrVqtdQ606ZNy6GHHpr27dtn2LBhX3gf7703J9XVNV961lJa2f+AANSX6dNn1/cIrAR8zgKlsqJ/DjVoUPZfD3yV7NTNDh06ZPLkyZk5c2bmz5+fe+65Jx07dqx9vqqqKocffni6dOmS4cOHf+rRPgAAAL64kh3Ra926dYYOHZr+/ftn8eLF6dOnT9q1a5dBgwZlyJAheffdd/P888+nqqoqd999d5Jkyy23zKhRo0o1EgAAwCqhZKGXJN27d0/37t2Xeuyyyy5Lkmy11VZ58cUXS7l7AACAVVJJb5gOAADAV0/oAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAghF6AAAABSP0AAAACkboAQAAFIzQAwAAKBihBwAAUDBCDwAAoGCEHgAAQMEIPQAAgIIRegAAAAUj9AAAAApG6AEAABSM0AMAACgYoQcAAFAwQg8AAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMEIPAACgYIQeAABAwQg9AACAgilp6E2YMCF77bVX9thjj1x77bWfeP6FF15Ir1690rlz5wwfPjxLliwp5TgAAACrhJKF3tSpU3PBBRfkuuuuy/jx43PDDTfklVdeWWqd4447Lj//+c9z9913p6amJjfeeGOpxgEAAFhllJdqw5MmTUr79u3TvHnzJEnnzp1z11135aijjkqSvP3221mwYEG22WabJEmvXr0yevTo9O3bt877aNCgbHmPXRLrtFi9vkcoqYpma9f3CCW1zhpr1fcIJbPaOsX+Z7dm8yb1PUJJNWte7H9+K8u/46l/PodWbj6LVl4+h+rX581XVlNTU1OKHV966aWZN29ehg4dmiS56aab8vTTT+f0009Pkjz55JM599xz8/vf/z5J8sYbb+Swww7L3XffXYpxAAAAVhklO3Wzuro6ZWX/V5k1NTVLLX/e8wAAACybkoVemzZtMn369Nrl6dOnp1WrVp/5/IwZM5Z6HgAAgGVTstDr0KFDJk+enJkzZ2b+/Pm555570rFjx9rnv/a1r6WysjKPP/54kuS2225b6nkAAACWTcm+o5d8dHuFSy+9NIsXL06fPn0yaNCgDBo0KEOGDMlWW22VF198MSNGjMicOXOyxRZb5KyzzkpFRUWpxgEAAFgllDT0AAAA+OqV9IbpAAAAfPWEHgAAQMEIPQAAgIIRegAAAAUj9GAFctddd6VXr17p0aNHunfvnt/+9rdJktGjR+exxx5bpm1edNFFueiiiz71uWuvvTY9e/ZMjx490rNnz4wfP772uZNOOilvv/12nfczZcqU7Lbbbss0IwDL35QpU7LZZpvlwQcfXOrx3XbbLVOmTPnM1yyvf5ffeOONueOOO2qX//73v2fAgAHp0aNHunXrllNOOSULFiz4r9v498+wzTbbbJnm+PfP0OHDh+eZZ55Zpu3AykbowQpi6tSpOeecczJu3Ljcfvvtuf766zNx4sTcd999efTRR1NVVbVc9/fUU0/lpptuyg033JDbb789l19+eX71q1/lxRdfTJI8/PDDcVFegJVbo0aNMnLkyMyZM+cr3/cTTzyRRYsWJUlefPHFHHXUUTnmmGNy++23Z/z48ampqcnIkSNLPse/f4aOGjUqW221Vcn3CSsCoQcriFmzZmXx4sW1f7u5+uqr5+yzz84LL7yQZ599NiNGjMg//vGPPPLIIznggAOyzz775Ac/+EHuvffeJMmJJ56Yk08+Ob169Urnzp2XOjr39NNPZ//998+uu+5a+zej06dPT01NTebPn58kWXvttTN69Oi0aNEiY8eOzbRp03LYYYdl1qxZS/3t78MPP5x+/folSZ5//vnss88+2WeffXLxxRcnSebMmZPvfve7tf9RMWXKlOy1116l/wUC8AmtWrVKhw4dcs455yz1eFVVVUaMGJH99tsvP/jBD3LEEUfUfv4sWLAgQ4cOTbdu3dK3b9/MmjUrSXLNNdfkRz/6Ubp165Z99tknr776apKPjhD+6le/Sp8+fdK1a9c8++yzmTRpUu6///6MHj06f/vb3zJu3Lj07t07W2+9dZKkvLw8xx13XHbfffckyYwZMzJ48OB07949++yzT/76179+5nuaO3duTjjhhPTq1Ss9e/asPWq4cOHCDBs2LJ07d063bt0yceLEjB8/fqnP0H79+uXhhx9OklxyySXZa6+90r1795x99tmpqqrKlClTsvfee+e4445Lt27dMmDAgLz//vvL7x8IfIWEHqwg2rZtmx/84AfZfffd06dPn5x33nmprq7OUUcdlS233DJnnHFGNttss1xzzTU544wzcuutt+aMM87IhRdeWLuNt956KzfccEOuuuqqnHvuuZk+fXqS5L333svVV1+dP/zhDxk3blzmzJmTjh075mtf+1p23nnnHHTQQbnooovSvHnztG7dOocddlhatWqVsWPHpkWLFp858wknnJBjjz02t956a9Zff/0kyRprrJFOnTrlrrvuSpKMHz8+e++9d+l+cQD8VyeeeGL+93//d6lTOKdNm5ZGjRrlhhtuyJ/+9KfMnj07f/nLX5IkM2fOzMEHH5w77rgj66yzTiZOnJg5c+bk3nvvze9+97vccccd6dSpU6699tra7TVv3jw333xz9t9//1x66aXp0KFDdttttwwZMiQ777xzXnjhhWyxxRZLzbXGGmukc+fOSZLTTz897du3z4QJEzJ69OgMGzYsM2bM+NT385vf/CZbbLFFbrnlllx77bW55JJL8tZbb+V3v/td5s2blz/+8Y+54oorcvHFF2evvfZa6jP0Y3/5y19y//335w9/+ENuvfXWvPHGG7n++uuTfHT08eP336xZs0yYMGH5/IOAr5jQgxXIqaeemvvvvz8HHHBA/vWvf2XffffNPffcs9Q65513Xl5++eVcfPHFueKKKzJ37tza53r16pVGjRqlTZs22W677fL4448nSXbeeedUVFRkrbXWSosWLfLBBx+koqIiY8aMyZ133pkuXbrkueeeS48ePfL3v/+9TrPOnDkz06ZNy/e///3afX+sd+/eue2225Ikd9xxR3r27Pllfi0AfAlrrLFGTj/99KVO4Vx33XXTt2/fXHvttRk1alRef/31zJs3L8lHRwHbtWuXJNlkk00ya9asrLHGGjn//PNz55135vzzz8+f//zn2vWTjz5nkuRb3/rWpx4BKysrS2Vl5WfO+NBDD6VPnz5Jkq9//evZeuut89RTT33qupMmTcr111+fnj175sADD8y8efPy8ssv59FHH0337t3ToEGDtGzZMnfeeWcqKio+c39du3bNaqutlvLy8vTu3TuTJ09O8tEZLptvvnnt+/nggw8+c25YkQk9WEE88MADmThxYlq3bp3evXvnggsuyIgRI3LzzTcvtV7fvn3z9NNPZ8stt8zhhx++1HMNGzas/bm6ujrl5eVJUvu/yUcftjU1NRk/fnwmT56cDTbYIAceeGAuueSSDBgwoDbQ/tPH39dbsmTJUtv5tH3vuOOOmTZtWu65556sv/76ad269bL8SgBYTnbaaaelTuG8//77c+yxx6Zx48bp1atXdtxxx9p/p3/aZ8Y777yT/fbbL7Nnz07Hjh2zzz77LPUZ8HHElZWVfer+t9xyy09cBGXOnDk5/PDDs2jRok98J7ympuYzv5teXV2d8847L7fddltuu+223Hjjjdl5551TXl6+1P7feOON2u8Ifto2/tPHn2//HqT/+VkHKxOhByuIxo0b5/zzz6/9LlxNTU1eeOGFfPvb307Dhg1TVVWV999/P6+//nqOPvrodOzYMffdd99SH4R//OMfU1NTk7fffjtPP/10tt9++8/cX1VVVc4///zMnDkzSbJo0aK8/PLLtX+L+fE+k6RFixZ55ZVXkiT33Xdf7WPrrbdeHnjggSRZ6spqZWVl2XvvvXPGGWcsdaQPgPrz8Smc06ZNy6RJk9KlS5f07t07zZo1y8MPP/xfL/r1zDPPZIMNNsjAgQOz1VZb5d577/3ci4T9++fIwIED8/vf/z5PP/10kmTx4sU5++yzs8Yaa6SioiLt27ev/YvNt956K0888US22WabT91u+/bt8/vf/z7JR6eg9ujRI++880523HHHTJw4MTU1NXnvvfdy0EEHZdGiRUvN8e/buPPOO7NgwYIsWbIkf/jDH9K+ffs6/R5hZSH0YAXRvn37HHXUUTn88MPTuXPn7LnnnmnYsGGOPPLI7Lzzzjn55JPz6quv1n7ZvUuXLpk7d24WLFhQe/rMggUL0rt37wwePDinnXbaf/1+Xe/evbPnnnvmgAMOSJcuXdKzZ89svvnmtafOdOrUKYcddljeeuutDBkyJKNGjUrv3r3TtGnT2m2cd955+fWvf5299947b7755lLb79q1a+bPn1/7RXsA6tfHp3AuXrw4Q4cOzZ133pnu3bvn6KOPznbbbfeZt1xIku9///uprq7OXnvtlX322ScbbbTRf10/STp06JBLLrkkd911VzbbbLOcd955GTVqVHr06JEePXqksrIyZ5xxRpKPbnvw0EMPpXv37jniiCNyxhlnpFWrVp+63aOOOioLFiyovVjKcccdl2984xvp27dvmjRpkh49emTgwIEZOXJk1lhjjdrP0CeeeKJ2G7vuums6deqU3r17p2vXrllvvfVy0EEHLcNvFVZcZTWOR0MhnHjiifnOd76zQhxBq66uzu9///u89tprGTFiRH2PAwCwyin//FUAvpijjjoq77zzTsaNG1ffowAArJIc0QMAACgY39EDAAAoGKEHAABQMEIPAACgYIQeAKuMKVOm5Nvf/nZ69uyZnj17pnv37tl///0zceLEz33tr3/969x7770lmeuQQw6pvaclACwPrroJwCqlcePGue2222qX33777QwcODANGzZM586dP/N1Dz/8cDbZZJOSzPTggw+WZLsArLqEHgCrtK997WsZMmRIxo0bl0033TSnnXZa5s6dm+nTp6dt27b51a9+lZtvvjnPPvtszj333DRs2DCbbLLJp65XWVmZ0aNH509/+lMaNWqUFi1a5KyzzkqrVq3yz3/+M6NGjcr777+fqqqq9OvXL3369MlJJ52UJBkwYEDGjh2bddddt55/IwAUgdADYJXXtm3bvPTSS7nxxhuz9957p2fPnlm8eHF69eqVBx54IAceeGDuuuuuHHjggfnhD3+Yc84551PXa9euXa666qpMnjw5FRUVufzyy/P000+nU6dOGTJkSM4999xsscUWmT17dvbbb79ssskmOeuss3LLLbfkqquuylprrVXfvwoACkLoAbDKKysrS+PGjXPcccflwQcfzGWXXZbXX38906ZNy7x58z6x/met17p167Rt2zb77LNPOnbsmI4dO+Z73/teXnnllbz55psZNmxY7TYWLFiQ559/Pttss81X+E4BWFUIPQBWec8880w23XTTHHPMMamqqkqXLl3SqVOnvPPOO6mpqfnE+p+1XoMGDXLNNdfkmWeeyeTJk3PmmWdm5513Ts+ePdO0adOlvhs4Y8aMNG3a9Kt8mwCsQlx1E4BV2muvvZYxY8bkkEMOyf/+7//myCOPzF577ZUkeeqpp1JVVZUkadiwYZYsWZIkn7neiy++mG7duuWb3/xmBg8enIEDB+aZZ57JRhtttNRFYN55551069Ytzz777Ce2DQDLgyN6AKxSFixYkJ49eyZJGjRokMrKyhxzzDHp1KlThg4dmiOPPDJNmjTJGmuskR133DFvvvlmkmS33XbLL3/5yyxevPgz1/vRj36ULl26pHfv3mnSpEkaN26cESNGpKKiImPGjMmoUaPy29/+NkuWLMnRRx+d7bffPkmy5557pl+/frnooouy6aab1tvvBoDiKKv5tHNSAAAAWGk5dRMAAKBghB4AAEDBCD0AAICCEXoAAAAFI/QAAAAKRugBAAAUjNADAAAoGKEHAABQMP8Pm9QWNdWAa5gAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1080x720 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.set(rc={'figure.figsize':(15,10)})\n",
    "g = sns.barplot(data=results[results['Tool'] == 'VHMnet'],\n",
    "                x='Dataset', y='perc_accuracy', hue='Taxa_lvl',\n",
    "                ci = None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 ('HostRange')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "vscode": {
   "interpreter": {
    "hash": "ac3782f64e42e2ed2c7a5c6ad0e171638978992b65ec2d90108602644db05ad0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
